{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMv6eSQHmH7hhWKFuK1oU8i",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/dtabuena/Workshop/blob/main/RNA_Workshop/BMX_Ingest.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 43,
      "metadata": {
        "id": "e-U1uRiBKykm"
      },
      "outputs": [],
      "source": [
        "!pip install scanpy --quiet\n",
        "!pip install pybiomart --quiet\n",
        "!pip install python-igraph --quiet\n",
        "!pip install louvain --quiet\n",
        "!pip install pynndescent --quiet\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "import h5py\n",
        "import numpy as np\n",
        "import scipy as sci\n",
        "from matplotlib import pyplot as plt\n",
        "import scanpy as sc\n",
        "import tarfile\n",
        "import os\n",
        "import anndata as ad\n",
        "import pandas as pd\n",
        "import pybiomart\n",
        "import shutil\n",
        "from tqdm import tqdm\n",
        "import urllib.request\n",
        "from IPython.display import clear_output\n",
        "from matplotlib.pyplot import rc_context\n",
        "from scipy import stats as st\n",
        "import logging\n",
        "logger = logging.basicConfig(level=logging.INFO,format='%(asctime)s  %(message)s', datefmt='%Y/%m/%d %I:%M:%S %p',force=True)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def z_score(x,axis=-1):\n",
        "    x=np.array(x)\n",
        "    return (x-np.mean(x,axis=axis))/np.std(x,axis=axis)\n",
        "\n",
        "\n",
        "def publishable_plots(FS=6):\n",
        "    plt.rcParams.update({'font.size': FS,'axes.linewidth':.5,'figure.dpi':300,\n",
        "                         'xtick.major.width': 0.5,'ytick.major.width': 0.5,\n",
        "                         'figure.titlesize':FS,'axes.titlesize': FS,'xtick.labelsize': FS,\n",
        "                         'ytick.labelsize':FS,'axes.labelsize': FS,'legend.fontsize': FS,\n",
        "                         'figure.labelsize':FS})\n",
        "\n",
        "    import urllib.request\n",
        "    arial_link = 'https://raw.githubusercontent.com/dtabuena/Resources/main/Fonts/arial.ttf'\n",
        "    filename = './arial.ttf'\n",
        "    urllib.request.urlretrieve(arial_link, filename)\n",
        "    plt.rcParams.update({'font.family': 'arial'})\n",
        "    return None\n",
        "\n",
        "\n",
        "\n",
        "def pull_gene_annots(csv_loc='./mmusculus_coding_noncoding.csv',\n",
        "                     my_git='https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv',\n",
        "                     biomart_name='mmusculus',\n",
        "                     biomart_keys=[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\",\"peptide\"]):\n",
        "\n",
        "    logging.info('pulling gene annotations...')\n",
        "    if os.path.exists('./mmusculus_coding_noncoding.csv'):\n",
        "        logging.info( 'Use local copy of musmus')\n",
        "        annot_dd = pd.read_csv('./mmusculus_coding_noncoding.csv').set_index(\"external_gene_name\")\n",
        "    else:\n",
        "        try:\n",
        "            logging.info( 'attempting to pull mus mus from git...')\n",
        "            musmus_link = 'https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv'\n",
        "            filename = './mmusculus_coding_noncoding.csv'\n",
        "            urllib.request.urlretrieve(musmus_link, filename)\n",
        "            annot_dd = pd.read_csv(filename).set_index(\"external_gene_name\")\n",
        "        except:\n",
        "            logging.info('attempting to pull mus mus from biomart...')\n",
        "            annot = sc.queries.biomart_annotations(\"mmusculus\",[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\"],).set_index('ensembl_gene_id')\n",
        "            uniq_inds = list(set(list(annot.index)))\n",
        "            for r in tqdm(uniq_inds):\n",
        "                match_bool = annot.index.str.contains(r)\n",
        "                if np.sum(match_bool)>1:\n",
        "                    new_val ='__'.join(list(annot.loc[r,'transcript_biotype']))\n",
        "                    annot.at[r,'transcript_biotype']=new_val\n",
        "            annot['is_coding']= annot.transcript_biotype.str.contains('coding')\n",
        "            annot_dd = annot.drop_duplicates().set_index(\"external_gene_name\")\n",
        "            annot_dd.to_csv('./mmusculus_coding_noncoding.csv')\n",
        "\n",
        "    coding_list = annot_dd.index[ annot_dd['is_coding'] ].to_list()\n",
        "    return coding_list, annot_dd\n",
        "\n",
        "publishable_plots(6)\n",
        "sc.settings.verbosity = 'error'             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
        "sc.settings.set_figure_params(dpi=300, facecolor='white',fontsize=6,figsize=(1.5,1.5))\n",
        "pd.set_option('display.max_rows',100)"
      ],
      "metadata": {
        "id": "y5DpMQR0t-7H"
      },
      "execution_count": 44,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "def trim_key(k):\n",
        "    floxed_dict = {'GSM5106175_YH_KZ03_01':('E3fKI_Syn_Cre602_15m','GSM5106175_602_E3fKI_15_XX'),\n",
        "                   'GSM5106176_YH_KZ03_03':('E4fKI_Syn_Cre475_15m','GSM5106176_475_E4fKI_15_XX')}\n",
        "    for f in floxed_dict.keys():\n",
        "        if f in k: return floxed_dict[f][1]\n",
        "    k = k.replace('_raw_gene_bc_matrices_h5.h5',\"\")\n",
        "    return k\n",
        "\n",
        "def high_var_genes_dim_reduc(adata,min_mean = 0.25,max_mean = 4,min_disp=0.55,pca_list=[],to_plot=True):\n",
        "    ''' The gene expression matrices were then log-normalized with a scale factor of 10,000,\n",
        "    using the Seurat NormalizeData function57,58. Highly dispersed genes were selected using\n",
        "    the Seurat FindVariableGenes function57,58,filtering for an average expression range of\n",
        "    0.25–4 and a minimum dispersion of 0.55, resulting in a list of 2,197 genes.'''\n",
        "    adata.raw = adata\n",
        "    sc.pp.normalize_total(adata, target_sum=10000)\n",
        "    sc.pp.log1p(adata)\n",
        "    sc.pp.highly_variable_genes(adata, min_mean=min_mean, max_mean=max_mean, min_disp=min_disp)\n",
        "    with rc_context({'figure.figsize': (1.5, 1.5)}):\n",
        "        sc.pl.highly_variable_genes(adata)\n",
        "    plt.tight_layout()\n",
        "    num_hv =  str(np.sum(adata.var['highly_variable']))\n",
        "    logging.info(f'{num_hv} hv genes')\n",
        "\n",
        "\n",
        "    #### PCA\n",
        "    sc.tl.pca(adata, svd_solver='arpack',n_comps=100)\n",
        "    if to_plot and len(pca_list)>0:\n",
        "        fig,ax=plt.subplots(figsize=(1.5,1))\n",
        "        ax.plot(adata.uns['pca']['variance_ratio'][:35],'ok',markersize=1)\n",
        "        quiet_PCA_plots(adata,pca_list,pc_pairs=[(0,1),(2,3),(4,5),(6,7)])\n",
        "\n",
        "    return adata\n",
        "\n",
        "def quiet_PCA_plots(adata,key_list,figsize=(1.5, 1.5),pc_pairs=[(0,1)]):\n",
        "    fig,ax=plt.subplots(1*len(pc_pairs),len(key_list),figsize=(figsize[0]*len(key_list),figsize[1]*len(pc_pairs)))\n",
        "    for ip,pair in enumerate(pc_pairs):\n",
        "        if len(key_list) == 1: ax=[ax]\n",
        "        for key_ind,key in enumerate(key_list):\n",
        "            key_types = sorted(list(set( adata.obs[key] )))\n",
        "            for k in key_types:\n",
        "                is_k = adata.obs[key]==k\n",
        "                ax[ip,key_ind].scatter(adata.obsm['X_pca'][is_k,pair[0]],adata.obsm['X_pca'][is_k,pair[1]],s=2,marker='.',linewidth=0,edgecolors=None,label=k)\n",
        "                ax[ip,key_ind].set_xlabel(f'PC{pair[0]}')\n",
        "                ax[ip,key_ind].set_ylabel(f'PC{pair[1]}')\n",
        "            if len(key_types)<8: ax[ip,key_ind].legend(key_types,loc='best',markerscale=3)\n",
        "            ax[ip,key_ind].set_title(key)\n",
        "            plt.tight_layout()\n",
        "    return None\n",
        "\n",
        "def preprocess_andata10x(adata_og,pct_mito=0.25,min_genes=200,max_genes=2400,min_counts=100,max_counts=4500):\n",
        "\n",
        "\n",
        "    coding_list, _ = pull_gene_annots()\n",
        "    coding_list_lower=[str(g).lower() for g in coding_list]\n",
        "\n",
        "    adata_og.var['mt'] = adata_og.var_names.str.startswith('mt-')\n",
        "    adata_og.var['coding'] = [gene in coding_list for gene in adata_og.var_names]\n",
        "    sc.pp.calculate_qc_metrics(adata_og, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
        "\n",
        "    adata_filt = adata_og.copy()\n",
        "\n",
        "    adata_filt.var['mt'] = adata_filt.var_names.str.startswith('mt-')\n",
        "    adata_filt.var['coding'] = [gene.lower() in coding_list_lower for gene in adata_filt.var_names]\n",
        "    sc.pp.calculate_qc_metrics(adata_filt, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
        "    display(adata_filt.var.head())\n",
        "    adata_filt.var['not_mt'] =  np.logical_not( adata_filt.var['mt'])\n",
        "\n",
        "    logging.info('Filtering...')\n",
        "    adata_filt = adata_filt[adata_filt.obs.pct_counts_mt < pct_mito, :]\n",
        "    logging.info(str(np.sum(adata_og.obs.pct_counts_mt <pct_mito)) + f' cells with >{pct_mito}% mt-RNA removed')\n",
        "    adata_filt = adata_filt[:, adata_filt.var.coding]\n",
        "    logging.info(str(np.sum(np.logical_not(adata_og.var.coding))) + ' non coding genes removed')\n",
        "    sc.pp.filter_cells(adata_filt, min_genes=min_genes)\n",
        "    sc.pp.filter_cells(adata_filt, max_genes=max_genes)\n",
        "    sc.pp.filter_cells(adata_filt, min_counts=min_counts)\n",
        "    sc.pp.filter_cells(adata_filt, max_counts=max_counts)\n",
        "    fig,ax=plt.subplots(1,figsize=(1.5,1.5))\n",
        "    sc.pl.scatter(adata_filt, x='total_counts', y='n_genes_by_counts',ax=ax)\n",
        "    logging.info(str(len(adata_filt.obs))+' cells remaining')\n",
        "    return adata_filt,adata_og\n",
        "\n",
        "def umap_and_cluster(adata, n_neighbors=10, n_pcs=20,resolution=.6,plot_keys=['Cluster (nn)'],size = 1,to_plot=True):\n",
        "    sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=n_pcs,random_state=42)\n",
        "    sc.tl.louvain(adata,resolution=resolution,random_state=42)\n",
        "    sc.tl.paga(adata)\n",
        "    sc.tl.umap(adata,random_state=42)\n",
        "    adata.obs['Cluster (nn)']= adata.obs['louvain']\n",
        "    if to_plot:\n",
        "        with rc_context({'figure.figsize': (2.5, 2.5)}):\n",
        "            sc.pl.umap(adata,add_outline=False, legend_loc='on data', color=plot_keys,size=size)\n",
        "    return adata\n",
        "\n",
        "def explore_umap(adata_GABA,key_list=[],size=1,legend_loc=None):\n",
        "    with rc_context({'figure.figsize': (1.5,1.5)}):\n",
        "        sc.pl.umap(adata_GABA, legend_loc=legend_loc, color=key_list,vmin=0,size=size) # add_outline=True,\n",
        "        plt.tight_layout()\n",
        "\n",
        "def get_gene_cluster(adata,gene):\n",
        "    gene_bool = z_score(sc.get.obs_df(adata,gene))>2\n",
        "    gene_cluster_list = np.array(adata.obs['louvain'][gene_bool])\n",
        "    gene_clust = st.mode(gene_cluster_list)\n",
        "    return gene_clust[0][0]"
      ],
      "metadata": {
        "id": "RDGHgiy1rhOr"
      },
      "execution_count": 45,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#### Get BMX DATA\n",
        "subfolder='/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna/'\n",
        "parent_folder='//hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/'\n",
        "needed_files = ['/barcodes.tsv', '/genes.tsv', '/matrix.mtx']\n",
        "\n",
        "folder_list = [f.name for f in os.scandir(parent_folder) if f.is_dir()]\n",
        "folder_list = [f for f in folder_list if \"_\" in f]\n",
        "_=[logging.info(f) for f in folder_list if \"_\" in f]\n",
        "\n",
        "\n",
        "\n",
        "f = folder_list[0]\n",
        "f_list = [''.join([parent_folder,f,subfolder,n]) for n in needed_files] #\n",
        "logging.info(f_list)\n",
        "\n",
        "dest_folder = 'C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/'\n",
        "for f in folder_list:\n",
        "    for n in needed_files:\n",
        "        src = ''.join([parent_folder,f,subfolder,n])\n",
        "        try:\n",
        "            logging.info(''.join([dest_folder,f,'/']))\n",
        "            os.makedirs(''.join([dest_folder,f,'/']))\n",
        "        except: None\n",
        "        dst = ''.join([dest_folder,f,n])\n",
        "        logging.info(src)\n",
        "        logging.info(dst)\n",
        "        shutil.copyfile(src, dst)\n",
        "key_loc = 'https://raw.githubusercontent.com/dtabuena/Workshop/61a957d8c34d5c46c6ab7cc52e1f0330607aaf86/RNA_Workshop/UNOFFICIAL_apoE_BMX_key.csv'\n",
        "data_key = pd.read_csv(key_loc).set_index('Animal')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oLEfucd8i_he",
        "outputId": "456a765e-2535-439d-d197-59313befb83c"
      },
      "execution_count": 46,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "2023/09/22 12:01:15 PM  01_E3Ki3611\n",
            "2023/09/22 12:01:15 PM  02_E3Ki3607\n",
            "2023/09/22 12:01:15 PM  03_E3Ki3608\n",
            "2023/09/22 12:01:15 PM  04_E3Ki3612\n",
            "2023/09/22 12:01:15 PM  05_E3Ki3613\n",
            "2023/09/22 12:01:15 PM  06_E3Ki3614\n",
            "2023/09/22 12:01:15 PM  07_E4Ki2737\n",
            "2023/09/22 12:01:15 PM  08_E4Ki2738\n",
            "2023/09/22 12:01:15 PM  09_E4Ki2739\n",
            "2023/09/22 12:01:15 PM  10_E4Ki2710\n",
            "2023/09/22 12:01:15 PM  11_E4Ki2711\n",
            "2023/09/22 12:01:15 PM  12_E4Ki2712\n",
            "2023/09/22 12:01:15 PM  ['//hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/01_E3Ki3611/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv', '//hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/01_E3Ki3611/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv', '//hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/01_E3Ki3611/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx']\n",
            "2023/09/22 12:01:15 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/01_E3Ki3611/\n",
            "2023/09/22 12:01:15 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/01_E3Ki3611/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:15 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/01_E3Ki3611/barcodes.tsv\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/01_E3Ki3611/\n",
            "2023/09/22 12:01:16 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/01_E3Ki3611/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/01_E3Ki3611/genes.tsv\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/01_E3Ki3611/\n",
            "2023/09/22 12:01:16 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/01_E3Ki3611/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/01_E3Ki3611/matrix.mtx\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/02_E3Ki3607/\n",
            "2023/09/22 12:01:16 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/02_E3Ki3607/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/02_E3Ki3607/barcodes.tsv\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/02_E3Ki3607/\n",
            "2023/09/22 12:01:16 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/02_E3Ki3607/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/02_E3Ki3607/genes.tsv\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/02_E3Ki3607/\n",
            "2023/09/22 12:01:16 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/02_E3Ki3607/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:16 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/02_E3Ki3607/matrix.mtx\n",
            "2023/09/22 12:01:17 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/03_E3Ki3608/\n",
            "2023/09/22 12:01:17 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/03_E3Ki3608/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:17 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/03_E3Ki3608/barcodes.tsv\n",
            "2023/09/22 12:01:17 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/03_E3Ki3608/\n",
            "2023/09/22 12:01:17 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/03_E3Ki3608/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:17 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/03_E3Ki3608/genes.tsv\n",
            "2023/09/22 12:01:17 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/03_E3Ki3608/\n",
            "2023/09/22 12:01:17 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/03_E3Ki3608/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:17 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/03_E3Ki3608/matrix.mtx\n",
            "2023/09/22 12:01:18 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/04_E3Ki3612/\n",
            "2023/09/22 12:01:18 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/04_E3Ki3612/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:18 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/04_E3Ki3612/barcodes.tsv\n",
            "2023/09/22 12:01:18 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/04_E3Ki3612/\n",
            "2023/09/22 12:01:18 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/04_E3Ki3612/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:18 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/04_E3Ki3612/genes.tsv\n",
            "2023/09/22 12:01:18 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/04_E3Ki3612/\n",
            "2023/09/22 12:01:18 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/04_E3Ki3612/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:18 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/04_E3Ki3612/matrix.mtx\n",
            "2023/09/22 12:01:18 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/05_E3Ki3613/\n",
            "2023/09/22 12:01:18 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/05_E3Ki3613/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:18 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/05_E3Ki3613/barcodes.tsv\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/05_E3Ki3613/\n",
            "2023/09/22 12:01:19 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/05_E3Ki3613/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/05_E3Ki3613/genes.tsv\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/05_E3Ki3613/\n",
            "2023/09/22 12:01:19 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/05_E3Ki3613/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/05_E3Ki3613/matrix.mtx\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/06_E3Ki3614/\n",
            "2023/09/22 12:01:19 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/06_E3Ki3614/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/06_E3Ki3614/barcodes.tsv\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/06_E3Ki3614/\n",
            "2023/09/22 12:01:19 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/06_E3Ki3614/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/06_E3Ki3614/genes.tsv\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/06_E3Ki3614/\n",
            "2023/09/22 12:01:19 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/06_E3Ki3614/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:19 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/06_E3Ki3614/matrix.mtx\n",
            "2023/09/22 12:01:20 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/07_E4Ki2737/\n",
            "2023/09/22 12:01:20 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/07_E4Ki2737/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:20 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/07_E4Ki2737/barcodes.tsv\n",
            "2023/09/22 12:01:20 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/07_E4Ki2737/\n",
            "2023/09/22 12:01:20 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/07_E4Ki2737/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:20 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/07_E4Ki2737/genes.tsv\n",
            "2023/09/22 12:01:20 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/07_E4Ki2737/\n",
            "2023/09/22 12:01:20 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/07_E4Ki2737/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:20 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/07_E4Ki2737/matrix.mtx\n",
            "2023/09/22 12:01:20 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/08_E4Ki2738/\n",
            "2023/09/22 12:01:20 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/08_E4Ki2738/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:20 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/08_E4Ki2738/barcodes.tsv\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/08_E4Ki2738/\n",
            "2023/09/22 12:01:21 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/08_E4Ki2738/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/08_E4Ki2738/genes.tsv\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/08_E4Ki2738/\n",
            "2023/09/22 12:01:21 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/08_E4Ki2738/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/08_E4Ki2738/matrix.mtx\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/09_E4Ki2739/\n",
            "2023/09/22 12:01:21 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/09_E4Ki2739/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/09_E4Ki2739/barcodes.tsv\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/09_E4Ki2739/\n",
            "2023/09/22 12:01:21 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/09_E4Ki2739/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/09_E4Ki2739/genes.tsv\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/09_E4Ki2739/\n",
            "2023/09/22 12:01:21 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/09_E4Ki2739/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:21 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/09_E4Ki2739/matrix.mtx\n",
            "2023/09/22 12:01:22 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/10_E4Ki2710/\n",
            "2023/09/22 12:01:22 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/10_E4Ki2710/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:22 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/10_E4Ki2710/barcodes.tsv\n",
            "2023/09/22 12:01:22 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/10_E4Ki2710/\n",
            "2023/09/22 12:01:22 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/10_E4Ki2710/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:22 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/10_E4Ki2710/genes.tsv\n",
            "2023/09/22 12:01:22 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/10_E4Ki2710/\n",
            "2023/09/22 12:01:22 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/10_E4Ki2710/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:22 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/10_E4Ki2710/matrix.mtx\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/11_E4Ki2711/\n",
            "2023/09/22 12:01:23 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/11_E4Ki2711/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/11_E4Ki2711/barcodes.tsv\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/11_E4Ki2711/\n",
            "2023/09/22 12:01:23 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/11_E4Ki2711/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/11_E4Ki2711/genes.tsv\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/11_E4Ki2711/\n",
            "2023/09/22 12:01:23 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/11_E4Ki2711/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/11_E4Ki2711/matrix.mtx\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/12_E4Ki2712/\n",
            "2023/09/22 12:01:23 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/12_E4Ki2712/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//barcodes.tsv\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/12_E4Ki2712/barcodes.tsv\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/12_E4Ki2712/\n",
            "2023/09/22 12:01:23 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/12_E4Ki2712/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//genes.tsv\n",
            "2023/09/22 12:01:23 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/12_E4Ki2712/genes.tsv\n",
            "2023/09/22 12:01:24 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/12_E4Ki2712/\n",
            "2023/09/22 12:01:24 PM  //hive.gladstone.internal/Huang-Lab/kzalocus/bumex_snSeq/12_E4Ki2712/outs/raw_gene_bc_matrices/mm10-1.2.0_premrna//matrix.mtx\n",
            "2023/09/22 12:01:24 PM  C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/12_E4Ki2712/matrix.mtx\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "### IMPORT IMPORT DATA #########\n",
        "\n",
        "working_dir = 'C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/snBMX/'\n",
        "os.chdir(working_dir)\n",
        "folder_list = [f.path for f in os.scandir(working_dir) if f.is_dir() and '_' in f.name]\n",
        "adata_dict = {}\n",
        "for f in tqdm(folder_list):\n",
        "    a = sc.read_10x_mtx(f)\n",
        "    sample_ID = f.split('/')[-1]\n",
        "    adata_dict[sample_ID]=a\n",
        "    a=[]\n",
        "adata_ali = ad.concat(adata_dict,axis = 0,label=\"sample\",index_unique=\"_\")\n",
        "adata_dict = {}\n",
        "# adata_ali.var.index = [i.upper() for i in adata_ali.var.index]\n",
        "\n",
        "adata_ali.obs['treatment'] = [ data_key.loc[bc.split('_')[-1],'Treatment']  for bc in adata_ali.obs.index]\n",
        "adata_ali.obs['apoe_iso'] = [s.split('_')[-1][:4] for s in adata_ali.obs['sample']]\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nP6pcC7jwDS0",
        "outputId": "718e004b-b16e-48a1-c498-8113b0e0ffde"
      },
      "execution_count": 48,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████████████████████████████████████████████████████████████████████████████| 12/12 [01:04<00:00,  5.42s/it]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "##### Import Reference Adata\n",
        "os.chdir(\"C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021\")\n",
        "url = 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE167497&format=file'\n",
        "filename = './'+'zalocusky_indiv.tar'\n",
        "\n",
        "try:\n",
        "    for f in os.listdir('./indiv_animal_results'):\n",
        "        print(f)\n",
        "except:\n",
        "    urllib.request.urlretrieve(url, filename)\n",
        "    my_tar = tarfile.open(filename)\n",
        "    my_tar.extractall('./indiv_animal_results') # specify which folder to extract to\n",
        "    my_tar.close()\n",
        "    for f in os.listdir('./indiv_animal_results'):\n",
        "        print(f)\n",
        "\n",
        "### Read, Combine, and Sample Split multiple 10x's\n",
        "adata_dict = {}\n",
        "for f in tqdm( os.listdir('./indiv_animal_results') ):\n",
        "    a = sc.read_10x_h5('./indiv_animal_results/'+f)\n",
        "    a.var_names_make_unique()\n",
        "    sample_code = trim_key(f)\n",
        "    a.obs['age_bin'] = str(int(np.ceil( int(sample_code.split(\"_\")[3])/5)*5))+'m'\n",
        "    a.obs['apoe_iso'] = sample_code.split(\"_\")[2]\n",
        "    # a.obs['mouse_ID'] = sample_code.split(\"_\")[1]\n",
        "    # a.obs['well'] = sample_code.split(\"_\")[4]\n",
        "    # a.obs['GSM'] = sample_code.split(\"_\")[0]\n",
        "    adata_dict[sample_code.split(\"_\")[0]] = a\n",
        "adata_zalo = ad.concat(adata_dict,axis = 0,label=\"sample\",index_unique=\"_\")\n",
        "adata_zalo = adata_zalo[['fKI' not in t for t in adata_zalo.obs.apoe_iso], :]\n",
        "adata_dict = {}\n",
        "clear_output()\n"
      ],
      "metadata": {
        "id": "DxeiCvNAhlok",
        "outputId": "7b7474f1-8bed-477e-bab1-131a798152cd",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "GSM5106143_02_E4_14_H0_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106144_03_E3_10_H10_raw_gene_bc_matrics_h5.h5\n",
            "GSM5106145_04_E4_10_H11_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106146_05_E3_14_H12_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106147_06_E3_5_A01_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106148_07_E4_5_A02_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106149_08_E3_10_A03_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106150_09_E4_10_A04_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106151_10_E3_10_A05_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106152_11_E4_10_A06_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106153_12_E3_20_A07_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106154_13_E4_20_A08_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106155_14_E3_20_A09_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106156_15_E4_20_A10_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106157_16_E3_15_A11_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106158_17_E4_15_A12_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106159_18_E3_15_B01_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106160_19_E4_15_B02_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106161_20_E3_5_B03_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106162_21_E4_5_B04_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106163_22_E3_5_B05_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106164_23_E4_5_B06_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106165_24_E3_20_B07_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106166_25_E4_20_B08_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106167_26_E3_20_C01_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106168_27_E4_20_C02_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106169_28_E3_5_C03_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106170_29_E4_5_C04_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106171_30_E3_10_C05_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106172_31_E4_10_C06_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106173_32_E3_15_C07_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106174_33_E4_15_C08_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106175_YH_KZ03_01_raw_gene_bc_matrices_h5.h5\n",
            "GSM5106176_YH_KZ03_03_raw_gene_bc_matrices_h5.h5\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "  0%|                                                                                           | 0/34 [00:00<?, ?it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            "  3%|██▍                                                                                | 1/34 [00:00<00:23,  1.39it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            "  6%|████▉                                                                              | 2/34 [00:01<00:22,  1.42it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            "  9%|███████▎                                                                           | 3/34 [00:02<00:20,  1.49it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 12%|█████████▊                                                                         | 4/34 [00:03<00:25,  1.20it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 15%|████████████▏                                                                      | 5/34 [00:04<00:24,  1.17it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 18%|██████████████▋                                                                    | 6/34 [00:04<00:24,  1.16it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 21%|█████████████████                                                                  | 7/34 [00:05<00:22,  1.18it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 24%|███████████████████▌                                                               | 8/34 [00:06<00:21,  1.20it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 26%|█████████████████████▉                                                             | 9/34 [00:07<00:19,  1.30it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 29%|████████████████████████                                                          | 10/34 [00:07<00:17,  1.38it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 32%|██████████████████████████▌                                                       | 11/34 [00:08<00:15,  1.47it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 35%|████████████████████████████▉                                                     | 12/34 [00:09<00:14,  1.48it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 38%|███████████████████████████████▎                                                  | 13/34 [00:09<00:14,  1.49it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 41%|█████████████████████████████████▊                                                | 14/34 [00:10<00:13,  1.52it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 44%|████████████████████████████████████▏                                             | 15/34 [00:10<00:12,  1.51it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 47%|██████████████████████████████████████▌                                           | 16/34 [00:11<00:12,  1.48it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 50%|█████████████████████████████████████████                                         | 17/34 [00:12<00:11,  1.51it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 53%|███████████████████████████████████████████▍                                      | 18/34 [00:12<00:10,  1.53it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 56%|█████████████████████████████████████████████▊                                    | 19/34 [00:13<00:09,  1.55it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 59%|████████████████████████████████████████████████▏                                 | 20/34 [00:14<00:09,  1.53it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 62%|██████████████████████████████████████████████████▋                               | 21/34 [00:14<00:08,  1.53it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 65%|█████████████████████████████████████████████████████                             | 22/34 [00:15<00:07,  1.52it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 68%|███████████████████████████████████████████████████████▍                          | 23/34 [00:16<00:07,  1.53it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 71%|█████████████████████████████████████████████████████████▉                        | 24/34 [00:16<00:06,  1.59it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 74%|████████████████████████████████████████████████████████████▎                     | 25/34 [00:17<00:05,  1.51it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 76%|██████████████████████████████████████████████████████████████▋                   | 26/34 [00:18<00:05,  1.48it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 79%|█████████████████████████████████████████████████████████████████                 | 27/34 [00:18<00:04,  1.48it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 82%|███████████████████████████████████████████████████████████████████▌              | 28/34 [00:19<00:04,  1.41it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 85%|█████████████████████████████████████████████████████████████████████▉            | 29/34 [00:20<00:03,  1.38it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 88%|████████████████████████████████████████████████████████████████████████▎         | 30/34 [00:21<00:02,  1.38it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 91%|██████████████████████████████████████████████████████████████████████████▊       | 31/34 [00:21<00:02,  1.37it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 94%|█████████████████████████████████████████████████████████████████████████████▏    | 32/34 [00:22<00:01,  1.36it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            " 97%|███████████████████████████████████████████████████████████████████████████████▌  | 33/34 [00:23<00:00,  1.45it/s]C:\\Users\\dennis.tabuena\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
            "  utils.warn_names_duplicates(\"var\")\n",
            "100%|██████████████████████████████████████████████████████████████████████████████████| 34/34 [00:23<00:00,  1.43it/s]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "######### Reference Data Filter Dim Red #########\n",
        "adata_zalo_filt = preprocess_andata10x(adata_zalo)[0]\n",
        "adata_zalo_filt = high_var_genes_dim_reduc(adata_zalo_filt,min_mean = 0.25,max_mean = 4,min_disp=0.55)\n",
        "\n",
        "######### Import Data Filter Dim Red #########\n",
        "adata_ali_filt = preprocess_andata10x(adata_ali)[0]\n",
        "adata_ali_filt = high_var_genes_dim_reduc(adata_ali_filt,min_mean = 0.25,max_mean = 4,min_disp=0.55)\n"
      ],
      "metadata": {
        "id": "GLDW7FgBuBkt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "adata_ali = []\n",
        "adata_zalo = []"
      ],
      "metadata": {
        "id": "CDr0RnB-3uBk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "#### Parameter Match ####\n",
        "adata_zalo_filt.obs['treatment'] = None\n",
        "\n",
        "display(adata_zalo_filt.obs.head(2))\n",
        "display(adata_ali_filt.obs.head(2))\n",
        "\n",
        "display(adata_zalo_filt.var.head(2))\n",
        "display(adata_ali_filt.var.head(2))\n",
        "\n",
        "\n",
        "\n",
        "adata_zalo_filt.obs['Gad1_pos'] = z_score(sc.get.obs_df(adata_zalo_filt,'Gad1'))>0.0\n",
        "adata_zalo_filt.obs['Gad2_pos'] = z_score(sc.get.obs_df(adata_zalo_filt,'Gad2'))>0.0\n",
        "adata_zalo_filt.obs['Syn1_pos'] = z_score(sc.get.obs_df(adata_zalo_filt,'Syn1'))>-100\n",
        "is_gaba = np.logical_and( np.logical_or(adata_zalo_filt.obs['Gad1_pos'] , adata_zalo_filt.obs['Gad2_pos'] ), adata_zalo_filt.obs['Syn1_pos'])\n",
        "adata_zalo_filt.obs['Gabaergic'] = is_gaba\n",
        "adata_zalo_filt = adata_zalo_filt[ adata_zalo_filt.obs['Gabaergic'] ,:]\n",
        "\n",
        "adata_ali_filt.obs['Gad1_pos'] = z_score(sc.get.obs_df(adata_ali_filt,'Gad1'))>0.0\n",
        "adata_ali_filt.obs['Gad2_pos'] = z_score(sc.get.obs_df(adata_ali_filt,'Gad2'))>0.0\n",
        "adata_ali_filt.obs['Syn1_pos'] = z_score(sc.get.obs_df(adata_ali_filt,'Syn1'))>-100\n",
        "is_gaba = np.logical_and( np.logical_or(adata_ali_filt.obs['Gad1_pos'] , adata_ali_filt.obs['Gad2_pos'] ), adata_ali_filt.obs['Syn1_pos'])\n",
        "adata_ali_filt.obs['Gabaergic'] = is_gaba\n",
        "adata_ali_filt = adata_ali_filt[ adata_ali_filt.obs['Gabaergic'] ,:]"
      ],
      "metadata": {
        "id": "J7Qjoae3qecL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "adata_ali_filt.obs['Gad1_pos']=None\n",
        "display(adata_ali_filt.obs.head(2))"
      ],
      "metadata": {
        "id": "YQMMawXwKwGC",
        "outputId": "cd19aff1-b661-4fe2-f91b-b0ed6f31e0f4",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 112
        }
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                                     sample treatment apoe_iso  \\\n",
              "AAACCTGAGATAGGAG-1_01_E3Ki3611  01_E3Ki3611       BMX     E3Ki   \n",
              "AAACCTGGTACGCACC-1_01_E3Ki3611  01_E3Ki3611       BMX     E3Ki   \n",
              "\n",
              "                                n_genes_by_counts  total_counts  \\\n",
              "AAACCTGAGATAGGAG-1_01_E3Ki3611                700        1030.0   \n",
              "AAACCTGGTACGCACC-1_01_E3Ki3611                648         960.0   \n",
              "\n",
              "                                total_counts_mt  pct_counts_mt  n_genes  \\\n",
              "AAACCTGAGATAGGAG-1_01_E3Ki3611              0.0            0.0      638   \n",
              "AAACCTGGTACGCACC-1_01_E3Ki3611              0.0            0.0      600   \n",
              "\n",
              "                                n_counts Gad1_pos  \n",
              "AAACCTGAGATAGGAG-1_01_E3Ki3611     897.0     None  \n",
              "AAACCTGGTACGCACC-1_01_E3Ki3611     826.0     None  "
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample</th>\n",
              "      <th>treatment</th>\n",
              "      <th>apoe_iso</th>\n",
              "      <th>n_genes_by_counts</th>\n",
              "      <th>total_counts</th>\n",
              "      <th>total_counts_mt</th>\n",
              "      <th>pct_counts_mt</th>\n",
              "      <th>n_genes</th>\n",
              "      <th>n_counts</th>\n",
              "      <th>Gad1_pos</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>AAACCTGAGATAGGAG-1_01_E3Ki3611</th>\n",
              "      <td>01_E3Ki3611</td>\n",
              "      <td>BMX</td>\n",
              "      <td>E3Ki</td>\n",
              "      <td>700</td>\n",
              "      <td>1030.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>638</td>\n",
              "      <td>897.0</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AAACCTGGTACGCACC-1_01_E3Ki3611</th>\n",
              "      <td>01_E3Ki3611</td>\n",
              "      <td>BMX</td>\n",
              "      <td>E3Ki</td>\n",
              "      <td>648</td>\n",
              "      <td>960.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>600</td>\n",
              "      <td>826.0</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "adata_zalo_filt = umap_and_cluster(adata_zalo_filt, n_neighbors=10, n_pcs=25,resolution=.6,to_plot=False)\n",
        "explore_umap(adata_zalo_filt,['Cluster (nn)','Sst','Pvalb'],legend_loc='on data',size=3)\n",
        "explore_umap(adata_zalo_filt,['Cluster (nn)','Gad1','Gad2'],legend_loc='on data',size=3)\n",
        "explore_umap(adata_zalo_filt,['Cluster (nn)','Vip','Reln'],legend_loc='on data',size=3)\n",
        "explore_umap(adata_zalo_filt,['Cluster (nn)','Syn1','Slc1a3'],legend_loc='on data',size=3)\n",
        "\n",
        "\n",
        "# sc.pp.pca(adata_zalo)\n",
        "# sc.pp.neighbors(adata_zalo)\n",
        "# sc.tl.umap(adata_zalo)\n",
        "\n",
        "# sc.tl.ingest(adata, adata_ref, obs='louvain')"
      ],
      "metadata": {
        "id": "_O-lmvy8uIdk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "var_names = adata_zalo_filt.var_names.intersection(adata_ali_filt.var_names)\n",
        "adata_zalo_filt = adata_zalo_filt[:, var_names]\n",
        "adata_ali_filt = adata_ali_filt[:, var_names]"
      ],
      "metadata": {
        "id": "xZcrrJir1kFZ",
        "outputId": "034aefe8-3479-4428-cf38-877cffb6e95b",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 731
        }
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "error",
          "ename": "InvalidIndexError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[1;31mInvalidIndexError\u001b[0m                         Traceback (most recent call last)",
            "Cell \u001b[1;32mIn[12], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m var_names \u001b[38;5;241m=\u001b[39m adata_zalo_filt\u001b[38;5;241m.\u001b[39mvar_names\u001b[38;5;241m.\u001b[39mintersection(adata_ali_filt\u001b[38;5;241m.\u001b[39mvar_names)\n\u001b[0;32m      2\u001b[0m adata_zalo_filt \u001b[38;5;241m=\u001b[39m adata_zalo_filt[:, var_names]\n\u001b[1;32m----> 3\u001b[0m adata_ali_filt \u001b[38;5;241m=\u001b[39m adata_ali_filt[:, var_names]\n",
            "File \u001b[1;32m~\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1108\u001b[0m, in \u001b[0;36mAnnData.__getitem__\u001b[1;34m(self, index)\u001b[0m\n\u001b[0;32m   1106\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__getitem__\u001b[39m(\u001b[38;5;28mself\u001b[39m, index: Index) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAnnData\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m   1107\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Returns a sliced view of the object.\"\"\"\u001b[39;00m\n\u001b[1;32m-> 1108\u001b[0m     oidx, vidx \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_normalize_indices(index)\n\u001b[0;32m   1109\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m AnnData(\u001b[38;5;28mself\u001b[39m, oidx\u001b[38;5;241m=\u001b[39moidx, vidx\u001b[38;5;241m=\u001b[39mvidx, asview\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
            "File \u001b[1;32m~\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\anndata.py:1089\u001b[0m, in \u001b[0;36mAnnData._normalize_indices\u001b[1;34m(self, index)\u001b[0m\n\u001b[0;32m   1088\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_normalize_indices\u001b[39m(\u001b[38;5;28mself\u001b[39m, index: Optional[Index]) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Tuple[\u001b[38;5;28mslice\u001b[39m, \u001b[38;5;28mslice\u001b[39m]:\n\u001b[1;32m-> 1089\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _normalize_indices(index, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobs_names, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvar_names)\n",
            "File \u001b[1;32m~\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\index.py:33\u001b[0m, in \u001b[0;36m_normalize_indices\u001b[1;34m(index, names0, names1)\u001b[0m\n\u001b[0;32m     31\u001b[0m ax0, ax1 \u001b[38;5;241m=\u001b[39m unpack_index(index)\n\u001b[0;32m     32\u001b[0m ax0 \u001b[38;5;241m=\u001b[39m _normalize_index(ax0, names0)\n\u001b[1;32m---> 33\u001b[0m ax1 \u001b[38;5;241m=\u001b[39m _normalize_index(ax1, names1)\n\u001b[0;32m     34\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m ax0, ax1\n",
            "File \u001b[1;32m~\\AppData\\Local\\anaconda3\\Lib\\site-packages\\anndata\\_core\\index.py:95\u001b[0m, in \u001b[0;36m_normalize_index\u001b[1;34m(indexer, index)\u001b[0m\n\u001b[0;32m     93\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m positions  \u001b[38;5;66;03m# np.ndarray[int]\u001b[39;00m\n\u001b[0;32m     94\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:  \u001b[38;5;66;03m# indexer should be string array\u001b[39;00m\n\u001b[1;32m---> 95\u001b[0m     positions \u001b[38;5;241m=\u001b[39m index\u001b[38;5;241m.\u001b[39mget_indexer(indexer)\n\u001b[0;32m     96\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m np\u001b[38;5;241m.\u001b[39many(positions \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m0\u001b[39m):\n\u001b[0;32m     97\u001b[0m         not_found \u001b[38;5;241m=\u001b[39m indexer[positions \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m0\u001b[39m]\n",
            "File \u001b[1;32m~\\AppData\\Local\\anaconda3\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3904\u001b[0m, in \u001b[0;36mIndex.get_indexer\u001b[1;34m(self, target, method, limit, tolerance)\u001b[0m\n\u001b[0;32m   3901\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_method(method, limit, tolerance)\n\u001b[0;32m   3903\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_index_as_unique:\n\u001b[1;32m-> 3904\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_requires_unique_msg)\n\u001b[0;32m   3906\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(target) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m   3907\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m np\u001b[38;5;241m.\u001b[39marray([], dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mintp)\n",
            "\u001b[1;31mInvalidIndexError\u001b[0m: Reindexing only valid with uniquely valued Index objects"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "sc.tl.ingest(adata_ali_filt, adata_zalo_filt, obs='louvain')"
      ],
      "metadata": {
        "id": "bwVG29Ib1Ydy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "def high_var_genes_dim_reduc(adata,min_mean = 0.25,max_mean = 4,min_disp=0.55,pca_list=[],to_plot=True):\n",
        "    ''' The gene expression matrices were then log-normalized with a scale factor of 10,000,\n",
        "    using the Seurat NormalizeData function57,58. Highly dispersed genes were selected using\n",
        "    the Seurat FindVariableGenes function57,58,filtering for an average expression range of\n",
        "    0.25–4 and a minimum dispersion of 0.55, resulting in a list of 2,197 genes.'''\n",
        "    adata.raw = adata\n",
        "    sc.pp.normalize_total(adata, target_sum=10000)\n",
        "    sc.pp.log1p(adata)\n",
        "    sc.pp.highly_variable_genes(adata, min_mean=min_mean, max_mean=max_mean, min_disp=min_disp)\n",
        "    with rc_context({'figure.figsize': (1.5, 1.5)}):\n",
        "        sc.pl.highly_variable_genes(adata)\n",
        "    plt.tight_layout()\n",
        "    num_hv =  str(np.sum(adata.var['highly_variable']))\n",
        "    logging.info(f'{num_hv} hv genes')\n",
        "\n",
        "\n",
        "    #### PCA\n",
        "    sc.tl.pca(adata, svd_solver='arpack',n_comps=100)\n",
        "    if to_plot and len(pca_list)>0:\n",
        "        fig,ax=plt.subplots(figsize=(1.5,1))\n",
        "        ax.plot(adata.uns['pca']['variance_ratio'][:35],'ok',markersize=1)\n",
        "        quiet_PCA_plots(adata,pca_list,pc_pairs=[(0,1),(2,3),(4,5),(6,7)])\n",
        "\n",
        "    return adata\n",
        "\n",
        "def quiet_PCA_plots(adata,key_list,figsize=(1.5, 1.5),pc_pairs=[(0,1)]):\n",
        "    fig,ax=plt.subplots(1*len(pc_pairs),len(key_list),figsize=(figsize[0]*len(key_list),figsize[1]*len(pc_pairs)))\n",
        "    for ip,pair in enumerate(pc_pairs):\n",
        "        if len(key_list) == 1: ax=[ax]\n",
        "        for key_ind,key in enumerate(key_list):\n",
        "            key_types = sorted(list(set( adata.obs[key] )))\n",
        "            for k in key_types:\n",
        "                is_k = adata.obs[key]==k\n",
        "                ax[ip,key_ind].scatter(adata.obsm['X_pca'][is_k,pair[0]],adata.obsm['X_pca'][is_k,pair[1]],s=2,marker='.',linewidth=0,edgecolors=None,label=k)\n",
        "                ax[ip,key_ind].set_xlabel(f'PC{pair[0]}')\n",
        "                ax[ip,key_ind].set_ylabel(f'PC{pair[1]}')\n",
        "            if len(key_types)<8: ax[ip,key_ind].legend(key_types,loc='best',markerscale=3)\n",
        "            ax[ip,key_ind].set_title(key)\n",
        "            plt.tight_layout()\n",
        "    return None\n",
        "\n"
      ],
      "metadata": {
        "id": "g_T_taBqr8BM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "eXQx0PZLurI3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# adata_filt = umap_and_cluster(adata_filt, n_neighbors=35, n_pcs=25,resolution=.6,to_plot=False)\n",
        "# explore_umap(adata_filt,['Cluster (nn)','SST','PVALB'],legend_loc='on data',size=3)\n",
        "# explore_umap(adata_filt,['Cluster (nn)','GAD1','GAD2'],legend_loc='on data',size=3)\n",
        "# explore_umap(adata_filt,['Cluster (nn)','VIP','RELN'],legend_loc='on data',size=3)\n",
        "# explore_umap(adata_filt,['Cluster (nn)','SYN1','SLC1A3'],legend_loc='on data',size=3)"
      ],
      "metadata": {
        "id": "2bXA8hOEsUmL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# ############## GABAERGIC Filter #####################\n",
        "# adata_GABA = adata_filt.copy()\n",
        "# adata_GABA.obs['Gad1_pos'] = z_score(sc.get.obs_df(adata_GABA,'Gad1'))>0.0\n",
        "# adata_GABA.obs['Gad2_pos'] = z_score(sc.get.obs_df(adata_GABA,'Gad2'))>0.0\n",
        "# adata_GABA.obs['Syn1_pos'] = z_score(sc.get.obs_df(adata_GABA,'Syn1'))>-100\n",
        "# is_gaba = np.logical_and( np.logical_or(adata_GABA.obs['Gad1_pos'] , adata_GABA.obs['Gad2_pos'] ), adata_GABA.obs['Syn1_pos'])\n",
        "# adata_GABA.obs['Gabaergic'] = is_gaba\n",
        "# adata_GABA = adata_GABA[ adata_GABA.obs['Gabaergic'] ,:]\n",
        "\n",
        "\n",
        "\n",
        "# adata_GABA = umap_and_cluster(adata_GABA, n_neighbors=5, n_pcs=25,resolution=.6,to_plot=False)\n",
        "# explore_umap(adata_GABA,['Cluster (nn)','Sst','Pvalb'],legend_loc='on data',size=8)\n",
        "# # explore_umap(adata_GABA,['sample','treatment','apoe_iso'],size=8)\n",
        "# # explore_umap(adata_GABA,[],size=8)\n"
      ],
      "metadata": {
        "id": "JMgAXJatxzXX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Ld_EqGP6wPvG"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}