{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/dtabuena/Workshop/blob/main/RNA_Workshop/Gabaergic_Analysis.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "b294428a",
      "metadata": {
        "id": "b294428a",
        "scrolled": true,
        "tags": []
      },
      "outputs": [],
      "source": [
        "!pip install scanpy\n",
        "!pip install pybiomart\n",
        "!pip install python-igraph\n",
        "!pip install louvain\n",
        "!pip install pynndescent\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "16d3f9dc",
      "metadata": {
        "id": "16d3f9dc",
        "tags": []
      },
      "outputs": [],
      "source": [
        "import h5py\n",
        "import numpy as np\n",
        "import scipy as sci\n",
        "from matplotlib import pyplot as plt\n",
        "import scanpy as sc\n",
        "import tarfile\n",
        "import os\n",
        "import anndata as ad\n",
        "import pandas as pd\n",
        "import pybiomart\n",
        "from tqdm import tqdm\n",
        "import urllib.request\n",
        "from IPython.display import clear_output\n",
        "from matplotlib.pyplot import rc_context\n",
        "from scipy import stats as st\n",
        "os.chdir(\"C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021\")\n",
        "\n",
        "def publishable_plots(FS=6):\n",
        "    plt.rcParams.update({'font.size': FS,'axes.linewidth':.5,'figure.dpi':300,\n",
        "                         'xtick.major.width': 0.5,'ytick.major.width': 0.5,\n",
        "                         'figure.titlesize':FS,'axes.titlesize': FS,'xtick.labelsize': FS,\n",
        "                         'ytick.labelsize': FS,'axes.labelsize': FS,'legend.fontsize': FS})\n",
        "\n",
        "    import urllib.request\n",
        "    arial_link = 'https://raw.githubusercontent.com/dtabuena/Resources/main/Fonts/arial.ttf'\n",
        "    filename = './arial.ttf'\n",
        "    urllib.request.urlretrieve(arial_link, filename)\n",
        "    plt.rcParams.update({'font.family': 'arial'})\n",
        "    return None\n",
        "publishable_plots(6)\n",
        "sc.settings.verbosity = 'error'             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
        "sc.settings.set_figure_params(dpi=300, facecolor='white',fontsize=6,)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4514aef3",
      "metadata": {
        "id": "4514aef3",
        "tags": []
      },
      "outputs": [],
      "source": [
        "def trim_key(k):\n",
        "    floxed_dict = {'GSM5106175_YH_KZ03_01':('E3fKI_Syn_Cre602_15m','GSM5106175_602_E3fKI_15_XX'),\n",
        "                   'GSM5106176_YH_KZ03_03':('E4fKI_Syn_Cre475_15m','GSM5106176_475_E4fKI_15_XX')}\n",
        "    for f in floxed_dict.keys():\n",
        "        if f in k: return floxed_dict[f][1]\n",
        "    k = k.replace('_raw_gene_bc_matrices_h5.h5',\"\")\n",
        "    return k\n",
        "def query_capitilaziation(gene,adata):\n",
        "    try:\n",
        "        return adata.var.index[ [g.lower() for g in list(adata.var.index)].index(gene.lower()) ]\n",
        "    except:\n",
        "        return gene + ' not_found'\n",
        "def z_score(x,axis=-1):\n",
        "    x=np.array(x)\n",
        "    return (x-np.mean(x,axis=axis))/np.std(x,axis=axis)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "04bde4bf",
      "metadata": {
        "scrolled": true,
        "tags": [],
        "id": "04bde4bf"
      },
      "outputs": [],
      "source": [
        "os.chdir(\"C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021\")\n",
        "url = 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE167497&format=file'\n",
        "filename = './'+'zalocusky_indiv.tar'\n",
        "\n",
        "try:\n",
        "    for f in os.listdir('./indiv_animal_results'):\n",
        "        print(f)\n",
        "except:\n",
        "    urllib.request.urlretrieve(url, filename)\n",
        "    my_tar = tarfile.open(filename)\n",
        "    my_tar.extractall('./indiv_animal_results') # specify which folder to extract to\n",
        "    my_tar.close()\n",
        "    for f in os.listdir('./indiv_animal_results'):\n",
        "        print(f)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f115f8e4",
      "metadata": {
        "id": "f115f8e4",
        "scrolled": true,
        "tags": []
      },
      "outputs": [],
      "source": [
        "### Read, Combine, and Sample Split multiple 10x's\n",
        "\n",
        "adata_dict = {}\n",
        "for f in tqdm( os.listdir('./indiv_animal_results') ):\n",
        "    a = sc.read_10x_h5('./indiv_animal_results/'+f)\n",
        "    a.var_names_make_unique()\n",
        "    sample_code = trim_key(f)\n",
        "    a.obs['age_bin'] = str(int(np.ceil( int(sample_code.split(\"_\")[3])/5)*5))+'m'\n",
        "    a.obs['E_type'] = sample_code.split(\"_\")[2]\n",
        "    a.obs['mouse_ID'] = sample_code.split(\"_\")[1]\n",
        "    a.obs['well'] = sample_code.split(\"_\")[4]\n",
        "    a.obs['GSM'] = sample_code.split(\"_\")[0]\n",
        "    adata_dict[sample_code.split(\"_\")[0]] = a\n",
        "adata = ad.concat(adata_dict,axis = 0,label=\"Sample\",index_unique=\"_\")\n",
        "adata = adata[['fKI' not in t for t in adata.obs.E_type], :]\n",
        "adata_dict = {}\n",
        "clear_output()\n",
        "print('data_loaded.')\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "c5a6ad5b",
      "metadata": {
        "tags": [],
        "id": "c5a6ad5b"
      },
      "outputs": [],
      "source": [
        "def pull_gene_annots(csv_loc='./mmusculus_coding_noncoding.csv',\n",
        "                     my_git='https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv',\n",
        "                     biomart_name='mmusculus',\n",
        "                     biomart_keys=[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\"]):\n",
        "\n",
        "\n",
        "    if os.path.exists('./mmusculus_coding_noncoding.csv'):\n",
        "        print( 'Use local copy of musmus')\n",
        "        annot_dd = pd.read_csv('./mmusculus_coding_noncoding.csv').set_index(\"external_gene_name\")\n",
        "    else:\n",
        "        try:\n",
        "            print( 'attempting to pull mus mus from git...')\n",
        "            musmus_link = 'https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv'\n",
        "            filename = './mmusculus_coding_noncoding.csv'\n",
        "            urllib.request.urlretrieve(musmus_link, filename)\n",
        "            annot_dd = pd.read_csv('./mmusculus_coding_noncoding.csv').set_index(\"external_gene_name\")\n",
        "        except:\n",
        "            print('attempting to pull mus mus from biomart...')\n",
        "            annot = sc.queries.biomart_annotations(\"mmusculus\",[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\"],).set_index('ensembl_gene_id')\n",
        "            uniq_inds = list(set(list(annot.index)))\n",
        "            for r in tqdm(uniq_inds):\n",
        "                match_bool = annot.index.str.contains(r)\n",
        "                if np.sum(match_bool)>1:\n",
        "                    new_val ='__'.join(list(annot.loc[r,'transcript_biotype']))\n",
        "                    annot.at[r,'transcript_biotype']=new_val\n",
        "            annot['is_coding']= annot.transcript_biotype.str.contains('coding')\n",
        "            annot_dd = annot.drop_duplicates().set_index(\"external_gene_name\")\n",
        "            annot_dd.to_csv('./mmusculus_coding_noncoding.csv')\n",
        "\n",
        "    coding_list = annot_dd.index[ annot_dd['is_coding'] ].to_list()\n",
        "    return coding_list, annot_dd\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def preprocess_andata10x(adata_og,pct_mito=0.25,min_genes=500,max_genes=2400,min_counts=500,max_counts=4500):\n",
        "\n",
        "    print('pulling gene annotations...')\n",
        "    coding_list, _ = pull_gene_annots()\n",
        "    adata_og.var['mt'] = adata_og.var_names.str.startswith('mt-')\n",
        "    adata_og.var['coding'] = [gene in coding_list for gene in adata_og.var_names]\n",
        "    sc.pp.calculate_qc_metrics(adata_og, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
        "\n",
        "    adata_QC = adata_og.copy()\n",
        "\n",
        "    print('Filtering...')\n",
        "    adata_QC = adata_QC[adata_QC.obs.pct_counts_mt < pct_mito, :]\n",
        "    print(str(np.sum(adata_og.obs.pct_counts_mt <pct_mito)) + f' cells with >{pct_mito}% removed')\n",
        "    adata_QC = adata_QC[:, adata_QC.var.coding]\n",
        "    print(str(np.sum(np.logical_not(adata_og.var.coding))) + ' non coding genes removed')\n",
        "    sc.pp.filter_cells(adata_QC, min_genes=min_genes)\n",
        "    sc.pp.filter_cells(adata_QC, max_genes=max_genes)\n",
        "    sc.pp.filter_cells(adata_QC, min_counts=min_counts)\n",
        "    sc.pp.filter_cells(adata_QC, max_counts=max_counts)\n",
        "    fig,ax=plt.subplots(1,figsize=(1.5,1.5))\n",
        "    sc.pl.scatter(adata_QC, x='total_counts', y='n_genes_by_counts',ax=ax)\n",
        "\n",
        "    return adata_QC,adata_og\n",
        "\n",
        "adata_QC = preprocess_andata10x(adata)[0]\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def high_var_genes_dim_reduc(adata,min_mean = 0.25,max_mean = 4,min_disp=0.55):\n",
        "    ''' The gene expression matrices were then log-normalized with a scale factor of 10,000,\n",
        "    using the Seurat NormalizeData function57,58. Highly dispersed genes were selected using\n",
        "    the Seurat FindVariableGenes function57,58,filtering for an average expression range of\n",
        "    0.25–4 and a minimum dispersion of 0.55, resulting in a list of 2,197 genes.'''\n",
        "    adata.raw = adata\n",
        "    sc.pp.normalize_total(adata, target_sum=10000)\n",
        "    sc.pp.log1p(adata)\n",
        "    sc.pp.highly_variable_genes(adata, min_mean=min_mean, max_mean=max_mean, min_disp=min_disp)\n",
        "    with rc_context({'figure.figsize': (1.5, 1.5)}):\n",
        "        sc.pl.highly_variable_genes(adata)\n",
        "    plt.tight_layout()\n",
        "    print(np.sum(adata.var['highly_variable']),'hv genes')\n",
        "\n",
        "\n",
        "    #### PCA\n",
        "    sc.tl.pca(adata, svd_solver='arpack',n_comps=50)\n",
        "    fig,ax=plt.subplots(figsize=(1,1))\n",
        "    ax.plot(adata.uns['pca']['variance_ratio'][:20],'ok',markersize=1)\n",
        "    quiet_PCA_plots(adata,['E_type','age_bin','mouse_ID'])\n",
        "\n",
        "    return adata\n",
        "\n",
        "def quiet_PCA_plots(adata,key_list,figsize=(2,2)):\n",
        "    fig,ax=plt.subplots(1,len(key_list),figsize=(figsize[0]*len(key_list),figsize[1]))\n",
        "    if len(key_list) == 1: ax=[ax]\n",
        "    for key in key_list:\n",
        "        key_ind = key_list.index(key)\n",
        "        key_types = sorted(list(set( adata.obs[key] )))\n",
        "        for k in key_types:\n",
        "            is_k = adata.obs[key]==k\n",
        "            ax[key_ind].scatter(adata.obsm['X_pca'][is_k,0],adata.obsm['X_pca'][is_k,1],s=2,marker='.',linewidth=0,edgecolors=None,label=k)\n",
        "        if len(key_types)<8: ax[key_ind].legend(key_types,loc='best',markerscale=3)\n",
        "        ax[key_ind].set_title(key)\n",
        "        plt.tight_layout()\n",
        "    return None\n",
        "\n",
        "############## GABAERGIC Filter #####################\n",
        "adata_GABA = adata_QC.copy()\n",
        "age_dict = {'5m':'05m', '10m': '10m','15m': '15_20m','20m': '15_20m'}\n",
        "adata_GABA.obs['age_bin'] = [ age_dict[a] for a in adata_GABA.obs['age_bin'] ]\n",
        "\n",
        "adata_GABA.obs['Gad1_Pos'] = z_score(sc.get.obs_df(adata_GABA,'Gad1'))>0.5\n",
        "adata_GABA.obs['NeuN_pos'] = z_score(sc.get.obs_df(adata_GABA,'Rbfox3'))>0.5\n",
        "adata_GABA.obs['Syn1_pos'] = z_score(sc.get.obs_df(adata_GABA,'Syn1'))>0.5\n",
        "\n",
        "# is_gaba = np.logical_and(adata_GABA.obs['Gad1_pos'],adata_GABA.obs['Syn1_pos'])\n",
        "is_gaba = adata_GABA.obs['Gad1_Pos']\n",
        "adata_GABA.obs['Gabaergic'] = is_gaba\n",
        "adata_GABA = adata_GABA[ adata_GABA.obs['Gabaergic'] ,:]\n",
        "sc.pp.filter_genes(adata_GABA, min_counts=50)\n",
        "\n",
        "\n",
        "############## Dim Reduction #####################\n",
        "adata_GABA = high_var_genes_dim_reduc(adata_GABA)"
      ],
      "metadata": {
        "id": "cx-L5v6n2FGP"
      },
      "id": "cx-L5v6n2FGP",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f246d0a6",
      "metadata": {
        "scrolled": true,
        "tags": [],
        "id": "f246d0a6"
      },
      "outputs": [],
      "source": [
        "def umap_and_cluster(adata, n_neighbors=10, n_pcs=20,resolution=0.6,plot_keys=['Cluster (nn)'],size = 1):\n",
        "    sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=n_pcs,random_state=42)\n",
        "    sc.tl.louvain(adata,resolution=resolution)\n",
        "    sc.tl.paga(adata)\n",
        "    sc.tl.umap(adata,random_state=42)\n",
        "    adata.obs['Cluster (nn)']= adata.obs['louvain']\n",
        "    with rc_context({'figure.figsize': (2.5, 2.5)}):\n",
        "        sc.pl.umap(adata,add_outline=False, legend_loc='on data', color=plot_keys,size=size)\n",
        "    return adata\n",
        "\n",
        "adata_GABA = umap_and_cluster(adata_GABA)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "69e3bc29",
      "metadata": {
        "id": "69e3bc29"
      },
      "outputs": [],
      "source": [
        "def explore_umap(adata_GABA,key_list=[],size=1,legend_loc=None):\n",
        "    with rc_context({'figure.figsize': (1.5,1.5)}):\n",
        "        sc.pl.umap(adata_GABA, legend_loc=legend_loc, color=key_list,vmin=0,size=size) # add_outline=True,\n",
        "        plt.tight_layout()\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Gad1','Syn1'],legend_loc='on data')\n",
        "explore_umap(adata_GABA,['E_type', 'Apoe'])\n",
        "explore_umap(adata_GABA,['Sst' ,'Pvalb','Reln','Vip'])\n",
        "\n",
        "louvain = adata_GABA.obs['louvain']\n",
        "clusters= sorted(list(set(louvain)))\n",
        "sst_clust=str(st.mode([int(c) for c in clusters],keepdims=False)[0])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1eb43f97-5b2e-4e67-b3d1-75e78f11eefd",
      "metadata": {
        "scrolled": true,
        "tags": [],
        "id": "1eb43f97-5b2e-4e67-b3d1-75e78f11eefd"
      },
      "outputs": [],
      "source": [
        "#### Get Marker Genes\n",
        "age = adata_GABA.obs['age_bin']\n",
        "e_type = adata_GABA.obs['E_type']\n",
        "louvain = adata_GABA.obs['louvain']\n",
        "clusters= sorted(list(set(louvain)))\n",
        "ages  = sorted(list(set(age)))\n",
        "# ages = np.array(ages)[np.array([3, 0, 1, 2])]\n",
        "# ages = list(ages)\n",
        "e_types  =list(set(e_type))\n",
        "\n",
        "\n",
        "\n",
        "sc.tl.rank_genes_groups(adata_GABA, 'louvain', method='wilcoxon',key_added='maker_genes',pts=True,use_raw =True)\n",
        "# sc.tl.rank_genes_groups(adata_GABA, 'louvain', method='logreg',pts=True,use_raw =True)\n",
        "\n",
        "maker_genes_df = pd.DataFrame(adata_GABA.uns['maker_genes']['names'])\n",
        "maker_genes_df_LFC = pd.DataFrame(adata_GABA.uns['maker_genes']['logfoldchanges'])\n",
        "\n",
        "top_30 = maker_genes_df[:30]\n",
        "lfc_thresh=4\n",
        "\n",
        "marker_genes = [m for m in set(top_30[maker_genes_df_LFC[:30]>lfc_thresh].values.flatten()) if isinstance(m,str)]\n",
        "# display(marker_genes)\n",
        "marker_genes = marker_genes + ['Sst', 'Vip', 'Pvalb']\n",
        "\n",
        "\n",
        "import os\n",
        "import shutil\n",
        "path = './cluster_markers/'\n",
        "if not os.path.exists(path): os.makedirs(path)\n",
        "\n",
        "# sc.pl.stacked_violin(adata_GABA, marker_genes, groupby='louvain');\n",
        "\n",
        "for c in clusters:\n",
        "     # with rc_context({'figure.figsize': (4,1),'lines.linewidth': 0}): sc.pl.rank_genes_groups_violin(adata_GABA, groups=c, n_genes=16,size=0,scale='width')\n",
        "    mark_df = sc.get.rank_genes_groups_df(adata_GABA, group =c,key='maker_genes')\n",
        "    new_file = os.path.join(path,f'cluster{c}_markers.csv')\n",
        "    if os.path.exists(new_file): os.remove(new_file)\n",
        "    mark_df[:20].to_csv(new_file)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "591887f3-f155-4898-8cf8-e8297165e5b1",
      "metadata": {
        "tags": [],
        "id": "591887f3-f155-4898-8cf8-e8297165e5b1"
      },
      "outputs": [],
      "source": [
        "########## Calc CLUSTER-AGE DEGs\n",
        "\n",
        "adata_GABA.obs['age_type_clust'] =  ['_'.join([age[i],e_type[i],louvain[i]]) for i in range(len(age))]\n",
        "ages = sorted(list(set(adata_GABA.obs['age_bin'])))\n",
        "\n",
        "for c in tqdm(clusters):\n",
        "    for a in ages:\n",
        "            sc.tl.rank_genes_groups(adata_GABA,use_raw=False,groupby='age_type_clust',groups=[a+\"_E4_\"+c,a+\"_E3_\"+c],reference=a+\"_E3_\"+c,\n",
        "                                    key_added=a+\"_c\"+c+'_4wrt3', method='wilcoxon',tie_correct=True)\n",
        "\n",
        "clear_output()\n",
        "import shutil\n",
        "path ='./DEGs'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "\n",
        "\n",
        "clear_output()\n",
        "df_dict = {}\n",
        "for c in tqdm(clusters):\n",
        "    age_dict={}\n",
        "    for a in ages:\n",
        "        ai =ages.index(a)\n",
        "        key = a+\"_E4_c\"+c\n",
        "        df = sc.get.rank_genes_groups_df(adata_GABA, group=a+\"_E4_\"+c,key=a+\"_c\"+c+'_4wrt3',pval_cutoff=1,log2fc_min=-15, log2fc_max=15, )\n",
        "        new_file = os.path.join(path,c+\"c_\"+a+'_4wrt3'+'.csv')\n",
        "        if os.path.exists(new_file): os.remove(new_file)\n",
        "        df.to_csv(new_file)\n",
        "        age_dict[a]=df\n",
        "    df_dict[c]=age_dict\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d4d8a2fa",
      "metadata": {
        "id": "d4d8a2fa"
      },
      "outputs": [],
      "source": [
        "kcn_list = [n for n in adata_GABA.var.index if 'kcn' in str(n).lower()]\n",
        "kcn_list.sort()\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "09d093bd",
      "metadata": {
        "id": "09d093bd"
      },
      "outputs": [],
      "source": [
        "###### SST Cluster KCN DEGs #####\n",
        "path='./SST_CLUSTER_DEG/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sst_cluster = adata_GABA.copy()\n",
        "adata_sst_cluster = adata_sst_cluster[adata_sst_cluster.obs['louvain']==sst_clust,:]\n",
        "adata_sst_cluster.obs['age_geno'] = adata_sst_cluster.obs['age_bin'].str.cat(adata_sst_cluster.obs['E_type'],sep='_')\n",
        "display(adata_sst_cluster)\n",
        "\n",
        "# adata_sst_cluster_fig,ax_p=plt.subplots(1,figsize=(3,10))\n",
        "# sc.pl.dotplot(adata_sst_cluster, kcn_list, groupby='age_geno', title='SST Cluster (n=' +str(len(adata_sst_cluster)) +')',\n",
        "#              swap_axes=True,ax=ax_p,expression_cutoff=.2)\n",
        "# adata_sst_cluster_fig.savefig('./adata_sst_cluster_fig.svg',dpi=300)\n",
        "\n",
        "deg_df_dict={}\n",
        "for a in ages:\n",
        "        group_key = 'age_geno'\n",
        "        analysis_groups = [a+\"_E4\", a+\"_E3\"]\n",
        "        deg_group = analysis_groups[0]\n",
        "        reference_group = analysis_groups[1]\n",
        "        key_added = a+'_E4_wrt_E3'\n",
        "        sc.tl.rank_genes_groups(adata_sst_cluster,use_raw=False,groupby=group_key,groups=analysis_groups,reference=reference_group,\n",
        "                                key_added=key_added, method='wilcoxon',tie_correct=True)\n",
        "        df = sc.get.rank_genes_groups_df(adata_sst_cluster, group=deg_group,key=key_added,pval_cutoff=1,log2fc_min=-15, log2fc_max=15, )\n",
        "        deg_df_dict[a] = df.set_index('names')\n",
        "        new_file = os.path.join(path,'SST_KCN_'+key_added+'.csv')\n",
        "        df.to_csv(new_file)\n",
        "\n",
        "timecourse_df = pd.DataFrame({'kcn':kcn_list}).set_index('kcn')\n",
        "\n",
        "\n",
        "for k,v in deg_df_dict.items():\n",
        "    timecourse_df[k+'_pval'] = np.nan\n",
        "    for ch in timecourse_df.index:\n",
        "        try:\n",
        "            timecourse_df.at[ch,k+'_pval'] = deg_df_dict[k].loc[ch,'pvals']\n",
        "            timecourse_df.at[ch,k+'_LFC'] = deg_df_dict[k].loc[ch,'logfoldchanges']\n",
        "        except:\n",
        "            None\n",
        "\n",
        "\n",
        "lfc = timecourse_df.iloc[:,[1, 3, 5]]\n",
        "max_lfc = np.ceil(np.nanmax(np.abs(lfc.values.flatten())))\n",
        "p_vals = timecourse_df.iloc[:,[0, 2, 4]]\n",
        "min_p = np.ceil(-np.log10(np.nanmin(p_vals.values.flatten())))\n",
        "\n",
        "\n",
        "\n",
        "fig_sst_volcano,ax=plt.subplots(1,3,figsize=(6,3))\n",
        "for ai in range(len(ax)):\n",
        "    ax[ai].set_title('SST E4vsE3 at '+ list(deg_df_dict.keys())[ai])\n",
        "    ax[ai].set_xlabel('Log2 Fold Change')\n",
        "    ax[ai].set_ylabel('-log10(pvalue)')\n",
        "    ax[ai].axhline(-np.log10(0.05),c='k',linestyle=\":\",linewidth=.6,label='p=0.05')\n",
        "    ax[ai].legend(loc='upper left')\n",
        "    ax[ai].set_ylim(0,min_p+.25)\n",
        "    ax[ai].set_xlim(-max_lfc,max_lfc)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "for i in timecourse_df.index:\n",
        "    x_vals = timecourse_df.loc[i,:].values[np.array([1,3,5])]\n",
        "    y_vals = -np.log10(timecourse_df.loc[i,:].values[np.array([0,2,4])])\n",
        "\n",
        "    if not np.isnan(x_vals[0]*y_vals[0]):\n",
        "        ax[0].scatter(x_vals[0],y_vals[0],c='r',s=3)\n",
        "        ax[0].text(x_vals[0],y_vals[0],i)\n",
        "\n",
        "    if not np.isnan(x_vals[1]*y_vals[1]):\n",
        "        ax[1].scatter(x_vals[1],y_vals[1],c='k',s=3)\n",
        "        ax[1].text(x_vals[1],y_vals[1],i)\n",
        "\n",
        "\n",
        "    if not np.isnan(x_vals[2]*y_vals[2]):\n",
        "        ax[2].scatter(x_vals[2],y_vals[2],c='b',s=3)\n",
        "        ax[2].text(x_vals[2],y_vals[2],i)\n",
        "\n",
        "\n",
        "plt.tight_layout()\n",
        "fig_sst_volcano.savefig(os.path.join(path,'fig_sst_cluster_volcano.svg'),dpi=300)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "7b42a87d",
      "metadata": {
        "scrolled": false,
        "id": "7b42a87d"
      },
      "outputs": [],
      "source": [
        "#### SST POS DEG\n",
        "\n",
        "adata_sst_POS = adata_GABA.copy()\n",
        "adata_sst_POS.obs['sst_Pos'] = z_score(sc.get.obs_df(adata_sst_POS,'Sst'))>2\n",
        "adata_sst_POS = adata_sst_POS[adata_sst_POS.obs['sst_Pos'], :]\n",
        "adata_sst_POS.obs['age_geno'] = adata_sst_POS.obs['age_bin'].str.cat(adata_sst_POS.obs['E_type'],sep='_')\n",
        "display(adata_sst_POS)\n",
        "\n",
        "# ### kcn dotplot\n",
        "# fig_sst_clust,ax_c=plt.subplots(1,figsize=(3,10))\n",
        "# dp = sc.pl.dotplot(adata_sst_POS, kcn_list, groupby='age_geno', title='SST Positives (n=' +str(len(adata_sst_POS)) +')',\n",
        "#              swap_axes=True,ax=ax_c,expression_cutoff=.2)\n",
        "# fig_sst_clust.savefig('./fig_sst_clust.svg',dpi=300)\n",
        "\n",
        "\n",
        "path='./SST_POS_DEG/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "deg_df_dict={}\n",
        "for a in ages:\n",
        "        group_key = 'age_geno'\n",
        "        analysis_groups = [a+\"_E4\", a+\"_E3\"]\n",
        "        deg_group = analysis_groups[0]\n",
        "        reference_group = analysis_groups[1]\n",
        "        key_added = a+'_E4_wrt_E3'\n",
        "        sc.tl.rank_genes_groups(adata_sst_POS,use_raw=True,groupby=group_key,groups=analysis_groups,reference=reference_group,\n",
        "                                key_added=key_added, method='wilcoxon',tie_correct=True)\n",
        "        df = sc.get.rank_genes_groups_df(adata_sst_POS, group=deg_group,key=key_added,pval_cutoff=1,log2fc_min=-15, log2fc_max=15)\n",
        "        deg_df_dict[a] = df.set_index('names')\n",
        "        new_file = os.path.join(path,'SST_KCN_'+key_added+'.csv')\n",
        "        df.to_csv(new_file)\n",
        "\n",
        "\n",
        "timecourse_df = pd.DataFrame({'kcn':kcn_list}).set_index('kcn')\n",
        "for k,v in deg_df_dict.items():\n",
        "    timecourse_df[k+'_pval'] = np.nan\n",
        "    for ch in timecourse_df.index:\n",
        "        try:\n",
        "            timecourse_df.at[ch,k+'_pval'] = deg_df_dict[k].loc[ch,'pvals']\n",
        "            timecourse_df.at[ch,k+'_LFC'] = deg_df_dict[k].loc[ch,'logfoldchanges']\n",
        "        except:\n",
        "            None\n",
        "\n",
        "# display(timecourse_df.head(50))\n",
        "lfc = timecourse_df.iloc[:,[1, 3, 5]]\n",
        "max_lfc = np.ceil(np.nanmax(np.abs(lfc.values.flatten())))\n",
        "\n",
        "p_vals = timecourse_df.iloc[:,[0, 2, 4]]\n",
        "min_p = np.ceil(-np.log10(np.nanmin(p_vals.values.flatten())))\n",
        "\n",
        "\n",
        "fig_sst_volcano,ax=plt.subplots(1,3,figsize=(6,3))\n",
        "for ai in range(len(ax)):\n",
        "    ax[ai].set_title('SST E4vsE3 at '+ list(deg_df_dict.keys())[ai])\n",
        "    ax[ai].set_xlabel('Log2 Fold Change')\n",
        "    ax[ai].set_ylabel('-log10(pvalue)')\n",
        "    ax[ai].axhline(-np.log10(0.05),c='k',linestyle=\":\",linewidth=.6,label='p=0.05')\n",
        "    ax[ai].legend(loc='upper left')\n",
        "    ax[ai].set_ylim(0,min_p)\n",
        "    ax[ai].set_xlim(-max_lfc,max_lfc)\n",
        "\n",
        "\n",
        "for i in timecourse_df.index:\n",
        "    x_vals = timecourse_df.loc[i,:].values[np.array([1,3,5])]\n",
        "    y_vals = -np.log10(timecourse_df.loc[i,:].values[np.array([0,2,4])])\n",
        "\n",
        "    if not np.isnan(x_vals[0]*y_vals[0]):\n",
        "        ax[0].scatter(x_vals[0],y_vals[0],c='r',s=3)\n",
        "        ax[0].text(x_vals[0],y_vals[0],i)\n",
        "\n",
        "    if not np.isnan(x_vals[1]*y_vals[1]):\n",
        "        ax[1].scatter(x_vals[1],y_vals[1],c='k',s=3)\n",
        "        ax[1].text(x_vals[1],y_vals[1],i)\n",
        "\n",
        "\n",
        "    if not np.isnan(x_vals[2]*y_vals[2]):\n",
        "        ax[2].scatter(x_vals[2],y_vals[2],c='b',s=3)\n",
        "        ax[2].text(x_vals[2],y_vals[2],i)\n",
        "\n",
        "\n",
        "plt.tight_layout()\n",
        "fig_sst_volcano.savefig(os.path.join(path,'fig_sst_pos_volcano.svg'),dpi=300)\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "private_outputs": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.4"
    },
    "varInspector": {
      "cols": {
        "lenName": 16,
        "lenType": 16,
        "lenVar": 40
      },
      "kernels_config": {
        "python": {
          "delete_cmd_postfix": "",
          "delete_cmd_prefix": "del ",
          "library": "var_list.py",
          "varRefreshCmd": "print(var_dic_list())"
        },
        "r": {
          "delete_cmd_postfix": ") ",
          "delete_cmd_prefix": "rm(",
          "library": "var_list.r",
          "varRefreshCmd": "cat(var_dic_list()) "
        }
      },
      "types_to_exclude": [
        "module",
        "function",
        "builtin_function_or_method",
        "instance",
        "_Feature"
      ],
      "window_display": false
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}