{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/dtabuena/Workshop/blob/main/RNA_Workshop/Gabaergic_Analysis_GO.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "b294428a",
      "metadata": {
        "id": "b294428a",
        "scrolled": true,
        "tags": []
      },
      "outputs": [],
      "source": [
        "!pip install scanpy --quiet\n",
        "!pip install pybiomart --quiet\n",
        "!pip install python-igraph --quiet\n",
        "!pip install louvain --quiet\n",
        "!pip install pynndescent --quiet\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "16d3f9dc",
      "metadata": {
        "id": "16d3f9dc",
        "tags": []
      },
      "outputs": [],
      "source": [
        "import h5py\n",
        "import numpy as np\n",
        "import scipy as sci\n",
        "from matplotlib import pyplot as plt\n",
        "import scanpy as sc\n",
        "import tarfile\n",
        "import os\n",
        "import anndata as ad\n",
        "import pandas as pd\n",
        "import pybiomart\n",
        "from tqdm import tqdm\n",
        "import urllib.request\n",
        "from IPython.display import clear_output\n",
        "from matplotlib.pyplot import rc_context\n",
        "from scipy import stats as st\n",
        "os.chdir(\"C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021\")\n",
        "\n",
        "def publishable_plots(FS=6):\n",
        "    plt.rcParams.update({'font.size': FS,'axes.linewidth':.5,'figure.dpi':300,\n",
        "                         'xtick.major.width': 0.5,'ytick.major.width': 0.5,\n",
        "                         'figure.titlesize':FS,'axes.titlesize': FS,'xtick.labelsize': FS,\n",
        "                         'ytick.labelsize':FS,'axes.labelsize': FS,'legend.fontsize': FS,\n",
        "                         'figure.labelsize':FS})\n",
        "\n",
        "    import urllib.request\n",
        "    arial_link = 'https://raw.githubusercontent.com/dtabuena/Resources/main/Fonts/arial.ttf'\n",
        "    filename = './arial.ttf'\n",
        "    urllib.request.urlretrieve(arial_link, filename)\n",
        "    plt.rcParams.update({'font.family': 'arial'})\n",
        "    return None\n",
        "publishable_plots(6)\n",
        "sc.settings.verbosity = 'error'             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
        "sc.settings.set_figure_params(dpi=300, facecolor='white',fontsize=6,)\n",
        "pd.set_option('display.max_rows',100)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4514aef3",
      "metadata": {
        "id": "4514aef3",
        "tags": []
      },
      "outputs": [],
      "source": [
        "def trim_key(k):\n",
        "    floxed_dict = {'GSM5106175_YH_KZ03_01':('E3fKI_Syn_Cre602_15m','GSM5106175_602_E3fKI_15_XX'),\n",
        "                   'GSM5106176_YH_KZ03_03':('E4fKI_Syn_Cre475_15m','GSM5106176_475_E4fKI_15_XX')}\n",
        "    for f in floxed_dict.keys():\n",
        "        if f in k: return floxed_dict[f][1]\n",
        "    k = k.replace('_raw_gene_bc_matrices_h5.h5',\"\")\n",
        "    return k\n",
        "def query_capitilaziation(gene,adata):\n",
        "    try:\n",
        "        return adata.var.index[ [g.lower() for g in list(adata.var.index)].index(gene.lower()) ]\n",
        "    except:\n",
        "        return gene + ' not_found'\n",
        "def z_score(x,axis=-1):\n",
        "    x=np.array(x)\n",
        "    return (x-np.mean(x,axis=axis))/np.std(x,axis=axis)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "04bde4bf",
      "metadata": {
        "scrolled": true,
        "tags": [],
        "id": "04bde4bf"
      },
      "outputs": [],
      "source": [
        "os.chdir(\"C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021\")\n",
        "url = 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE167497&format=file'\n",
        "filename = './'+'zalocusky_indiv.tar'\n",
        "\n",
        "try:\n",
        "    for f in os.listdir('./indiv_animal_results'):\n",
        "        print(f)\n",
        "except:\n",
        "    urllib.request.urlretrieve(url, filename)\n",
        "    my_tar = tarfile.open(filename)\n",
        "    my_tar.extractall('./indiv_animal_results') # specify which folder to extract to\n",
        "    my_tar.close()\n",
        "    for f in os.listdir('./indiv_animal_results'):\n",
        "        print(f)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f115f8e4",
      "metadata": {
        "id": "f115f8e4",
        "scrolled": true,
        "tags": []
      },
      "outputs": [],
      "source": [
        "### Read, Combine, and Sample Split multiple 10x's\n",
        "\n",
        "adata_dict = {}\n",
        "for f in tqdm( os.listdir('./indiv_animal_results') ):\n",
        "    a = sc.read_10x_h5('./indiv_animal_results/'+f)\n",
        "    a.var_names_make_unique()\n",
        "    sample_code = trim_key(f)\n",
        "    a.obs['age_bin'] = str(int(np.ceil( int(sample_code.split(\"_\")[3])/5)*5))+'m'\n",
        "    a.obs['E_type'] = sample_code.split(\"_\")[2]\n",
        "    a.obs['mouse_ID'] = sample_code.split(\"_\")[1]\n",
        "    a.obs['well'] = sample_code.split(\"_\")[4]\n",
        "    a.obs['GSM'] = sample_code.split(\"_\")[0]\n",
        "    adata_dict[sample_code.split(\"_\")[0]] = a\n",
        "adata = ad.concat(adata_dict,axis = 0,label=\"Sample\",index_unique=\"_\")\n",
        "adata = adata[['fKI' not in t for t in adata.obs.E_type], :]\n",
        "adata_dict = {}\n",
        "clear_output()\n",
        "print('data_loaded.')\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "c5a6ad5b",
      "metadata": {
        "tags": [],
        "id": "c5a6ad5b"
      },
      "outputs": [],
      "source": [
        "def pull_gene_annots(csv_loc='./mmusculus_coding_noncoding.csv',\n",
        "                     my_git='https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv',\n",
        "                     biomart_name='mmusculus',\n",
        "                     biomart_keys=[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\",\"peptide\"]):\n",
        "\n",
        "\n",
        "    if os.path.exists('./mmusculus_coding_noncoding.csv'):\n",
        "        print( 'Use local copy of musmus')\n",
        "        annot_dd = pd.read_csv('./mmusculus_coding_noncoding.csv').set_index(\"external_gene_name\")\n",
        "    else:\n",
        "        try:\n",
        "            print( 'attempting to pull mus mus from git...')\n",
        "            musmus_link = 'https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv'\n",
        "            filename = './mmusculus_coding_noncoding.csv'\n",
        "            urllib.request.urlretrieve(musmus_link, filename)\n",
        "            annot_dd = pd.read_csv('./mmusculus_coding_noncoding.csv').set_index(\"external_gene_name\")\n",
        "        except:\n",
        "            print('attempting to pull mus mus from biomart...')\n",
        "            annot = sc.queries.biomart_annotations(\"mmusculus\",[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\"],).set_index('ensembl_gene_id')\n",
        "            uniq_inds = list(set(list(annot.index)))\n",
        "            for r in tqdm(uniq_inds):\n",
        "                match_bool = annot.index.str.contains(r)\n",
        "                if np.sum(match_bool)>1:\n",
        "                    new_val ='__'.join(list(annot.loc[r,'transcript_biotype']))\n",
        "                    annot.at[r,'transcript_biotype']=new_val\n",
        "            annot['is_coding']= annot.transcript_biotype.str.contains('coding')\n",
        "            annot_dd = annot.drop_duplicates().set_index(\"external_gene_name\")\n",
        "            annot_dd.to_csv('./mmusculus_coding_noncoding.csv')\n",
        "\n",
        "    coding_list = annot_dd.index[ annot_dd['is_coding'] ].to_list()\n",
        "    return coding_list, annot_dd\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def preprocess_andata10x(adata_og,pct_mito=0.25,min_genes=500,max_genes=2400,min_counts=500,max_counts=4500):\n",
        "\n",
        "    print('pulling gene annotations...')\n",
        "    coding_list, _ = pull_gene_annots()\n",
        "    adata_og.var['mt'] = adata_og.var_names.str.startswith('mt-')\n",
        "    adata_og.var['coding'] = [gene in coding_list for gene in adata_og.var_names]\n",
        "    sc.pp.calculate_qc_metrics(adata_og, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
        "\n",
        "    adata_QC = adata_og.copy()\n",
        "\n",
        "    print('Filtering...')\n",
        "    adata_QC = adata_QC[adata_QC.obs.pct_counts_mt < pct_mito, :]\n",
        "    print(str(np.sum(adata_og.obs.pct_counts_mt <pct_mito)) + f' cells with >{pct_mito}% removed')\n",
        "    adata_QC = adata_QC[:, adata_QC.var.coding]\n",
        "    print(str(np.sum(np.logical_not(adata_og.var.coding))) + ' non coding genes removed')\n",
        "    sc.pp.filter_cells(adata_QC, min_genes=min_genes)\n",
        "    sc.pp.filter_cells(adata_QC, max_genes=max_genes)\n",
        "    sc.pp.filter_cells(adata_QC, min_counts=min_counts)\n",
        "    sc.pp.filter_cells(adata_QC, max_counts=max_counts)\n",
        "    fig,ax=plt.subplots(1,figsize=(1.5,1.5))\n",
        "    sc.pl.scatter(adata_QC, x='total_counts', y='n_genes_by_counts',ax=ax)\n",
        "\n",
        "    return adata_QC,adata_og\n",
        "\n",
        "adata_QC = preprocess_andata10x(adata)[0]\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def high_var_genes_dim_reduc(adata,min_mean = 0.25,max_mean = 4,min_disp=0.55):\n",
        "    ''' The gene expression matrices were then log-normalized with a scale factor of 10,000,\n",
        "    using the Seurat NormalizeData function57,58. Highly dispersed genes were selected using\n",
        "    the Seurat FindVariableGenes function57,58,filtering for an average expression range of\n",
        "    0.25â€“4 and a minimum dispersion of 0.55, resulting in a list of 2,197 genes.'''\n",
        "    adata.raw = adata\n",
        "    sc.pp.normalize_total(adata, target_sum=10000)\n",
        "    sc.pp.log1p(adata)\n",
        "    sc.pp.highly_variable_genes(adata, min_mean=min_mean, max_mean=max_mean, min_disp=min_disp)\n",
        "    with rc_context({'figure.figsize': (1.5, 1.5)}):\n",
        "        sc.pl.highly_variable_genes(adata)\n",
        "    plt.tight_layout()\n",
        "    print(np.sum(adata.var['highly_variable']),'hv genes')\n",
        "\n",
        "\n",
        "    #### PCA\n",
        "    sc.tl.pca(adata, svd_solver='arpack',n_comps=50)\n",
        "    fig,ax=plt.subplots(figsize=(1,1))\n",
        "    ax.plot(adata.uns['pca']['variance_ratio'][:25],'ok',markersize=1)\n",
        "    quiet_PCA_plots(adata,['E_type','age_bin','mouse_ID'],pc_pairs=[(0,1),(2,3),(4,5),(6,7)])\n",
        "\n",
        "    return adata\n",
        "\n",
        "def quiet_PCA_plots(adata,key_list,figsize=(2,2),pc_pairs=[(0,1)]):\n",
        "    fig,ax=plt.subplots(1*len(pc_pairs),len(key_list),figsize=(figsize[0]*len(key_list),figsize[1]*len(pc_pairs)))\n",
        "    for ip,pair in enumerate(pc_pairs):\n",
        "        if len(key_list) == 1: ax=[ax]\n",
        "        for key_ind,key in enumerate(key_list):\n",
        "            key_types = sorted(list(set( adata.obs[key] )))\n",
        "            for k in key_types:\n",
        "                is_k = adata.obs[key]==k\n",
        "                ax[ip,key_ind].scatter(adata.obsm['X_pca'][is_k,pair[0]],adata.obsm['X_pca'][is_k,pair[1]],s=2,marker='.',linewidth=0,edgecolors=None,label=k)\n",
        "                ax[ip,key_ind].set_xlabel(f'PC{pair[0]}')\n",
        "                ax[ip,key_ind].set_ylabel(f'PC{pair[1]}')\n",
        "            if len(key_types)<8: ax[ip,key_ind].legend(key_types,loc='best',markerscale=3)\n",
        "            ax[ip,key_ind].set_title(key)\n",
        "            plt.tight_layout()\n",
        "    return None\n",
        "\n",
        "############## GABAERGIC Filter #####################\n",
        "adata_GABA = adata_QC.copy()\n",
        "age_dict = {'5m':'05m', '10m': '10m','15m': '15m+','20m': '15m+'}\n",
        "adata_GABA.obs['age_bin'] = [ age_dict[a] for a in adata_GABA.obs['age_bin'] ]\n",
        "\n",
        "adata_GABA.obs['Gad1_pos'] = z_score(sc.get.obs_df(adata_GABA,'Gad1'))>0.5\n",
        "adata_GABA.obs['Gad2_pos'] = z_score(sc.get.obs_df(adata_GABA,'Gad2'))>0.5\n",
        "adata_GABA.obs['Syn1_pos'] = z_score(sc.get.obs_df(adata_GABA,'Syn1'))>-100\n",
        "is_gaba = np.logical_and( np.logical_or(adata_GABA.obs['Gad1_pos'] , adata_GABA.obs['Gad2_pos'] ), adata_GABA.obs['Syn1_pos'])\n",
        "\n",
        "\n",
        "adata_GABA.obs['Gabaergic'] = is_gaba\n",
        "adata_GABA = adata_GABA[ adata_GABA.obs['Gabaergic'] ,:]\n",
        "sc.pp.filter_genes(adata_GABA, min_counts=50)\n",
        "\n",
        "\n",
        "############## Dim Reduction #####################\n",
        "adata_GABA = high_var_genes_dim_reduc(adata_GABA)"
      ],
      "metadata": {
        "id": "cx-L5v6n2FGP"
      },
      "id": "cx-L5v6n2FGP",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f246d0a6",
      "metadata": {
        "scrolled": true,
        "tags": [],
        "id": "f246d0a6"
      },
      "outputs": [],
      "source": [
        "def umap_and_cluster(adata, n_neighbors=10, n_pcs=20,resolution=.6,plot_keys=['Cluster (nn)'],size = 1,to_plot=True):\n",
        "    sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=n_pcs,random_state=42)\n",
        "    sc.tl.louvain(adata,resolution=resolution,random_state=42)\n",
        "    sc.tl.paga(adata)\n",
        "    sc.tl.umap(adata,random_state=42)\n",
        "    adata.obs['Cluster (nn)']= adata.obs['louvain']\n",
        "    if to_plot:\n",
        "        with rc_context({'figure.figsize': (2.5, 2.5)}):\n",
        "            sc.pl.umap(adata,add_outline=False, legend_loc='on data', color=plot_keys,size=size)\n",
        "    return adata\n",
        "\n",
        "def explore_umap(adata_GABA,key_list=[],size=1,legend_loc=None):\n",
        "    with rc_context({'figure.figsize': (1.5,1.5)}):\n",
        "        sc.pl.umap(adata_GABA, legend_loc=legend_loc, color=key_list,vmin=-.7,size=size,cmap='Purples') # add_outline=True,\n",
        "        plt.tight_layout()\n",
        "\n",
        "def get_gene_cluster(adata,gene):\n",
        "    gene_bool = z_score(sc.get.obs_df(adata,gene))>2\n",
        "    gene_cluster_list = np.array(adata.obs['louvain'][gene_bool])\n",
        "    gene_clust = st.mode(gene_cluster_list)\n",
        "    return gene_clust[0][0]\n",
        "\n",
        "def marker_analysis(adata):\n",
        "    #### Get Marker Genes\n",
        "    sc.tl.rank_genes_groups(adata, 'louvain', method='wilcoxon',key_added='maker_genes',pts=True,use_raw =True)\n",
        "    sc.tl.rank_genes_groups(adata, 'louvain', method='logreg',pts=True,use_raw =True)\n",
        "    maker_genes_df = pd.DataFrame(adata.uns['maker_genes']['names'])\n",
        "    maker_genes_df_LFC = pd.DataFrame(adata.uns['maker_genes']['logfoldchanges'])\n",
        "    top_30 = maker_genes_df[:30]\n",
        "    lfc_thresh=4\n",
        "    marker_genes = [m for m in set(top_30[maker_genes_df_LFC[:30]>lfc_thresh].values.flatten()) if isinstance(m,str)]\n",
        "    sc.pl.stacked_violin(adata, marker_genes, groupby='louvain');\n",
        "    return marker_genes\n",
        "\n",
        "adata_GABA = umap_and_cluster(adata_GABA, n_neighbors=15, n_pcs=20,resolution=.6,to_plot=False)\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Sst', 'Pvalb'],legend_loc='on data')\n",
        "explore_umap(adata_GABA,['Cluster (nn)','E_type', 'age_bin'])\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Vip', 'Reln'])\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Gfap', 'Syn1'])\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Apoe','E_type'])\n",
        "\n",
        "sst_clust = get_gene_cluster(adata_GABA,'Sst')\n",
        "pv_clust = get_gene_cluster(adata_GABA,'Pvalb')\n",
        "\n",
        "print('sst_clust',sst_clust)\n",
        "print('pv_clust',pv_clust)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d4d8a2fa",
      "metadata": {
        "id": "d4d8a2fa"
      },
      "outputs": [],
      "source": [
        "kcn_list = [n for n in adata_GABA.var.index if 'kcn' in str(n).lower()]\n",
        "kcn_list.sort()\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def deg_set_up(adata,keys_of_interest):\n",
        "    combo_key = '_'.join(keys_of_interest)\n",
        "    adata.obs[combo_key] = adata.obs[keys_of_interest[0]]\n",
        "    for i,v in enumerate(keys_of_interest[1:]):\n",
        "        # to_add = adata.obs[v]\n",
        "        adata.obs[combo_key] = adata.obs[combo_key].str.cat(adata.obs[v].astype(str),sep='_')\n",
        "    return adata,combo_key\n",
        "\n",
        "\n",
        "def deg_analysis(adata_temp,new_file_path,analysis_pairs,group_key,adata_name='volcano',log2fc_extrema=[-15,15],label_cutoff=1,to_plot=True):\n",
        "    '''\n",
        "    Takes in anndata RNA object and a stratifying key and returns\n",
        "    DEGs based on given grouppings. Also builds volcano plots.    '''\n",
        "    adata_temp.raw = adata_temp\n",
        "    '''Perform statistics and write to DF'''\n",
        "    # print(os.getcwd())\n",
        "    deg_df_dict={}\n",
        "    for key,pair in analysis_pairs.items():\n",
        "        sc.tl.rank_genes_groups(adata_temp,use_raw=True,groupby=group_key,groups=pair,reference=pair[1],key_added=key,method='wilcoxon',tie_correct=True)\n",
        "        deg_df = sc.get.rank_genes_groups_df(adata_temp, group=pair[0],key=key,pval_cutoff=1,log2fc_min=log2fc_extrema[0], log2fc_max=log2fc_extrema[1])\n",
        "        deg_df = deg_df.set_index('names')\n",
        "        deg_df_dict[key] = deg_df\n",
        "        new_file_name = os.path.join(new_file_path,adata_name+'_'+key+'.csv')\n",
        "        print(new_file_name)\n",
        "        deg_df.to_csv(new_file_name)\n",
        "    return adata_temp, deg_df_dict\n",
        "\n",
        "def deg_volc_plot(deg_df_dict,gene_list=[],label_cutoff=0.1,suptitle='',lock_y_max=True,special_genes=[],use_adj=False):\n",
        "    ''' Plot Volcanoes '''\n",
        "    fig_volcano,ax=plt.subplots(1,len(deg_df_dict),figsize=(len(deg_df_dict)*1.5,2))\n",
        "    if len(deg_df_dict) == 1:  ax=[ax]\n",
        "    for key,df in deg_df_dict.items():\n",
        "        key_ind = list(deg_df_dict.keys()).index(key)\n",
        "        ax[key_ind].set_title(key.replace('_','\\n'))\n",
        "        genes= df.index\n",
        "        lfc = df['logfoldchanges']\n",
        "        if use_adj: pval = df['pvals_adj']\n",
        "        else: pval = df['pvals']\n",
        "\n",
        "        neg_log10_pval = -np.log10(pval)\n",
        "        for i,g in enumerate(genes):\n",
        "            if str(g) in gene_list:\n",
        "                if not np.isnan(lfc[i]*neg_log10_pval[i]):\n",
        "                    ax[key_ind].scatter(lfc[i],neg_log10_pval[i],c='k',s=3)\n",
        "                    if pval[i]<label_cutoff:\n",
        "                        ax[key_ind].text(lfc[i],neg_log10_pval[i],genes[i],rotation=45,fontsize=4)\n",
        "\n",
        "                if g in special_genes:\n",
        "                    if not np.isnan(lfc[i]*neg_log10_pval[i]):\n",
        "                        ax[key_ind].scatter(lfc[i],neg_log10_pval[i],c='r',s=3)\n",
        "                        ax[key_ind].text(lfc[i],neg_log10_pval[i],genes[i],rotation=45,fontsize=4,color='r')\n",
        "    x_lim=[]\n",
        "    y_lim=[]\n",
        "    for a in ax:\n",
        "        x_lim.append(a.get_xlim())\n",
        "        y_lim.append(a.get_ylim())\n",
        "\n",
        "    x_etr = np.max(np.abs(x_lim))\n",
        "    y_etr = np.max(np.abs(y_lim))\n",
        "\n",
        "    for a in ax:\n",
        "        a.set_xlim(-x_etr*1.1,x_etr*1.1)\n",
        "        if lock_y_max: a.set_ylim(0,y_etr*1.1)\n",
        "        else: a.set_ylim(0,a.get_ylim()[1]*1.1)\n",
        "        a.set_xlabel('Log2 Fold Change')\n",
        "        if use_adj: a.set_ylabel('-log10(adj-pvalue)')\n",
        "        else: a.set_ylabel('-log10(raw-pvalue)')\n",
        "        a.axhline(-np.log10(0.05),c='k',linestyle=\":\",linewidth=.6,label='p=0.05')\n",
        "\n",
        "    fig_volcano.suptitle(suptitle)\n",
        "    plt.tight_layout()\n",
        "    return fig_volcano,ax\n",
        "\n",
        "def deg_heatmap(deg_dict,gene_list=[],ax=None,lfcmax = 2.5,ax_cbar=None,suptitle=''):\n",
        "\n",
        "    try: ax.grid(visible=False)\n",
        "    except:\n",
        "        fig,axs = plt.subplots(1,2,figsize=(1+.5*len(deg_dict.keys()),len(gene_list)/14),width_ratios=[9,.5])\n",
        "\n",
        "        ax=axs[0]\n",
        "        ax_cbar = axs[1]\n",
        "        ax.grid(visible=False)\n",
        "    if len(gene_list)==0:gene_list = list()\n",
        "\n",
        "    heat_map_null=np.zeros([len(gene_list),len(deg_dict)])\n",
        "    heat_map_lfc = heat_map_null.copy()*np.nan\n",
        "    heat_map_p = heat_map_null.copy()*np.nan\n",
        "    for key,deg_df in deg_dict.items():\n",
        "        key_id = list(deg_dict.keys()).index(key)\n",
        "        for gene in gene_list:\n",
        "            i_gene = gene_list.index(gene)\n",
        "            try:\n",
        "                heat_map_lfc[i_gene,key_id] = deg_df.loc[gene,'logfoldchanges']\n",
        "            except:\n",
        "                None\n",
        "    cbar = ax.pcolorfast(heat_map_lfc,vmin=-lfcmax,vmax=lfcmax,cmap='bwr')\n",
        "    ax.set_yticks(np.arange(len(gene_list))+0.5, labels=gene_list,rotation=0)\n",
        "    ax.tick_params(length=0)\n",
        "\n",
        "    fmt_xlabels = list(deg_dict.keys())\n",
        "    fmt_xlabels = [xl.replace('_','\\n') for xl in fmt_xlabels]\n",
        "    ax.set_xticks(np.arange(len(deg_dict.keys()))+0.5, labels=fmt_xlabels,rotation=0)\n",
        "    ax.grid(visible=False)\n",
        "\n",
        "\n",
        "    for key,deg_df in deg_dict.items():\n",
        "        key_id = list(deg_dict.keys()).index(key)\n",
        "        for gene in gene_list:\n",
        "            i_gene = gene_list.index(gene)\n",
        "            try:\n",
        "                p = deg_df.loc[gene,'pvals']\n",
        "                if p<.1:\n",
        "                    p_str = str(p)[:5]\n",
        "                    l2fc = str(2**deg_df.loc[gene,'logfoldchanges'])[:3]\n",
        "                    tag = f\"{l2fc}X; p={p_str}\"\n",
        "                    tag = pval_to_star(p)\n",
        "                    ax.text(key_id+.5,i_gene+.5,tag,ha='center',va='top',rotation=0)\n",
        "            except: None\n",
        "\n",
        "    ax.tick_params(top=True, labeltop=True, bottom=False, labelbottom=False)\n",
        "\n",
        "    try:\n",
        "        plt.colorbar(cbar,cax=ax_cbar)\n",
        "        ax_cbar.set_ylabel('log2(Fold Change)')\n",
        "    except: plt.colorbar(cbar)\n",
        "    plt.suptitle(suptitle)\n",
        "    plt.tight_layout()\n",
        "    return plt.gcf(), ax, cbar\n",
        "\n",
        "def pval_to_star(p,specifics={(.05,.1):'\\'',(.05,1):'ns'},max_star=5):\n",
        "    for range,tag in specifics.items():\n",
        "        if p<=np.max(range) and p>np.min(range): return tag\n",
        "    if p<=0.5 and p>0.01: return '*'\n",
        "    nlogp=-np.log10(p)\n",
        "    return ('*'*int(nlogp))[:max_star]"
      ],
      "metadata": {
        "id": "hru563c6bCFU"
      },
      "id": "hru563c6bCFU",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "E8xu4iK_641u"
      },
      "id": "E8xu4iK_641u",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "###### SST Cluster KCN DEGs #####\n",
        "path='./DEG_Reuslts/KCN/SST_Cluster_DEG_KCN_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sst_cluster = adata_GABA.copy()[adata_GABA.obs['louvain']==sst_clust,:]\n",
        "adata_sst_cluster.var['is_KCN'] = [g in kcn_list for g in adata_sst_cluster.var.index]\n",
        "adata_sst_cluster.obs['age_geno'] = adata_sst_cluster.obs['age_bin'].str.cat(adata_sst_cluster.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_sst_cluster.obs['age_bin'])))\n",
        "analysis_pairs = { a+'_E4vE3':(a+\"_E4\", a+\"_E3\") for a in ages}\n",
        "print(analysis_pairs)\n",
        "adata_name = 'SST_cluster_KCNs'\n",
        "sst_cluster_deg_data = deg_analysis(adata_sst_cluster,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "## Plot+Save\n",
        "num_sstc = len(adata_sst_cluster)\n",
        "fig_volcano,ax = deg_volc_plot(sst_cluster_deg_data[1],gene_list=kcn_list,suptitle=f'SST Cluster ({num_sstc})',special_genes=['Kcnt2','Kcnf1'])\n",
        "fig_volcano.savefig(path+'SST_Cluster_Age-Geno_volcano.svg', dpi=300, format='svg')\n",
        "fig_volcano.savefig(path+'SST_Cluster_Age-Geno_volcano.jpg', dpi=300, format='jpg')\n",
        "fig_heat,ax,cax = deg_heatmap(sst_cluster_deg_data[1],gene_list=kcn_list,lfcmax = 2.5,suptitle='SST Cluster')\n",
        "fig_heat.savefig(path+'SST_Cluster_Age-Geno_heatmap.svg', dpi=300, format='svg')\n",
        "fig_heat.savefig(path+'SST_Cluster_Age-Geno_heatmap.jpg', dpi=300, format='jpg')\n",
        "\n",
        "###### PV Cluster KCN DEGs #####\n",
        "path='./DEG_Reuslts/KCN/PV_Cluster_DEG_KCN_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_pv_cluster = adata_GABA.copy()[adata_GABA.obs['louvain']==pv_clust,:]\n",
        "adata_pv_cluster.var['is_KCN'] = [g in kcn_list for g in adata_pv_cluster.var.index]\n",
        "adata_pv_cluster.obs['age_geno'] = adata_pv_cluster.obs['age_bin'].str.cat(adata_pv_cluster.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_pv_cluster.obs['age_bin'])))\n",
        "analysis_pairs = { a+'_E4vE3':(a+\"_E4\", a+\"_E3\") for a in ages}\n",
        "adata_name = 'pv_cluster_KCNs'\n",
        "pv_cluster_deg_data = deg_analysis(adata_pv_cluster,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "### Plot+Save\n",
        "num_pvc = len(adata_pv_cluster)\n",
        "fig_volcano,ax = deg_volc_plot(pv_cluster_deg_data[1],gene_list=kcn_list,suptitle=f'PV Cluster ({num_pvc})',special_genes=['Kcnt2','Kcnf1'])\n",
        "fig_volcano.savefig(path+'PV_Cluster_Age-Geno_volcano.svg', dpi=300, format='svg')\n",
        "fig_volcano.savefig(path+'PV_Cluster_Age-Geno_volcano.jpg', dpi=300, format='jpg')\n",
        "fig_heat,ax,cax = deg_heatmap(pv_cluster_deg_data[1],gene_list=kcn_list,lfcmax = 2.5,suptitle='PV Cluster')\n",
        "fig_heat.savefig(path+'PV_Cluster_Age-Geno_heatmap.svg', dpi=300, format='svg')\n",
        "fig_heat.savefig(path+'PV_Cluster_Age-Geno_heatmap.jpg', dpi=300, format='jpg')\n",
        "\n",
        "###### SST Positive KCN DEGs #####\n",
        "path='./DEG_Reuslts/KCN/SST_Positive_DEG_KCN_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sst_positive = adata_GABA.copy()[sc.get.obs_df(adata_GABA,'Sst')>1,:]\n",
        "adata_sst_positive.var['is_KCN'] = [g in kcn_list for g in adata_sst_positive.var.index]\n",
        "adata_sst_positive.obs['age_geno'] = adata_sst_positive.obs['age_bin'].str.cat(adata_sst_positive.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_sst_positive.obs['age_bin'])))\n",
        "analysis_pairs = { a+'_E4vE3':(a+\"_E4\", a+\"_E3\") for a in ages}\n",
        "adata_name = 'SST_cluster_KCNs'\n",
        "sst_positive_deg_data = deg_analysis(adata_sst_positive,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "### Plot+Save\n",
        "num_sstp = len(adata_sst_positive)\n",
        "fig_volcano,ax = deg_volc_plot(sst_positive_deg_data[1],gene_list=kcn_list,suptitle=f'SST Pos ({num_sstp})',special_genes=['Kcnt2','Kcnf1'])\n",
        "fig_volcano.savefig(path+'SST_Pos_Age-Geno_volcano.svg', dpi=300, format='svg')\n",
        "fig_volcano.savefig(path+'SST_Pos_Age-Geno_volcano.jpg', dpi=300, format='jpg')\n",
        "\n",
        "fig_heat,ax,cax =deg_heatmap(sst_positive_deg_data[1],gene_list=kcn_list,lfcmax = 2.5,suptitle='SST Pos')\n",
        "fig_heat.savefig(path+'SST_Pos_Age-Geno_heatmap.svg', dpi=300, format='svg')\n",
        "fig_heat.savefig(path+'SST_Pos_Age-Geno_heatmap.jpg', dpi=300, format='jpg')\n",
        "\n",
        "\n",
        "###### PV positive KCN DEGs #####\n",
        "path='./DEG_Reuslts/KCN/PV_Positive_DEG_KCN_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_pv_positive = adata_GABA.copy()[sc.get.obs_df(adata_GABA,'Pvalb')>1,:]\n",
        "adata_pv_positive.var['is_KCN'] = [g in kcn_list for g in adata_pv_positive.var.index]\n",
        "adata_pv_positive.obs['age_geno'] = adata_pv_positive.obs['age_bin'].str.cat(adata_pv_positive.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_pv_positive.obs['age_bin'])))\n",
        "analysis_pairs = { a+'_E4vE3':(a+\"_E4\", a+\"_E3\") for a in ages}\n",
        "pv_positive_deg_data = deg_analysis(adata_pv_positive,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "### Plot+Save\n",
        "num_pvp = len(adata_pv_positive)\n",
        "fig_volcano,ax = deg_volc_plot(pv_positive_deg_data[1],gene_list=kcn_list,suptitle=f'PV Pos ({num_pvp})',special_genes=['Kcnt2','Kcnf1'])\n",
        "fig_volcano.savefig(path+'PV_Pos_Age-Geno_volcano.svg', dpi=300, format='svg')\n",
        "fig_volcano.savefig(path+'PV_Pos_Age-Geno_volcano.jpg', dpi=300, format='jpg')\n",
        "fig_heat,ax,cax =deg_heatmap(pv_positive_deg_data[1],gene_list=kcn_list,lfcmax = 2.5,suptitle='PV Pos')\n",
        "fig_heat.savefig(path+'PV_Pos_Age-Geno_heatmap.svg', dpi=300, format='svg')\n",
        "fig_heat.savefig(path+'PV_Pos_Age-Geno_heatmap.jpg', dpi=300, format='jpg')"
      ],
      "metadata": {
        "id": "CWKYhY5ZqOXc"
      },
      "id": "CWKYhY5ZqOXc",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "########### MULTIMODAL DEGS #######################\n",
        "#### Clusters\n",
        "path = './DEG_Reuslts/KCN/Cluster_DEGs_multivar/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sstpv_clusters = adata_GABA.copy()[np.logical_or(adata_GABA.obs['louvain']==sst_clust,adata_GABA.obs['louvain']==pv_clust),:]\n",
        "keys_of_interest = ['age_bin','louvain','E_type']\n",
        "adata_sstpv_clusters,combo_key = deg_set_up(adata_sstpv_clusters,keys_of_interest)\n",
        "analysis_pairs = {'15m_SST_E4 vs E3_(genotype)': (f'15m+_{sst_clust}_E4', f'15m+_{sst_clust}_E3'),\n",
        "                #   'SST_E4_15m vs 5m_(age)': (f'15m+_{sst_clust}_E4', f'05m_{sst_clust}_E4'),\n",
        "                #   'SST_E4_15m vs 10m_(age)': (f'15m+_{sst_clust}_E4', f'10m_{sst_clust}_E4'),\n",
        "                  '15m_E4_SST vs PV_(celltype)': (f'15m+_{sst_clust}_E4', f'15m+_{pv_clust}_E4')}\n",
        "multimodal_cluster_deg_data = deg_analysis(adata_sstpv_clusters,path,analysis_pairs,group_key = combo_key,log2fc_extrema=[-15,15],to_plot=False)\n",
        "\n",
        "n=str(np.sum(adata_GABA.obs['louvain']==sst_clust))+'v'+str(np.sum(adata_GABA.obs['louvain']==pv_clust))\n",
        "fig_volcano,ax = deg_volc_plot(multimodal_cluster_deg_data[1],gene_list=kcn_list,suptitle=f'Cluster Analysis\\n({n})',lock_y_max=False,special_genes=['Kcnt2','Kcnf1'])\n",
        "fig_volcano.savefig(path+'volc_Cluster_DEGs_multivar.svg', dpi=300, format='svg')\n",
        "fig_volcano.savefig(path+'volc_Cluster_DEGs_multivar.jpg', dpi=300, format='svg')\n",
        "fig_heat,ax,cax =deg_heatmap(multimodal_cluster_deg_data[1],gene_list=kcn_list,lfcmax = 2.5,suptitle=f'Cluster Analysis\\n({n})')\n",
        "fig_heat.savefig(path+'heatmap_Cluster_DEGs_multivar.svg', dpi=300, format='svg')\n",
        "fig_heat.savefig(path+'heatmap_Cluster_DEGs_multivar.jpg', dpi=300, format='jpg')\n",
        "\n",
        "\n",
        "#### Marker Positive\n",
        "path = './DEG_Reuslts/KCN/MarkerPos_DEGs_multivar/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sstpv_positive = adata_GABA.copy()[np.logical_or(sc.get.obs_df(adata_GABA,'Sst')>1,sc.get.obs_df(adata_GABA,'Pvalb')>1),:]\n",
        "adata_sstpv_positive.obs['Sst_Pos'] = sc.get.obs_df(adata_sstpv_positive,'Sst')>1\n",
        "adata_sstpv_positive.obs['Pv_Pos'] = sc.get.obs_df(adata_sstpv_positive,'Pvalb')>1\n",
        "keys_of_interest = ['age_bin','Sst_Pos','E_type']\n",
        "\n",
        "adata_sstpv_positive,combo_key = deg_set_up(adata_sstpv_positive,keys_of_interest)\n",
        "display(adata_sstpv_positive.obs.head(3))\n",
        "analysis_pairs = {'15m_SST_E4 vs E3_(genotype)': (f'15m+_True_E4', f'15m+_True_E3'),\n",
        "                #   'SST_E4_15m vs 5m_(age)': (f'15m+_True_E4', f'05m_True_E4'),\n",
        "                #   'SST_E4_15m vs 10m_(age)': (f'15m+_True_E4', f'10m_True_E4'),\n",
        "                  '15m_E4_SST vs PV_(celltype)': (f'15m+_True_E4', f'15m+_False_E4')}\n",
        "display(analysis_pairs)\n",
        "multimodal_markerpos_deg_data = deg_analysis(adata_sstpv_positive,path,analysis_pairs,group_key = combo_key,log2fc_extrema=[-15,15],to_plot=False)\n",
        "n=str(np.sum(sc.get.obs_df(adata_GABA,'Sst')>1))+'v'+str(np.sum(sc.get.obs_df(adata_GABA,'Pvalb')>1))\n",
        "fig_volcano,ax = deg_volc_plot(multimodal_markerpos_deg_data[1],gene_list=kcn_list,suptitle=f'MarkerPos Analysis\\n({n})',lock_y_max=False,special_genes=['Kcnt2','Kcnf1'])\n",
        "fig_volcano.savefig(path+'volc_MarkerPos_DEGs_multivar.svg', dpi=300, format='svg')\n",
        "fig_volcano.savefig(path+'volc_MarkerPos_DEGs_multivar.jpg', dpi=300, format='jpg')\n",
        "fig_heat,ax,cax =deg_heatmap(multimodal_markerpos_deg_data[1],gene_list=kcn_list,lfcmax = 2.5,suptitle=f'MarkerPos Analysis\\n({n})')\n",
        "fig_heat.savefig(path+'heatmap_MarkerPos_DEGs_multivar.svg', dpi=300, format='svg')\n",
        "fig_heat.savefig(path+'heatmap_MarkerPos_DEGs_multivar.jpg', dpi=300, format='jpg')"
      ],
      "metadata": {
        "id": "29ALUAijx3pP"
      },
      "id": "29ALUAijx3pP",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "MMm4Pp8a0bDy"
      },
      "id": "MMm4Pp8a0bDy"
    },
    {
      "cell_type": "code",
      "source": [
        "# ######### CHLORIDE #######################\n",
        "gene_sheets = ['C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021/Chloride Gene Lists/HUGO_Chloride_Channels.xlsx',\n",
        "                'C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021/Chloride Gene Lists/HUGO_SLC_(N)KCC.xlsx',\n",
        "                'C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021/Chloride Gene Lists/HUGO_SLC_22.xlsx',\n",
        "                'C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021/Chloride Gene Lists/HUGO_SLC_AnionicNT.xlsx',\n",
        "                'C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021/Chloride Gene Lists/HUGO_SLC_AnionicTransporter.xlsx',\n",
        "                'C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021/Chloride Gene Lists/HUGO_SLC_AnionicTransporter_Multi.xlsx',\n",
        "                ]\n",
        "Chloride_list = []\n",
        "for s in gene_sheets:\n",
        "    df = pd.read_excel(s)\n",
        "    Chloride_list = Chloride_list + list(df['Approved symbol'])\n",
        "Chloride_list = [n.capitalize() for n in Chloride_list]\n",
        "print(Chloride_list)"
      ],
      "metadata": {
        "id": "xScePdiJwOql"
      },
      "id": "xScePdiJwOql",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "###### SST Cluster Chloride DEGs #####\n",
        "path='./DEG_Reuslts/Chloride/DEG_Reuslts/SST_Cluster_DEG_Chloride_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sst_cluster = adata_GABA.copy()[adata_GABA.obs['louvain']==sst_clust,:]\n",
        "adata_sst_cluster.var['is_Chloride'] = [g in Chloride_list for g in adata_sst_cluster.var.index]\n",
        "adata_sst_cluster.obs['age_geno'] = adata_sst_cluster.obs['age_bin'].str.cat(adata_sst_cluster.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_sst_cluster.obs['age_bin'])))\n",
        "analysis_pairs = { a+'_E4vE3':(a+\"_E4\", a+\"_E3\") for a in ages}\n",
        "print(analysis_pairs)\n",
        "adata_name = 'SST_cluster_Chlorides'\n",
        "sst_cluster_deg_data = deg_analysis(adata_sst_cluster,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "## Plot+Save\n",
        "fig_volcano,ax = deg_volc_plot(sst_cluster_deg_data[1],gene_list=Chloride_list,suptitle='SST Cluster')\n",
        "fig_volcano.savefig(path+'SST_Cluster_Age-Geno_volcano.svg', dpi=300, format='svg')\n",
        "fig_heat,ax,cax = deg_heatmap(sst_cluster_deg_data[1],gene_list=Chloride_list,lfcmax = 2.5,suptitle='SST Cluster')\n",
        "fig_heat.savefig(path+'SST_Cluster_Age-Geno_heatmap.svg', dpi=300, format='svg')\n",
        "\n",
        "###### PV Cluster Chloride DEGs #####\n",
        "path='./DEG_Reuslts/Chloride/PV_Cluster_DEG_Chloride_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_pv_cluster = adata_GABA.copy()[adata_GABA.obs['louvain']==pv_clust,:]\n",
        "adata_pv_cluster.var['is_Chloride'] = [g in Chloride_list for g in adata_pv_cluster.var.index]\n",
        "adata_pv_cluster.obs['age_geno'] = adata_pv_cluster.obs['age_bin'].str.cat(adata_pv_cluster.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_pv_cluster.obs['age_bin'])))\n",
        "analysis_pairs = { a+'_E4vE3':(a+\"_E4\", a+\"_E3\") for a in ages}\n",
        "adata_name = 'pv_cluster_Chlorides'\n",
        "pv_cluster_deg_data = deg_analysis(adata_pv_cluster,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "### Plot+Save\n",
        "fig_volcano,ax = deg_volc_plot(pv_cluster_deg_data[1],gene_list=Chloride_list,suptitle='PV Cluster')\n",
        "fig_volcano.savefig(path+'PV_Cluster_Age-Geno_volcano.svg', dpi=300, format='svg')\n",
        "fig_heat,ax,cax = deg_heatmap(pv_cluster_deg_data[1],gene_list=Chloride_list,lfcmax = 2.5,suptitle='PV Cluster')\n",
        "fig_heat.savefig(path+'PV_Cluster_Age-Geno_heatmap.svg', dpi=300, format='svg')\n",
        "\n",
        "###### SST Positive Chloride DEGs #####\n",
        "path='./DEG_Reuslts/Chloride/SST_Positive_DEG_Chloride_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sst_positive = adata_GABA.copy()[sc.get.obs_df(adata_GABA,'Sst')>1,:]\n",
        "adata_sst_positive.var['is_Chloride'] = [g in Chloride_list for g in adata_sst_positive.var.index]\n",
        "adata_sst_positive.obs['age_geno'] = adata_sst_positive.obs['age_bin'].str.cat(adata_sst_positive.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_sst_positive.obs['age_bin'])))\n",
        "analysis_pairs = { a+'_E4vE3':(a+\"_E4\", a+\"_E3\") for a in ages}\n",
        "adata_name = 'SST_cluster_Chlorides'\n",
        "sst_positive_deg_data = deg_analysis(adata_sst_positive,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "### Plot+Save\n",
        "fig_volcano,ax = deg_volc_plot(sst_positive_deg_data[1],gene_list=Chloride_list,suptitle='SST Pos')\n",
        "fig_volcano.savefig(path+'SST_Pos_Age-Geno_volcano.svg', dpi=300, format='svg')\n",
        "fig_heat,ax,cax =deg_heatmap(sst_positive_deg_data[1],gene_list=Chloride_list,lfcmax = 2.5,suptitle='SST Pos')\n",
        "fig_heat.savefig(path+'SST_Pos_Age-Geno_heatmap.svg', dpi=300, format='svg')\n",
        "\n",
        "\n",
        "\n",
        "###### PV positive Chloride DEGs #####\n",
        "path='./DEG_Reuslts/Chloride/PV_Positive_DEG_Chloride_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_pv_positive = adata_GABA.copy()[sc.get.obs_df(adata_GABA,'Pvalb')>1,:]\n",
        "adata_pv_positive.var['is_Chloride'] = [g in Chloride_list for g in adata_pv_positive.var.index]\n",
        "adata_pv_positive.obs['age_geno'] = adata_pv_positive.obs['age_bin'].str.cat(adata_pv_positive.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_pv_positive.obs['age_bin'])))\n",
        "analysis_pairs = { a+'_E4vE3':(a+\"_E4\", a+\"_E3\") for a in ages}\n",
        "pv_positive_deg_data = deg_analysis(adata_pv_positive,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "### Plot+Save\n",
        "fig_volcano,ax = deg_volc_plot(pv_positive_deg_data[1],gene_list=Chloride_list,suptitle='PV Pos')\n",
        "fig_volcano.savefig(path+'PV_Pos_Age-Geno_volcano.svg', dpi=300, format='svg')\n",
        "fig_heat,ax,cax =deg_heatmap(pv_positive_deg_data[1],gene_list=Chloride_list,lfcmax = 2.5,suptitle='PV Pos')\n",
        "fig_heat.savefig(path+'PV_Pos_Age-Geno_heatmap.svg', dpi=300, format='svg')"
      ],
      "metadata": {
        "id": "j71bKqSgEXeH"
      },
      "id": "j71bKqSgEXeH",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "########### MULTIMODAL DEGS #######################\n",
        "#### Clusters\n",
        "path = './DEG_Reuslts/Chloride/Cluster_DEGs_multivar/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sstpv_clusters = adata_GABA.copy()[np.logical_or(adata_GABA.obs['louvain']==sst_clust,adata_GABA.obs['louvain']==pv_clust),:]\n",
        "keys_of_interest = ['age_bin','louvain','E_type']\n",
        "adata_sstpv_clusters,combo_key = deg_set_up(adata_sstpv_clusters,keys_of_interest)\n",
        "analysis_pairs = {'15m_SST_E4 vs E3_(genotype)': (f'15m+_{sst_clust}_E4', f'15m+_{sst_clust}_E3'),\n",
        "                  'SST_E4_15m vs 5m_(age)': (f'15m+_{sst_clust}_E4', f'05m_{sst_clust}_E4'),\n",
        "                  'SST_E4_15m vs 10m_(age)': (f'15m+_{sst_clust}_E4', f'10m_{sst_clust}_E4'),\n",
        "                  '15m_E4_SST vs PV_(celltype)': (f'15m+_{sst_clust}_E4', f'15m+_{pv_clust}_E4')}\n",
        "multimodal_cluster_deg_data = deg_analysis(adata_sstpv_clusters,path,analysis_pairs,group_key = combo_key,log2fc_extrema=[-15,15],to_plot=False)\n",
        "\n",
        "n=str(np.sum(adata_GABA.obs['louvain']==sst_clust))+'v'+str(np.sum(adata_GABA.obs['louvain']==pv_clust))\n",
        "fig_volcano,ax = deg_volc_plot(multimodal_cluster_deg_data[1],gene_list=Chloride_list,suptitle=f'Cluster Analysis\\n({n})',lock_y_max=False)\n",
        "fig_volcano.savefig(path+'volc_Cluster_DEGs_multivar.svg', dpi=300, format='svg')\n",
        "fig_heat,ax,cax =deg_heatmap(multimodal_cluster_deg_data[1],gene_list=Chloride_list,lfcmax = 2.5,suptitle=f'Cluster Analysis\\n({n})')\n",
        "fig_heat.savefig(path+'heatmap_Cluster_DEGs_multivar.svg', dpi=300, format='svg')\n",
        "\n",
        "#### Marker Positive\n",
        "path = './DEG_Reuslts/Chloride/MarkerPos_DEGs_multivar/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sstpv_positive = adata_GABA.copy()[np.logical_or(sc.get.obs_df(adata_GABA,'Sst')>1,sc.get.obs_df(adata_GABA,'Pvalb')>1),:]\n",
        "adata_sstpv_positive.obs['Sst_Pos'] = sc.get.obs_df(adata_sstpv_positive,'Sst')>1\n",
        "adata_sstpv_positive.obs['Pv_Pos'] = sc.get.obs_df(adata_sstpv_positive,'Pvalb')>1\n",
        "keys_of_interest = ['age_bin','Sst_Pos','E_type']\n",
        "\n",
        "adata_sstpv_positive,combo_key = deg_set_up(adata_sstpv_positive,keys_of_interest)\n",
        "display(adata_sstpv_positive.obs.head(3))\n",
        "analysis_pairs = {'15m_SST_E4 vs E3_(genotype)': (f'15m+_True_E4', f'15m+_True_E3'),\n",
        "                  'SST_E4_15m vs 5m_(age)': (f'15m+_True_E4', f'05m_True_E4'),\n",
        "                  'SST_E4_15m vs 10m_(age)': (f'15m+_True_E4', f'10m_True_E4'),\n",
        "                  '15m_E4_SST vs PV_(celltype)': (f'15m+_True_E4', f'15m+_False_E4')}\n",
        "display(analysis_pairs)\n",
        "multimodal_markerpos_deg_data = deg_analysis(adata_sstpv_positive,path,analysis_pairs,group_key = combo_key,log2fc_extrema=[-15,15],to_plot=False)\n",
        "n=str(np.sum(sc.get.obs_df(adata_GABA,'Sst')>1))+'v'+str(np.sum(sc.get.obs_df(adata_GABA,'Pvalb')>1))\n",
        "fig_volcano,ax = deg_volc_plot(multimodal_markerpos_deg_data[1],gene_list=Chloride_list,suptitle=f'MarkerPos Analysis\\n({n})',lock_y_max=False)\n",
        "fig_volcano.savefig(path+'volc_MarkerPos_DEGs_multivar.svg', dpi=300, format='svg')\n",
        "fig_heat,ax,cax =deg_heatmap(multimodal_markerpos_deg_data[1],gene_list=Chloride_list,lfcmax = 2.5,suptitle=f'MarkerPos Analysis\\n({n})')\n",
        "fig_heat.savefig(path+'heatmap_MarkerPos_DEGs_multivar.svg', dpi=300, format='svg')"
      ],
      "metadata": {
        "id": "qdbfHXkvZvPx"
      },
      "id": "qdbfHXkvZvPx",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "explore_umap(adata_GABA,['Cluster (nn)','Sst','Kcnt2'])\n",
        "\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Reln','Lamp5'])"
      ],
      "metadata": {
        "id": "LVrje4QK19jQ"
      },
      "id": "LVrje4QK19jQ",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# adata_GABA.\n",
        "keys_of_interest = ['age_bin','louvain','E_type']\n",
        "adata_sstpv_clusters,combo_key = deg_set_up(adata_sstpv_clusters,keys_of_interest)"
      ],
      "metadata": {
        "id": "xO7BTKSfntNz"
      },
      "id": "xO7BTKSfntNz",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "# print(analysis_pairs)\n",
        "# # adata_name = 'SST_cluster_Chlorides'\n",
        "adata_GABA_deg = adata_GABA.copy()\n",
        "keys_of_interest = ['louvain','age_bin','E_type']\n",
        "adata_GABA_deg,combo_key = deg_set_up(adata_GABA_deg,keys_of_interest)\n",
        "print(combo_key)\n",
        "louvains = list(set(adata_GABA_deg.obs['louvain']))\n",
        "analysis_pairs = { l+'_E4vE3':(l+\"_E4\", l+\"_E3\") for l in louvains}\n",
        "\n",
        "\n",
        "analysis_pairs = {}\n",
        "for l in louvains:\n",
        "    for a in ages:\n",
        "        analysis_pairs[l+\"_\"+a+'_E4vE3']=(l+\"_\"+a+\"_E4\",l+\"_\"+a+\"_E3\")\n",
        "\n",
        "adata_GABA_deg.obs.head()\n",
        "\n",
        "# print(analysis_pairs)\n",
        "os.chdir(\"C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021\")\n",
        "path = './Inhib/DEGS/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "_,deg_data = deg_analysis(adata_GABA_deg,path,analysis_pairs,adata_name='Inhib_degs',group_key = 'louvain_age_bin_E_type',log2fc_extrema=[-15,15],to_plot=False)"
      ],
      "metadata": {
        "id": "zhdxhE3n7tFf"
      },
      "id": "zhdxhE3n7tFf",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "b0884x5mahxy"
      },
      "id": "b0884x5mahxy",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "3f__vINfbXKo"
      },
      "id": "3f__vINfbXKo",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "try: os.makedirs('./GABA GO')\n",
        "except: None\n",
        "os.chdir('./GABA GO')"
      ],
      "metadata": {
        "id": "5jLELvvZc4vX"
      },
      "id": "5jLELvvZc4vX",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install goatools --quiet\n",
        "# !mv C:/Users/dennis.tabuena/Downloads/gene_result.txt .\n",
        "!python C:/Users/dennis.tabuena/AppData/Local/anaconda3/Scripts/ncbi_gene_results_to_python.py -o genes_ncbi_mus_musculus_proteincoding.py gene_result.txt\n",
        "from genes_ncbi_mus_musculus_proteincoding import GENEID2NT as GeneID2nt_mus\n"
      ],
      "metadata": {
        "id": "F3eXTlEtbXBj"
      },
      "id": "F3eXTlEtbXBj",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "from goatools.base import download_go_basic_obo\n",
        "from goatools.base import download_ncbi_associations\n",
        "from goatools.obo_parser import GODag\n",
        "from goatools.anno.genetogo_reader import Gene2GoReader\n",
        "from goatools.goea.go_enrichment_ns import GOEnrichmentStudyNS\n",
        "\n",
        "obo_fname = download_go_basic_obo()\n",
        "fin_gene2go = download_ncbi_associations()\n",
        "obodag = GODag(\"go-basic.obo\")\n",
        "\n",
        "#run one time to initialize\n",
        "\n",
        "mapper = {}\n",
        "\n",
        "for key in GeneID2nt_mus:\n",
        "    mapper[GeneID2nt_mus[key].Symbol] = GeneID2nt_mus[key].GeneID\n",
        "\n",
        "inv_map = {v: k for k, v in mapper.items()}\n",
        "\n",
        "\n",
        "# Read NCBI's gene2go. Store annotations in a list of namedtuples\n",
        "objanno = Gene2GoReader(fin_gene2go, taxids=[10090])\n",
        "# Get namespace2association where:\n",
        "#    namespace is:\n",
        "#        BP: biological_process\n",
        "#        MF: molecular_function\n",
        "#        CC: cellular_component\n",
        "#    assocation is a dict:\n",
        "#        key: NCBI GeneID\n",
        "#        value: A set of GO IDs associated with that gene\n",
        "ns2assoc = objanno.get_ns2assc()\n",
        "\n",
        "\n",
        "#run one time to initialize\n",
        "goeaobj = GOEnrichmentStudyNS(\n",
        "        GeneID2nt_mus.keys(), # List of mouse protein-coding genes\n",
        "        ns2assoc, # geneid/GO associations\n",
        "        obodag, # Ontologies\n",
        "        propagate_counts = False,\n",
        "        alpha = 0.05, # default significance cut-off\n",
        "        methods = ['fdr_bh']) # defult multipletest correction method\n",
        "\n",
        "#run one time to initialize\n",
        "GO_items = []\n",
        "\n",
        "temp = goeaobj.ns2objgoea['BP'].assoc\n",
        "for item in temp:\n",
        "    GO_items += temp[item]\n",
        "\n",
        "\n",
        "temp = goeaobj.ns2objgoea['CC'].assoc\n",
        "for item in temp:\n",
        "    GO_items += temp[item]\n",
        "\n",
        "\n",
        "temp = goeaobj.ns2objgoea['MF'].assoc\n",
        "for item in temp:\n",
        "    GO_items += temp[item]\n",
        "\n",
        "\n",
        "def go_it(test_genes):\n",
        "    print(f'input genes: {len(test_genes)}')\n",
        "\n",
        "    mapped_genes = []\n",
        "    for gene in test_genes:\n",
        "        try:\n",
        "            mapped_genes.append(mapper[gene])\n",
        "        except:\n",
        "            pass\n",
        "    print(f'mapped genes: {len(mapped_genes)}')\n",
        "\n",
        "    goea_results_all = goeaobj.run_study(mapped_genes)\n",
        "    goea_results_sig = [r for r in goea_results_all if r.p_fdr_bh < 0.05]\n",
        "    GO = pd.DataFrame(list(map(lambda x: [x.GO, x.goterm.name, x.goterm.namespace, x.p_uncorrected, x.p_fdr_bh,\\\n",
        "                   x.ratio_in_study[0], x.ratio_in_study[1], GO_items.count(x.GO), list(map(lambda y: inv_map[y], x.study_items)),\\\n",
        "                   ], goea_results_sig)), columns = ['GO ID', 'GO term', 'class', 'p', 'p_corr', 'n_genes',\\\n",
        "                                                    'n_study', 'n_go', 'study_genes'])\n",
        "\n",
        "    GO = GO[GO.n_genes > 1]\n",
        "    GO['Percent_Enriched'] = GO.n_genes/GO.n_go\n",
        "    return GO"
      ],
      "metadata": {
        "id": "4s49refGbW0y"
      },
      "id": "4s49refGbW0y",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Dvi-GZqi-d7i"
      },
      "id": "Dvi-GZqi-d7i",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "display(deg_data['5_10m_E4vE3'].head())"
      ],
      "metadata": {
        "id": "SUiBDyy8fYpx"
      },
      "id": "SUiBDyy8fYpx",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_sig_list(deg_df):\n",
        "    deg_df_sig = deg_df.copy()\n",
        "    deg_df_sig = deg_df_sig[deg_df_sig['pvals']<0.05]\n",
        "\n",
        "    return list(deg_df_sig.index)\n",
        "\n",
        "go_dict = {}\n",
        "for k in deg_data.keys():\n",
        "    sig_list = get_sig_list(deg_data['5_10m_E4vE3'])\n",
        "    go_dict[k] = go_it(sig_list)\n"
      ],
      "metadata": {
        "id": "mRKAUH1mb60s"
      },
      "id": "mRKAUH1mb60s",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "os.chdir('C:/Users/dennis.tabuena/Dropbox (Gladstone)/0_Projects/_ReAnalyze_Zalocusky_2021/GABA GO')\n",
        "for k,v in go_dict.items():\n",
        "    v.to_csv(f\"./{k}.csv\")"
      ],
      "metadata": {
        "id": "jLWLDk15K80Q"
      },
      "id": "jLWLDk15K80Q",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "######### Cell COUNTING"
      ],
      "metadata": {
        "id": "PCS9Q1pUaYOb"
      },
      "id": "PCS9Q1pUaYOb",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "### Cell Counts\n",
        "############## GABAERGIC Filter #####################\n",
        "adata_GABA = adata_QC.copy()\n",
        "age_dict = {'5m':'05m', '10m': '10m','15m': '15m','20m': '20m'}\n",
        "adata_GABA.obs['age_bin'] = [ age_dict[a] for a in adata_GABA.obs['age_bin'] ]\n",
        "\n",
        "adata_GABA.obs['Gad1_pos'] = z_score(sc.get.obs_df(adata_GABA,'Gad1'))>0.5\n",
        "adata_GABA.obs['Gad2_pos'] = z_score(sc.get.obs_df(adata_GABA,'Gad2'))>0.5\n",
        "adata_GABA.obs['Syn1_pos'] = z_score(sc.get.obs_df(adata_GABA,'Syn1'))>-100\n",
        "is_gaba = np.logical_and( np.logical_or(adata_GABA.obs['Gad1_pos'] , adata_GABA.obs['Gad2_pos'] ), adata_GABA.obs['Syn1_pos'])\n",
        "\n",
        "\n",
        "adata_GABA.obs['Gabaergic'] = is_gaba\n",
        "adata_GABA = adata_GABA[ adata_GABA.obs['Gabaergic'] ,:]\n",
        "sc.pp.filter_genes(adata_GABA, min_counts=50)\n",
        "\n",
        "\n",
        "############## Dim Reduction #####################\n",
        "adata_GABA = high_var_genes_dim_reduc(adata_GABA)\n",
        "adata_GABA = umap_and_cluster(adata_GABA, n_neighbors=15, n_pcs=20,resolution=.6,to_plot=False)\n",
        "\n"
      ],
      "metadata": {
        "id": "cYL5gkcnj3aL"
      },
      "id": "cYL5gkcnj3aL",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "adata_GABA.obs\n",
        "ages = list(set(adata_GABA.obs['age_bin']))\n",
        "clusters = list(set(adata_GABA.obs['louvain']))\n",
        "etype = list(set(adata_GABA.obs['E_type']))\n",
        "\n",
        "count_dict_type = {}\n",
        "for a in ages:\n",
        "    for c in ['5','2']:\n",
        "        for t in etype:\n",
        "            count_dict_type[a+\"_\"+c+\"c_\"+t] = np.sum(np.logical_and(np.logical_and(adata_GABA.obs['age_bin']==a,adata_GABA.obs['louvain']==c), adata_GABA.obs['E_type']==t))\n",
        "\n",
        "cell_count_type_df = pd.DataFrame({'Group': count_dict_type.keys(), 'Counts':count_dict_type.values()}).set_index('Group')\n",
        "display(cell_count_type_df)\n",
        "cell_count_type_df.to_csv('./cell_count_type_df.csv')"
      ],
      "metadata": {
        "id": "reFNzNQBnuXo"
      },
      "id": "reFNzNQBnuXo",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "adata_GABA.obs['High_E'] ='low'\n",
        "adata_GABA.obs['High_E'][z_score(sc.get.obs_df(adata_GABA,'Apoe'))>2] = 'high'\n",
        "\n",
        "\n",
        "# display(adata_GABA.obs.head(4))\n",
        "ages = list(set(adata_GABA.obs['age_bin']))\n",
        "clusters = list(set(adata_GABA.obs['louvain']))\n",
        "etype = list(set(adata_GABA.obs['E_type']))\n",
        "e_level = list(set(adata_GABA.obs['High_E']))\n",
        "\n",
        "count_dict_HE = {}\n",
        "for a in ages:\n",
        "    for c in ['5','2']:\n",
        "        for t in etype:\n",
        "            for l in e_level:\n",
        "                count_dict_HE[a+\"_\"+c+\"c_\"+t+'_'+l] = np.sum(np.logical_and(np.logical_and(adata_GABA.obs['age_bin']==a, adata_GABA.obs['louvain']==c),\n",
        "                                                                           np.logical_and(adata_GABA.obs['E_type']==t, adata_GABA.obs['High_E']==l)))\n",
        "\n",
        "\n",
        "cell_count_HE_df = pd.DataFrame({'Group': count_dict_HE.keys(), 'Counts':count_dict_HE.values()}).set_index('Group')\n",
        "display(cell_count_HE_df)\n",
        "cell_count_HE_df.to_csv('./cell_count_HE_df.csv')"
      ],
      "metadata": {
        "id": "LdaJbbcBn-Xm"
      },
      "id": "LdaJbbcBn-Xm",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "K-DpjV5OcQZN"
      },
      "id": "K-DpjV5OcQZN",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "adata_GABA_full = adata_QC.copy()\n",
        "adata_GABA.obs\n",
        "adata_GABA_full.obs['gaba_louvain']=np.nan\n",
        "for o in adata_GABA.obs_names:\n",
        "    adata_GABA_full.obs['gaba_louvain'][o] = adata_GABA.obs['louvain'][o]\n",
        "\n",
        "\n",
        "\n",
        "age_dict = {'5m':'05m', '10m': '10m','15m': '15m','20m': '20m'}\n",
        "adata_GABA_full.obs['age_bin'] = [ age_dict[a] for a in adata_GABA_full.obs['age_bin'] ]\n",
        "adata_GABA_full.obs['Gad1_pos'] = z_score(sc.get.obs_df(adata_GABA_full,'Gad1'))>0.5\n",
        "adata_GABA_full.obs['Gad2_pos'] = z_score(sc.get.obs_df(adata_GABA_full,'Gad2'))>0.5\n",
        "adata_GABA_full.obs['Syn1_pos'] = z_score(sc.get.obs_df(adata_GABA_full,'Syn1'))>0\n",
        "is_gaba =  np.logical_or(adata_GABA_full.obs['Gad1_pos'] , adata_GABA_full.obs['Gad2_pos'] )\n",
        "adata_GABA_full.obs['Gabaergic'] = is_gaba\n",
        "adata_GABA_full.obs['High_E'] = z_score(sc.get.obs_df(adata_GABA_full,'Apoe'))>2\n",
        "\n",
        "adata_GABA_full.obs"
      ],
      "metadata": {
        "id": "QwTWViTRmYAa"
      },
      "id": "QwTWViTRmYAa",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "adata_GABA_full[first_entry].obs['E_type']"
      ],
      "metadata": {
        "id": "4nFIvRlqp5hb"
      },
      "id": "4nFIvRlqp5hb",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "\n",
        "ages = list(set(adata_GABA_full.obs['age_bin']))\n",
        "ages = list(set(adata_GABA_full.obs['age_bin']))\n",
        "clusters = list(set(adata_GABA_full.obs['gaba_louvain']))\n",
        "etype = list(set(adata_GABA_full.obs['E_type']))\n",
        "mices = list(set(adata_GABA_full.obs['mouse_ID']))\n",
        "\n",
        "\n",
        "\n",
        "cell_count_df = pd.DataFrame(columns=['Animal','age_bin','E_type','Total','Neur','Gaba','PV','SST','PV_HE','SST_HE'])\n",
        "cell_count_df['Animal'] = mices\n",
        "cell_count_df = cell_count_df.set_index('Animal')\n",
        "for m in mices:\n",
        "    first_entry = adata_GABA_full.obs_names[  list(adata_GABA_full.obs['mouse_ID']).index(m)   ]\n",
        "    cell_count_df.loc[m,'E_type'] = adata_GABA_full.obs['E_type'][first_entry]\n",
        "    cell_count_df.loc[m,'age_bin'] = adata_GABA_full.obs['age_bin'][first_entry]\n",
        "    cell_count_df.loc[m,'Total'] = np.sum(  adata_GABA_full.obs['mouse_ID']==m )\n",
        "    cell_count_df.loc[m,'Neur'] = np.sum(  np.logical_and(adata_GABA_full.obs['mouse_ID']==m, adata_GABA_full.obs['Syn1_pos']) )\n",
        "    cell_count_df.loc[m,'Gaba'] = np.sum(  np.logical_and(adata_GABA_full.obs['mouse_ID']==m, adata_GABA_full.obs['Gabaergic']) )\n",
        "    cell_count_df.loc[m,'PV'] = np.sum(  np.logical_and(adata_GABA_full.obs['mouse_ID']==m, adata_GABA_full.obs['gaba_louvain']=='5') )\n",
        "    cell_count_df.loc[m,'SST'] = np.sum(  np.logical_and(adata_GABA_full.obs['mouse_ID']==m, adata_GABA_full.obs['gaba_louvain']=='2') )\n",
        "    cell_count_df.loc[m,'PV_HE'] = np.sum(  np.logical_and(adata_GABA_full.obs['mouse_ID']==m, np.logical_and(adata_GABA_full.obs['gaba_louvain']=='5',adata_GABA_full.obs['High_E']) ))\n",
        "    cell_count_df.loc[m,'SST_HE'] = np.sum(  np.logical_and(adata_GABA_full.obs['mouse_ID']==m, np.logical_and(adata_GABA_full.obs['gaba_louvain']=='2',adata_GABA_full.obs['High_E']) ))\n",
        "\n",
        "cell_count_df.to_csv('./cell_count_df.csv')\n",
        "os.getcwd()"
      ],
      "metadata": {
        "id": "wS87oOAdaeB7"
      },
      "id": "wS87oOAdaeB7",
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "private_outputs": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.4"
    },
    "varInspector": {
      "cols": {
        "lenName": 16,
        "lenType": 16,
        "lenVar": 40
      },
      "kernels_config": {
        "python": {
          "delete_cmd_postfix": "",
          "delete_cmd_prefix": "del ",
          "library": "var_list.py",
          "varRefreshCmd": "print(var_dic_list())"
        },
        "r": {
          "delete_cmd_postfix": ") ",
          "delete_cmd_prefix": "rm(",
          "library": "var_list.r",
          "varRefreshCmd": "cat(var_dic_list()) "
        }
      },
      "types_to_exclude": [
        "module",
        "function",
        "builtin_function_or_method",
        "instance",
        "_Feature"
      ],
      "window_display": false
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}