{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/dtabuena/Workshop/blob/main/Ref_Map/KZ_BuildRef.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "b294428a",
      "metadata": {
        "id": "b294428a",
        "scrolled": true,
        "tags": []
      },
      "outputs": [],
      "source": [
        "!pip install scanpy --quiet\n",
        "!pip install pybiomart --quiet\n",
        "!pip install python-igraph --quiet\n",
        "!pip install louvain --quiet\n",
        "!pip install pynndescent --quiet\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import h5py\n",
        "import numpy as np\n",
        "import scipy as sci\n",
        "from matplotlib import pyplot as plt\n",
        "import scanpy as sc\n",
        "import tarfile\n",
        "import os\n",
        "import anndata as ad\n",
        "import pandas as pd\n",
        "import pybiomart\n",
        "from tqdm import tqdm\n",
        "import urllib.request\n",
        "from IPython.display import clear_output\n",
        "from matplotlib.pyplot import rc_context\n",
        "from scipy import stats as st\n",
        "import logging\n",
        "logging.Logger('my_log').setLevel('INFO')\n",
        "\n",
        "############# SPECIFIC CONFIG #############\n",
        "working_dir = r\"C:\\Users\\dennis.tabuena\\Gladstone Dropbox\\Dennis Tabuena\\0_Projects\\_KZ_All\"\n",
        "os.chdir(working_dir)\n",
        "zalocusky_url = 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE167497&format=file'\n",
        "geo_zalo_filename = 'zalocusky_indiv.tar'\n",
        "\n",
        "import urllib\n",
        "response = urllib.request.urlretrieve('https://raw.githubusercontent.com/dtabuena/Resources/main/Matplotlib_Config/Load_FS6.py','Load_FS6.py')\n",
        "%run Load_FS6.py\n",
        "\n"
      ],
      "metadata": {
        "id": "YoAmeyEda0Sq"
      },
      "id": "YoAmeyEda0Sq",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4514aef3",
      "metadata": {
        "id": "4514aef3",
        "tags": []
      },
      "outputs": [],
      "source": [
        "def trim_key(k):\n",
        "    floxed_dict = {'GSM5106175_YH_KZ03_01':('E3fKI_Syn_Cre602_15m','GSM5106175_602_E3fKI_15_XX'),\n",
        "                   'GSM5106176_YH_KZ03_03':('E4fKI_Syn_Cre475_15m','GSM5106176_475_E4fKI_15_XX')}\n",
        "    for f in floxed_dict.keys():\n",
        "        if f in k: return floxed_dict[f][1]\n",
        "    k = k.replace('_raw_gene_bc_matrices_h5.h5',\"\")\n",
        "    return k\n",
        "\n",
        "def query_capitilaziation(gene,adata):\n",
        "    try:\n",
        "        return adata.var.index[ [g.lower() for g in list(adata.var.index)].index(gene.lower()) ]\n",
        "    except:\n",
        "        return gene + ' not_found'\n",
        "\n",
        "def z_score(x,axis=-1):\n",
        "    x=np.array(x)\n",
        "    return (x-np.mean(x,axis=axis))/np.std(x,axis=axis)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "c5a6ad5b",
      "metadata": {
        "tags": [],
        "id": "c5a6ad5b"
      },
      "outputs": [],
      "source": [
        "def pull_gene_annots(csv_loc='./mmusculus_coding_noncoding.csv',\n",
        "                     my_git='https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv',\n",
        "                     biomart_name='mmusculus',\n",
        "                     biomart_keys=[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\",\"peptide\"]):\n",
        "\n",
        "    if os.path.exists('./mmusculus_coding_noncoding.csv'):\n",
        "        print( 'Use local copy of musmus')\n",
        "        annot_dd = pd.read_csv('./mmusculus_coding_noncoding.csv').set_index(\"external_gene_name\")\n",
        "    else:\n",
        "        try:\n",
        "            print( 'attempting to pull mus mus from git...')\n",
        "            musmus_link = 'https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv'\n",
        "            filename = './mmusculus_coding_noncoding.csv'\n",
        "            urllib.request.urlretrieve(musmus_link, filename)\n",
        "            annot_dd = pd.read_csv('./mmusculus_coding_noncoding.csv').set_index(\"external_gene_name\")\n",
        "        except:\n",
        "            print('attempting to pull mus mus from biomart...')\n",
        "            annot = sc.queries.biomart_annotations(\"mmusculus\",[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\"],).set_index('ensembl_gene_id')\n",
        "            uniq_inds = list(set(list(annot.index)))\n",
        "            for r in tqdm(uniq_inds):\n",
        "                match_bool = annot.index.str.contains(r)\n",
        "                if np.sum(match_bool)>1:\n",
        "                    new_val ='__'.join(list(annot.loc[r,'transcript_biotype']))\n",
        "                    annot.at[r,'transcript_biotype']=new_val\n",
        "            annot['is_coding']= annot.transcript_biotype.str.contains('coding')\n",
        "            annot_dd = annot.drop_duplicates().set_index(\"external_gene_name\")\n",
        "            annot_dd.to_csv('./mmusculus_coding_noncoding.csv')\n",
        "\n",
        "    coding_list = annot_dd.index[ annot_dd['is_coding'] ].to_list()\n",
        "    return coding_list, annot_dd\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def preprocess_andata10x(adata_og,pct_mito=0.25,min_genes=500,max_genes=2400,min_counts=500,max_counts=4500):\n",
        "\n",
        "    print('pulling gene annotations...')\n",
        "    coding_list, _ = pull_gene_annots()\n",
        "    adata_og.var['mt'] = adata_og.var_names.str.startswith('mt-')\n",
        "    adata_og.var['coding'] = [gene in coding_list for gene in adata_og.var_names]\n",
        "    sc.pp.calculate_qc_metrics(adata_og, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
        "\n",
        "    adata_QC = adata_og.copy()\n",
        "\n",
        "    print('Filtering...')\n",
        "    adata_QC = adata_QC[adata_QC.obs.pct_counts_mt < pct_mito, :]\n",
        "    print(str(np.sum(adata_og.obs.pct_counts_mt <pct_mito)) + f' cells with >{pct_mito}% removed')\n",
        "    adata_QC = adata_QC[:, adata_QC.var.coding]\n",
        "    print(str(np.sum(np.logical_not(adata_og.var.coding))) + ' non coding genes removed')\n",
        "    sc.pp.filter_cells(adata_QC, min_genes=min_genes)\n",
        "    sc.pp.filter_cells(adata_QC, max_genes=max_genes)\n",
        "    sc.pp.filter_cells(adata_QC, min_counts=min_counts)\n",
        "    sc.pp.filter_cells(adata_QC, max_counts=max_counts)\n",
        "    fig,ax=plt.subplots(1,figsize=(1.5,1.5))\n",
        "    sc.pl.scatter(adata_QC, x='total_counts', y='n_genes_by_counts',ax=ax)\n",
        "\n",
        "    return adata_QC,adata_og\n",
        "\n",
        "def high_var_genes_dim_reduc(adata,min_mean = 0.25,max_mean = 4,min_disp=0.55):\n",
        "    ''' The gene expression matrices were then log-normalized with a scale factor of 10,000,\n",
        "    using the Seurat NormalizeData function57,58. Highly dispersed genes were selected using\n",
        "    the Seurat FindVariableGenes function57,58,filtering for an average expression range of\n",
        "    0.25â€“4 and a minimum dispersion of 0.55, resulting in a list of 2,197 genes.'''\n",
        "    adata.raw = adata\n",
        "    sc.pp.normalize_total(adata, target_sum=10000)\n",
        "    sc.pp.log1p(adata)\n",
        "    adata.uns['log1p'] = {'base': None}\n",
        "    print(adata.uns['log1p'])\n",
        "    sc.pp.highly_variable_genes(adata, min_mean=min_mean, max_mean=max_mean, min_disp=min_disp)\n",
        "    with rc_context({'figure.figsize': (1.5, 1.5)}):\n",
        "        sc.pl.highly_variable_genes(adata)\n",
        "    plt.tight_layout()\n",
        "    print(np.sum(adata.var['highly_variable']),'hv genes')\n",
        "\n",
        "\n",
        "    #### PCA\n",
        "    sc.tl.pca(adata, svd_solver='arpack',n_comps=50)\n",
        "    fig,ax=plt.subplots(figsize=(1,1))\n",
        "    ax.plot(adata.uns['pca']['variance_ratio'][:25],'ok',markersize=1)\n",
        "    quiet_PCA_plots(adata,['E_type','age_bin','mouse_ID'],pc_pairs=[(0,1),(2,3)])\n",
        "\n",
        "    return adata\n",
        "\n",
        "def quiet_PCA_plots(adata,key_list,figsize=(2,2),pc_pairs=[(0,1)]):\n",
        "    fig,ax=plt.subplots(1*len(pc_pairs),len(key_list),figsize=(figsize[0]*len(key_list),figsize[1]*len(pc_pairs)))\n",
        "    for ip,pair in enumerate(pc_pairs):\n",
        "        if len(key_list) == 1: ax=[ax]\n",
        "        for key_ind,key in enumerate(key_list):\n",
        "            key_types = sorted(list(set( adata.obs[key] )))\n",
        "            for k in key_types:\n",
        "                is_k = adata.obs[key]==k\n",
        "                ax[ip,key_ind].scatter(adata.obsm['X_pca'][is_k,pair[0]],adata.obsm['X_pca'][is_k,pair[1]],s=2,marker='.',linewidth=0,edgecolors=None,label=k)\n",
        "                ax[ip,key_ind].set_xlabel(f'PC{pair[0]}')\n",
        "                ax[ip,key_ind].set_ylabel(f'PC{pair[1]}')\n",
        "            if len(key_types)<8: ax[ip,key_ind].legend(key_types,loc='best',markerscale=3)\n",
        "            ax[ip,key_ind].set_title(key)\n",
        "            plt.tight_layout()\n",
        "    return None\n",
        "\n",
        "\n",
        "def umap_and_cluster(adata, n_neighbors=10, n_pcs=20,resolution=.6,plot_keys=['Cluster (nn)'],size = 1,to_plot=True):\n",
        "    sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=n_pcs,random_state=42)\n",
        "    sc.tl.louvain(adata,resolution=resolution,random_state=42)\n",
        "    sc.tl.paga(adata)\n",
        "    sc.tl.umap(adata,random_state=42)\n",
        "    adata.obs['Cluster (nn)']= adata.obs['louvain']\n",
        "    if to_plot:\n",
        "        with rc_context({'figure.figsize': (2.5, 2.5)}):\n",
        "            sc.pl.umap(adata,add_outline=False, legend_loc='on data', color=plot_keys,size=size)\n",
        "    return adata\n",
        "\n",
        "def explore_umap(adata_GABA,key_list=[],size=1,legend_loc=None):\n",
        "    with rc_context({'figure.figsize': (1.5,1.5)}):\n",
        "        sc.pl.umap(adata_GABA, legend_loc=legend_loc, color=key_list,vmin=-0.5,size=size,cmap='Purples') # add_outline=True,\n",
        "        plt.tight_layout()\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def marker_analysis(adata):\n",
        "    #### Get Marker Genes\n",
        "    sc.tl.rank_genes_groups(adata, 'louvain', method='wilcoxon',key_added='maker_genes',pts=True,use_raw =True)\n",
        "    sc.tl.rank_genes_groups(adata, 'louvain', method='logreg',pts=True,use_raw =True)\n",
        "    maker_genes_df = pd.DataFrame(adata.uns['maker_genes']['names'])\n",
        "    maker_genes_df_LFC = pd.DataFrame(adata.uns['maker_genes']['logfoldchanges'])\n",
        "    top_30 = maker_genes_df[:30]\n",
        "    lfc_thresh=4\n",
        "    marker_genes = [m for m in set(top_30[maker_genes_df_LFC[:30]>lfc_thresh].values.flatten()) if isinstance(m,str)]\n",
        "    sc.pl.stacked_violin(adata, marker_genes, groupby='louvain');\n",
        "    return marker_genes\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "def show_clusters(adata_GABA, key, cmap='tab20', legend_fontsize='small'):\n",
        "    assert key in adata_GABA.obs, f\"Key '{key}' not found in adata_GABA.obs.\"\n",
        "\n",
        "    umap_coords = adata_GABA.obsm['X_umap']\n",
        "    size = 0.4\n",
        "\n",
        "    # Create the figure with subplots for scatter and legend\n",
        "    fig, (ax_scatter, ax_legend) = plt.subplots(1, 2, figsize=(3.5, 2.5), gridspec_kw={'width_ratios': [5, 1]})\n",
        "\n",
        "    # Get cluster colors\n",
        "    colors = adata_GABA.obs[key].values.astype(float)\n",
        "    scatter = ax_scatter.scatter(\n",
        "        umap_coords[:, 0], umap_coords[:, 1],\n",
        "        c=colors, cmap=cmap, s=size, vmin=-0.5, edgecolor='none'  # Set edgecolor to 'none'\n",
        "    )\n",
        "\n",
        "    # Set equal aspect ratio to make axes square\n",
        "    ax_scatter.set_aspect('equal', adjustable='box')\n",
        "    ax_scatter.set_xlabel('umap 1',ha='left')\n",
        "    ax_scatter.xaxis.set_label_coords(0, 0.0)\n",
        "    ax_scatter.set_ylabel('umap 2',ha='left')\n",
        "    ax_scatter.yaxis.set_label_coords(0.0, 0)\n",
        "    ax_scatter.set_xticks([])\n",
        "    ax_scatter.set_yticks([])\n",
        "\n",
        "    unique_clusters = np.unique(colors)\n",
        "    handles = []\n",
        "    labels = []\n",
        "\n",
        "    # Create custom legend handles for unique clusters\n",
        "    for cluster in unique_clusters:\n",
        "        color_i = scatter.cmap(scatter.norm(cluster))\n",
        "        handle = plt.Line2D([], [], marker='o', linestyle='None', color=color_i,\n",
        "                             markerfacecolor=color_i,\n",
        "                             markersize=4.5)  # Smaller markers\n",
        "        handles.append(handle)\n",
        "        labels.append(f'Cluster {int(cluster)}')\n",
        "\n",
        "    # Turn off the axis for the legend subplot\n",
        "    ax_legend.axis('off')\n",
        "\n",
        "    # Add the custom legend to the legend axis with specified font size\n",
        "    ax_legend.legend(handles, labels, loc='center', fontsize=legend_fontsize, markerscale=0.75,\n",
        "                     ncol=1, frameon=False)\n",
        "    plt.tight_layout()\n",
        "    plt.show()\n",
        "\n",
        "    return fig, (ax_scatter, ax_legend)\n",
        "\n",
        "# Example calls to the function\n",
        "# show_clusters(adata_GABA, 'Cluster (nn)')\n",
        "\n",
        "\n",
        "def show_gene_loading(adata_GABA, gene, cmap='Purples'):\n",
        "    assert gene in adata_GABA.var_names, f\"Gene '{gene}' not found in adata_GABA.var_names.\"\n",
        "\n",
        "    umap_coords = adata_GABA.obsm['X_umap']\n",
        "    size = 0.5\n",
        "\n",
        "    # Get gene expression values for the specified gene\n",
        "    gene_expression = adata_GABA[:, gene].X.toarray().flatten()  # Convert to dense array and flatten\n",
        "\n",
        "    # Create the figure\n",
        "    fig = plt.figure(figsize=(3, 1.5))\n",
        "\n",
        "    # Define positions for scatter and colorbar axes\n",
        "    ax_scatter = fig.add_axes([0.0, 0.1, 0.5, 0.8])  # [left, bottom, width, height]\n",
        "    ax_cbar = fig.add_axes([0.50, 0.1, 0.01, 0.8])  # Slightly to the right of the scatter plot\n",
        "\n",
        "    # Scatter plot\n",
        "    scatter = ax_scatter.scatter(\n",
        "        umap_coords[:, 0], umap_coords[:, 1],\n",
        "        c=gene_expression, cmap=cmap, s=size, edgecolor='none',vmin=-.45,vmax=2.75)\n",
        "\n",
        "    # Set equal aspect ratio to make axes square\n",
        "    ax_scatter.set_aspect('equal', adjustable='box')\n",
        "    ax_scatter.set_xlabel('umap 1',ha='left')\n",
        "    ax_scatter.xaxis.set_label_coords(0, 0.0)\n",
        "    ax_scatter.set_ylabel('umap 2',ha='left')\n",
        "    ax_scatter.yaxis.set_label_coords(0.0, 0)\n",
        "    ax_scatter.set_xticks([])\n",
        "    ax_scatter.set_yticks([])\n",
        "\n",
        "    # Colorbar\n",
        "    cbar = fig.colorbar(scatter, cax=ax_cbar, orientation='vertical')\n",
        "    cbar.set_label(f'stdev/mean', rotation=270, labelpad=10)\n",
        "    ax_scatter.set_title(gene.upper())\n",
        "    # Ensure colorbar is labeled\n",
        "    # ax_cbar.set_ylabel(f'{gene} Expression', rotation=270, labelpad=10)\n",
        "\n",
        "    # Adjust layout and show the plot\n",
        "    plt.tight_layout()\n",
        "    plt.show()\n",
        "\n",
        "    return fig, (ax_scatter, ax_cbar)\n",
        "\n",
        "# Example call to the function\n",
        "# show_gene_loading(adata_GABA, 'Sst')\n",
        "\n",
        "# cluster_fig, ax = show_clusters(adata_GABA, 'Cluster (nn)')\n",
        "# cluster_fig.savefig(\"cluster_plot.svg\", format=\"svg\")\n",
        "# cluster_fig.savefig(\"cluster_plot.jpeg\", format=\"jpeg\")"
      ],
      "metadata": {
        "id": "aM-_rLAPGNdx"
      },
      "id": "aM-_rLAPGNdx",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "################# INITIALIZE DIRECTORY DOWNLOAD FROM GEO\n",
        "\n",
        "# try: os.makedirs(working_dir)\n",
        "# except: None\n",
        "os.chdir(working_dir)\n",
        "try: os.makedirs('./indiv_animal_results')\n",
        "except: None\n",
        "urllib.request.urlretrieve(zalocusky_url, './indiv_animal_results/'+geo_zalo_filename)\n",
        "my_tar = tarfile.open('./indiv_animal_results/'+geo_zalo_filename)\n",
        "my_tar.extractall('./indiv_animal_results') # specify which folder to extract to\n",
        "my_tar.close()\n",
        "# for f in os.listdir('./indiv_animal_results'):\n",
        "#     print(f)\n"
      ],
      "metadata": {
        "id": "R4iZj93WnnKd"
      },
      "id": "R4iZj93WnnKd",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import h5py\n",
        "import scanpy as sc\n",
        "\n",
        "# Replace with your HDF5 file path\n",
        "file_path = 'raw_matix.h5'\n",
        "\n",
        "# Read the Cell Ranger data\n",
        "adata = sc.read_10x_h5(file_path)\n",
        "\n",
        "# Inspect the imported data\n",
        "print(adata)\n",
        "print(f\"Number of cells: {adata.n_obs}\")\n",
        "print(f\"Number of genes: {adata.n_vars}\")"
      ],
      "metadata": {
        "id": "1Mf7xHxPJLDW"
      },
      "id": "1Mf7xHxPJLDW",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "################# Read, Combine, and Sample Split multiple 10x's\n",
        "\n",
        "adata_dict = {}\n",
        "for f in os.listdir('./indiv_animal_results'):\n",
        "    if '.h5' in f:\n",
        "        a = sc.read_10x_h5('./indiv_animal_results/'+f)\n",
        "        a.var_names_make_unique()\n",
        "        sample_code = trim_key(f)\n",
        "        a.obs['age_bin'] = str(int(np.ceil( int(sample_code.split(\"_\")[3])/5)*5))+'m'\n",
        "        a.obs['E_type'] = sample_code.split(\"_\")[2]\n",
        "        a.obs['mouse_ID'] = sample_code.split(\"_\")[1]\n",
        "        a.obs['well'] = sample_code.split(\"_\")[4]\n",
        "        a.obs['GSM'] = sample_code.split(\"_\")[0]\n",
        "        adata_dict[sample_code.split(\"_\")[0]] = a\n",
        "adata = ad.concat(adata_dict,axis = 0,label=\"Sample\",index_unique=\"_\")\n",
        "adata.obs.E_type\n",
        "# adata = adata[['fKI' not in t for t in adata.obs.E_type], :]\n",
        "adata_dict = {}\n",
        "clear_output()\n",
        "adata.write_h5ad(filename='./kz_adata_raw.h5')\n",
        "\n",
        "\n",
        "adata_QC = preprocess_andata10x(adata)[0]\n",
        "adata_QC.write_h5ad(filename='./kz_adata_qc.h5')\n",
        "\n",
        "\n",
        "############## GABAERGIC Filter #####################\n",
        "adata_GABA = adata_QC.copy()\n",
        "age_dict = {'5m':'05m', '10m': '10m','15m': '15m+','20m': '15m+'}\n",
        "adata_GABA.obs['age_bin'] = [ age_dict[a] for a in adata_GABA.obs['age_bin'] ]\n",
        "\n",
        "adata_GABA.obs['Gad1_pos'] = z_score(sc.get.obs_df(adata_GABA,'Gad1'))>0.5\n",
        "adata_GABA.obs['Gad2_pos'] = z_score(sc.get.obs_df(adata_GABA,'Gad2'))>0.5\n",
        "adata_GABA.obs['Syn1_pos'] = z_score(sc.get.obs_df(adata_GABA,'Syn1'))>-100\n",
        "is_gaba = np.logical_and( np.logical_or(adata_GABA.obs['Gad1_pos'] , adata_GABA.obs['Gad2_pos'] ), adata_GABA.obs['Syn1_pos'])\n",
        "\n",
        "\n",
        "adata_GABA.obs['Gabaergic'] = is_gaba\n",
        "adata_GABA = adata_GABA[ adata_GABA.obs['Gabaergic'] ,:]\n",
        "sc.pp.filter_genes(adata_GABA, min_counts=50)\n",
        "\n",
        "############## Dim Reduction #####################\n",
        "adata_GABA = high_var_genes_dim_reduc(adata_GABA)\n",
        "adata_GABA.write_h5ad(filename='./kz_adata_GABA.h5')\n",
        "\n",
        "\n",
        "print(adata_GABA.uns['log1p'])"
      ],
      "metadata": {
        "id": "cx-L5v6n2FGP"
      },
      "id": "cx-L5v6n2FGP",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# ############## Dim Reduction #####################\n",
        "# # adata_GABA = high_var_genes_dim_reduc(adata_GABApub\n",
        "# # adata_GABA.write_h5ad(filename='./kz_adata_GABA.h5')\n",
        "\n",
        "# adata_GABA = ad.read_h5ad('./kz_adata_GABA.h5')\n",
        "# adata_GABA = umap_and_cluster(adata_GABA, n_neighbors=15, n_pcs=20,resolution=.6,to_plot=False)\n",
        "# show_clusters(adata_GABA, 'Cluster (nn)')"
      ],
      "metadata": {
        "id": "thajIZvhtaW0"
      },
      "id": "thajIZvhtaW0",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f246d0a6",
      "metadata": {
        "scrolled": true,
        "tags": [],
        "id": "f246d0a6"
      },
      "outputs": [],
      "source": [
        "show_clusters(adata_GABA, 'Cluster (nn)')\n",
        "show_gene_loading(adata_GABA, 'Sst')\n",
        "show_gene_loading(adata_GABA, 'Pvalb')\n",
        "show_gene_loading(adata_GABA, 'Vip')\n",
        "show_gene_loading(adata_GABA, 'Reln')\n",
        "show_gene_loading(adata_GABA, 'Apoe')\n",
        "show_gene_loading(adata_GABA, 'Syn1')\n",
        "show_gene_loading(adata_GABA, 'Gfap')\n",
        "\n",
        "\n",
        "adata_GABA = umap_and_cluster(adata_GABA, n_neighbors=15, n_pcs=20,resolution=.6,to_plot=False)\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Sst', 'Pvalb'],legend_loc='on data')\n",
        "explore_umap(adata_GABA,['Cluster (nn)','E_type', 'age_bin'])\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Vip', 'Reln'])\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Gfap', 'Syn1'])\n",
        "explore_umap(adata_GABA,['Cluster (nn)','Apoe','E_type'])\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "adata_GABA"
      ],
      "metadata": {
        "id": "WLmXXMsN0CiJ"
      },
      "id": "WLmXXMsN0CiJ",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d4d8a2fa",
      "metadata": {
        "id": "d4d8a2fa"
      },
      "outputs": [],
      "source": [
        "kcn_list = [n for n in adata_GABA.var.index if 'kcn' in str(n).lower()]\n",
        "kcn_list.sort()\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def deg_set_up(adata,keys_of_interest):\n",
        "    combo_key = '_'.join(keys_of_interest)\n",
        "    adata.obs[combo_key] = adata.obs[keys_of_interest[0]]\n",
        "    for i,v in enumerate(keys_of_interest[1:]):\n",
        "        # to_add = adata.obs[v]\n",
        "        adata.obs[combo_key] = adata.obs[combo_key].str.cat(adata.obs[v].astype(str),sep='_')\n",
        "    return adata,combo_key\n",
        "\n",
        "\n",
        "def deg_analysis(adata_temp,new_file_path,analysis_pairs,group_key,adata_name='volcano',log2fc_extrema=[-15,15],label_cutoff=1,to_plot=True):\n",
        "    '''\n",
        "    Takes in anndata RNA object and a stratifying key and returns\n",
        "    DEGs based on given grouppings. Also builds volcano plots.    '''\n",
        "    adata_temp.raw = adata_temp\n",
        "    '''Perform statistics and write to DF'''\n",
        "    # print(os.getcwd())\n",
        "    deg_df_dict={}\n",
        "    for key,pair in analysis_pairs.items():\n",
        "        sc.tl.rank_genes_groups(adata_temp,use_raw=True,groupby=group_key,groups=pair,reference=pair[1],key_added=key,method='wilcoxon',tie_correct=True)\n",
        "        deg_df = sc.get.rank_genes_groups_df(adata_temp, group=pair[0],key=key,pval_cutoff=1,log2fc_min=log2fc_extrema[0], log2fc_max=log2fc_extrema[1])\n",
        "        deg_df = deg_df.set_index('names')\n",
        "        deg_df_dict[key] = deg_df\n",
        "        new_file_name = os.path.join(new_file_path,adata_name+'_'+key+'.csv')\n",
        "        print(new_file_name)\n",
        "        deg_df.to_csv(new_file_name)\n",
        "    return adata_temp, deg_df_dict\n",
        "\n",
        "def deg_volc_plot(deg_df_dict,gene_list=[],label_cutoff=0.1,suptitle='',lock_y_max=True,special_genes=[],use_adj=False):\n",
        "    ''' Plot Volcanoes '''\n",
        "    fig_volcano,ax=plt.subplots(1,len(deg_df_dict),figsize=(len(deg_df_dict)*1.5,2))\n",
        "    if len(deg_df_dict) == 1:  ax=[ax]\n",
        "    for key,df in deg_df_dict.items():\n",
        "        key_ind = list(deg_df_dict.keys()).index(key)\n",
        "        ax[key_ind].set_title(key.replace('_','\\n'))\n",
        "        genes= df.index\n",
        "        lfc = df['logfoldchanges']\n",
        "        if use_adj: pval = df['pvals_adj']\n",
        "        else: pval = df['pvals']\n",
        "\n",
        "        neg_log10_pval = -np.log10(pval)\n",
        "        for i,g in enumerate(genes):\n",
        "            if str(g) in gene_list:\n",
        "                if not np.isnan(lfc[i]*neg_log10_pval[i]):\n",
        "                    ax[key_ind].scatter(lfc[i],neg_log10_pval[i],c='k',s=3)\n",
        "                    if pval[i]<label_cutoff:\n",
        "                        ax[key_ind].text(lfc[i],neg_log10_pval[i],genes[i],rotation=45,fontsize=4)\n",
        "\n",
        "                if g in special_genes:\n",
        "                    if not np.isnan(lfc[i]*neg_log10_pval[i]):\n",
        "                        ax[key_ind].scatter(lfc[i],neg_log10_pval[i],c='r',s=3)\n",
        "                        ax[key_ind].text(lfc[i],neg_log10_pval[i],genes[i],rotation=45,fontsize=4,color='r')\n",
        "    x_lim=[]\n",
        "    y_lim=[]\n",
        "    for a in ax:\n",
        "        x_lim.append(a.get_xlim())\n",
        "        y_lim.append(a.get_ylim())\n",
        "\n",
        "    x_etr = np.max(np.abs(x_lim))\n",
        "    y_etr = np.max(np.abs(y_lim))\n",
        "\n",
        "    for a in ax:\n",
        "        a.set_xlim(-x_etr*1.1,x_etr*1.1)\n",
        "        if lock_y_max: a.set_ylim(0,y_etr*1.1)\n",
        "        else: a.set_ylim(0,a.get_ylim()[1]*1.1)\n",
        "        a.set_xlabel('Log2 Fold Change')\n",
        "        if use_adj: a.set_ylabel('-log10(adj-pvalue)')\n",
        "        else: a.set_ylabel('-log10(raw-pvalue)')\n",
        "        a.axhline(-np.log10(0.05),c='k',linestyle=\":\",linewidth=.6,label='p=0.05')\n",
        "\n",
        "    fig_volcano.suptitle(suptitle)\n",
        "    plt.tight_layout()\n",
        "    return fig_volcano,ax\n",
        "\n",
        "def deg_heatmap(deg_dict,gene_list=[],ax=None,lfcmax = 2.5,ax_cbar=None,suptitle=''):\n",
        "\n",
        "    try: ax.grid(visible=False)\n",
        "    except:\n",
        "        fig,axs = plt.subplots(1,2,figsize=(1+.5*len(deg_dict.keys()),len(gene_list)/14),width_ratios=[9,.5])\n",
        "\n",
        "        ax=axs[0]\n",
        "        ax_cbar = axs[1]\n",
        "        ax.grid(visible=False)\n",
        "    if len(gene_list)==0:gene_list = list()\n",
        "\n",
        "    heat_map_null=np.zeros([len(gene_list),len(deg_dict)])\n",
        "    heat_map_lfc = heat_map_null.copy()*np.nan\n",
        "    heat_map_p = heat_map_null.copy()*np.nan\n",
        "    for key,deg_df in deg_dict.items():\n",
        "        key_id = list(deg_dict.keys()).index(key)\n",
        "        for gene in gene_list:\n",
        "            i_gene = gene_list.index(gene)\n",
        "            try:\n",
        "                heat_map_lfc[i_gene,key_id] = deg_df.loc[gene,'logfoldchanges']\n",
        "            except:\n",
        "                None\n",
        "    cbar = ax.pcolorfast(heat_map_lfc,vmin=-lfcmax,vmax=lfcmax,cmap='bwr')\n",
        "    ax.set_yticks(np.arange(len(gene_list))+0.5, labels=gene_list,rotation=0)\n",
        "    ax.tick_params(length=0)\n",
        "\n",
        "    fmt_xlabels = list(deg_dict.keys())\n",
        "    fmt_xlabels = [xl.replace('_','\\n') for xl in fmt_xlabels]\n",
        "    ax.set_xticks(np.arange(len(deg_dict.keys()))+0.5, labels=fmt_xlabels,rotation=0)\n",
        "    ax.grid(visible=False)\n",
        "\n",
        "\n",
        "    for key,deg_df in deg_dict.items():\n",
        "        key_id = list(deg_dict.keys()).index(key)\n",
        "        for gene in gene_list:\n",
        "            i_gene = gene_list.index(gene)\n",
        "            try:\n",
        "                p = deg_df.loc[gene,'pvals']\n",
        "                if p<.1:\n",
        "                    p_str = str(p)[:5]\n",
        "                    l2fc = str(2**deg_df.loc[gene,'logfoldchanges'])[:3]\n",
        "                    tag = f\"{l2fc}X; p={p_str}\"\n",
        "                    tag = pval_to_star(p)\n",
        "                    ax.text(key_id+.5,i_gene+.5,tag,ha='center',va='top',rotation=0)\n",
        "            except: None\n",
        "\n",
        "    ax.tick_params(top=True, labeltop=True, bottom=False, labelbottom=False)\n",
        "\n",
        "    try:\n",
        "        plt.colorbar(cbar,cax=ax_cbar)\n",
        "        ax_cbar.set_ylabel('log2(Fold Change)')\n",
        "    except: plt.colorbar(cbar)\n",
        "    plt.suptitle(suptitle)\n",
        "    plt.tight_layout()\n",
        "    return plt.gcf(), ax, cbar\n",
        "\n",
        "def pval_to_star(p,specifics={(.05,.1):'\\'',(.05,1):'ns'},max_star=5):\n",
        "    for range,tag in specifics.items():\n",
        "        if p<=np.max(range) and p>np.min(range): return tag\n",
        "    if p<=0.5 and p>0.01: return '*'\n",
        "    nlogp=-np.log10(p)\n",
        "    return ('*'*int(nlogp))[:max_star]"
      ],
      "metadata": {
        "id": "hru563c6bCFU"
      },
      "id": "hru563c6bCFU",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "\n",
        "def single_deg_volcano(\n",
        "    deg_df,\n",
        "    gene_list=[],\n",
        "    label_cutoff=0.1,\n",
        "    title=None,\n",
        "    comp_string=None,\n",
        "    lock_y_max=True,\n",
        "    special_genes=[],\n",
        "    use_adj=False,\n",
        "    marker_size = .5\n",
        "    )   :\n",
        "    '''\n",
        "    Plot a single volcano plot for a DE comparison.\n",
        "\n",
        "    Parameters:\n",
        "        deg_df (DataFrame): DataFrame of DE results with gene names as index and columns including\n",
        "                            'logfoldchanges', 'pvals', and optionally 'pvals_adj'.\n",
        "        gene_list (list, optional): List of genes to be highlighted in the plot (default is an empty list).\n",
        "        label_cutoff (float, optional): Minimum -log10(p-value) threshold for labeling genes (default is 0.1).\n",
        "        title (str, optional): Title for the plot (default is an empty string).\n",
        "        lock_y_max (bool, optional): If True, locks the y-axis max across subplots; adjust to data range otherwise.\n",
        "        special_genes (list, optional): Genes to be specially highlighted in red (default is an empty list).\n",
        "        use_adj (bool, optional): If True, use adjusted p-values; otherwise, use raw p-values (default is False).\n",
        "\n",
        "    Returns:\n",
        "        fig_volcano (Figure): Matplotlib figure object for the volcano plot.\n",
        "        ax (Axes): Matplotlib axes object for the volcano plot.\n",
        "    '''\n",
        "\n",
        "    # Initialize the figure and axis for the volcano plot\n",
        "    fig_volcano, ax = plt.subplots(1, 1, figsize=(1.75, 1.75))\n",
        "\n",
        "    # Set the title for the plot\n",
        "    ax.set_title(title.replace('_', '\\n'))\n",
        "\n",
        "    # Extract necessary data\n",
        "    genes = deg_df.index\n",
        "    lfc = deg_df['logfoldchanges']\n",
        "    pval = deg_df['pvals_adj'] if use_adj else deg_df['pvals']\n",
        "    neg_log10_pval = -np.log10(pval)\n",
        "\n",
        "    # Plot each gene, marking those in gene_list and special_genes\n",
        "    for i, g in enumerate(genes):\n",
        "        if str(g) in gene_list:\n",
        "            if not np.isnan(lfc[i] * neg_log10_pval[i]):\n",
        "                # Highlight genes in gene_list\n",
        "                ax.scatter(lfc[i], neg_log10_pval[i], c='k', s=marker_size)\n",
        "                if pval[i] < label_cutoff:\n",
        "                    ax.text(lfc[i], neg_log10_pval[i], genes[i], rotation=45, fontsize=4)\n",
        "\n",
        "            # Highlight special genes in red\n",
        "            if g in special_genes:\n",
        "                if not np.isnan(lfc[i] * neg_log10_pval[i]):\n",
        "                    ax.scatter(lfc[i], neg_log10_pval[i], c='r', s=marker_size)\n",
        "                    ax.text(lfc[i], neg_log10_pval[i], genes[i], rotation=45, fontsize=4, color='r')\n",
        "\n",
        "    # Set axis limits and labels\n",
        "    x_lim = ax.get_xlim()\n",
        "    y_lim = ax.get_ylim()\n",
        "    x_etr = np.max(np.abs(x_lim))\n",
        "    y_etr = np.max(np.abs(y_lim))\n",
        "\n",
        "    ax.set_xlim(-x_etr * 1.1, x_etr * 1.1)\n",
        "    if lock_y_max:\n",
        "        ax.set_ylim(0, y_etr * 1.1)\n",
        "    else:\n",
        "        ax.set_ylim(0, ax.get_ylim()[1] * 1.1)\n",
        "\n",
        "    if comp_string is not None:\n",
        "        xlabel = f'{comp_string}\\nLog2(fold change)'\n",
        "    else:\n",
        "        xlabel = 'Log2(fold change)'\n",
        "    ax.set_xlabel(xlabel)\n",
        "    ax.set_ylabel('-log10(adj-pvalue)' if use_adj else '-log10(raw-pvalue)')\n",
        "    ax.axhline(-np.log10(0.05), c='k', linestyle=\":\", linewidth=0.6, label='p=0.05')\n",
        "    ax.axvline(0, c='k', linestyle=\"-\", linewidth=0.6, label='p=0.05')\n",
        "\n",
        "    plt.tight_layout()\n",
        "\n",
        "    return fig_volcano, ax\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "H0eyzWEVJ0Ip"
      },
      "id": "H0eyzWEVJ0Ip",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_gene_cluster(adata,gene):\n",
        "    gene_bool = z_score(sc.get.obs_df(adata,gene))>2\n",
        "    gene_cluster_list = np.array(adata.obs['louvain'][gene_bool]).astype(int)\n",
        "    gene_clust = st.mode(gene_cluster_list)[0][0]\n",
        "    return gene_clust\n",
        "\n",
        "\n",
        "sst_clust = get_gene_cluster(adata_GABA,'Sst')\n",
        "pv_clust = get_gene_cluster(adata_GABA,'Pvalb')\n",
        "\n",
        "print('sst_clust',sst_clust)\n",
        "print('pv_clust',pv_clust)"
      ],
      "metadata": {
        "id": "Uaok1R9Mlg6a"
      },
      "id": "Uaok1R9Mlg6a",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sst_clust\n",
        "print(adata_GABA.uns['log1p'])\n",
        "adata_GABA.uns['log1p'] = {'base': None}\n",
        "print(adata_GABA.uns['log1p'])"
      ],
      "metadata": {
        "id": "BPCeB8KIlfqW"
      },
      "id": "BPCeB8KIlfqW",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "set(adata_GABA.obs.E_type)"
      ],
      "metadata": {
        "id": "mPPlZ5MdPB0e"
      },
      "id": "mPPlZ5MdPB0e",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "analysis_pairs_34 = { a+'_E4vE3':(a+\"_E4\", a+\"_E3\") for a in ages}\n",
        "analysis_pairs_4f4 = {'15m+_E4vE4fKI':(\"15m+_E4\", \"15m+_E4fKI\")}\n",
        "analysis_pairs = analysis_pairs_34 | analysis_pairs_4f4\n",
        "print(analysis_pairs)"
      ],
      "metadata": {
        "id": "2eP-dc0lOw1K"
      },
      "id": "2eP-dc0lOw1K",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "###### SST Cluster KCN DEGs #####\n",
        "path='./DEG_Reuslts/KCN/SST_Cluster_DEG_KCN_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sst_cluster = adata_GABA.copy()[adata_GABA.obs['Cluster (nn)'].astype(int)==sst_clust,:]\n",
        "adata_sst_cluster.var['is_KCN'] = [g in kcn_list for g in adata_sst_cluster.var.index]\n",
        "adata_sst_cluster.obs['age_geno'] = adata_sst_cluster.obs['age_bin'].str.cat(adata_sst_cluster.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_sst_cluster.obs['age_bin'])))\n",
        "print(analysis_pairs)\n",
        "adata_name = 'SST_cluster_KCNs'\n",
        "sst_cluster_deg_data = deg_analysis(adata_sst_cluster,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "single_deg_volcano(sst_cluster_deg_data[1]['15m+_E4vE3'],gene_list=kcn_list,title='SST Cluster 15-20m+\\nApoE4 vs ApoE3 ',special_genes=['Kcnt2'])\n",
        "\n",
        "single_deg_volcano(sst_cluster_deg_data[1]['15m+_E4vE4fKI'],gene_list=kcn_list,title='SST Cluster 15-20m+\\nApoE4 vs nEKO',special_genes=['Kcnt2'])\n",
        "\n",
        "###### PV Cluster KCN DEGs #####\n",
        "path='./DEG_Reuslts/KCN/PV_Cluster_DEG_KCN_4v3_across_ages/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_pv_cluster = adata_GABA.copy()[adata_GABA.obs['Cluster (nn)'].astype(int)==pv_clust,:]\n",
        "adata_pv_cluster.var['is_KCN'] = [g in kcn_list for g in adata_pv_cluster.var.index]\n",
        "adata_pv_cluster.obs['age_geno'] = adata_pv_cluster.obs['age_bin'].str.cat(adata_pv_cluster.obs['E_type'],sep='_')\n",
        "ages = sorted(list(set(adata_pv_cluster.obs['age_bin'])))\n",
        "adata_name = 'pv_cluster_KCNs'\n",
        "pv_cluster_deg_data = deg_analysis(adata_pv_cluster,path,analysis_pairs,group_key = 'age_geno',log2fc_extrema=[-15,15],to_plot=False)\n",
        "single_deg_volcano(pv_cluster_deg_data[1]['15m+_E4vE3'],gene_list=kcn_list,title='PV Cluster 15-20m+\\nApoE4 vs ApoE3',special_genes=['Kcnt2'])\n",
        "single_deg_volcano(pv_cluster_deg_data[1]['15m+_E4vE4fKI'],gene_list=kcn_list,title='PV Cluster 15-20m+\\nApoE4 vs nEKO',special_genes=['Kcnt2'])"
      ],
      "metadata": {
        "id": "CWKYhY5ZqOXc"
      },
      "id": "CWKYhY5ZqOXc",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "########### MULTIMODAL DEGS #######################\n",
        "#### Clusters\n",
        "path = './DEG_Reuslts/KCN/Cluster_DEGs_multivar/'\n",
        "try: os.makedirs(path)\n",
        "except: None\n",
        "adata_sstpv_clusters = adata_GABA.copy()[np.logical_or(adata_GABA.obs['Cluster (nn)'].astype(int)==sst_clust,adata_GABA.obs['Cluster (nn)'].astype(int)==pv_clust),:]\n",
        "keys_of_interest = ['age_bin','louvain','E_type']\n",
        "adata_sstpv_clusters,combo_key = deg_set_up(adata_sstpv_clusters,keys_of_interest)\n",
        "analysis_pairs = {'15m_SST_E4 vs E3_(genotype)': (f'15m+_{sst_clust}_E4', f'15m+_{sst_clust}_E3'),\n",
        "                #   'SST_E4_15m vs 5m_(age)': (f'15m+_{sst_clust}_E4', f'05m_{sst_clust}_E4'),\n",
        "                #   'SST_E4_15m vs 10m_(age)': (f'15m+_{sst_clust}_E4', f'10m_{sst_clust}_E4'),\n",
        "                  '15m_E4_SST vs PV_(celltype)': (f'15m+_{sst_clust}_E4', f'15m+_{pv_clust}_E4')}\n",
        "multimodal_cluster_deg_data = deg_analysis(adata_sstpv_clusters,path,analysis_pairs,group_key = combo_key,log2fc_extrema=[-15,15],to_plot=False)\n",
        "single_deg_volcano(multimodal_cluster_deg_data[1]['15m_E4_SST vs PV_(celltype)'],gene_list=kcn_list,title='15-20m+ ApoE4 SST vs PV Cluster',special_genes=['Kcnt2'])\n"
      ],
      "metadata": {
        "id": "29ALUAijx3pP"
      },
      "id": "29ALUAijx3pP",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "MMm4Pp8a0bDy"
      },
      "id": "MMm4Pp8a0bDy"
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Figure Plotting and Saving\n",
        "\"\"\"\n",
        "# Cluster plots\n",
        "cluster_fig, ax = show_clusters(adata_GABA, 'Cluster (nn)')\n",
        "cluster_fig.savefig(\"cluster_plot.svg\", format=\"svg\")\n",
        "cluster_fig.savefig(\"cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "sst_cluster_fig, ax = show_gene_loading(adata_GABA, 'Sst')\n",
        "sst_cluster_fig.savefig(\"sst_cluster_plot.svg\", format=\"svg\")\n",
        "sst_cluster_fig.savefig(\"sst_cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "pv_cluster_fig, ax = show_gene_loading(adata_GABA, 'Pvalb')\n",
        "pv_cluster_fig.savefig(\"pv_cluster_plot.svg\", format=\"svg\")\n",
        "pv_cluster_fig.savefig(\"pv_cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "vip_cluster_fig, ax = show_gene_loading(adata_GABA, 'Vip')\n",
        "vip_cluster_fig.savefig(\"vip_cluster_plot.svg\", format=\"svg\")\n",
        "vip_cluster_fig.savefig(\"vip_cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "reln_cluster_fig, ax = show_gene_loading(adata_GABA, 'Reln')\n",
        "reln_cluster_fig.savefig(\"reln_cluster_plot.svg\", format=\"svg\")\n",
        "reln_cluster_fig.savefig(\"reln_cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "Lamp5_cluster_fig, ax = show_gene_loading(adata_GABA, 'Lamp5')\n",
        "Lamp5_cluster_fig.savefig(\"lamp5_cluster_plot.svg\", format=\"svg\")\n",
        "Lamp5_cluster_fig.savefig(\"lamp5_cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "Apoe_cluster_fig, ax = show_gene_loading(adata_GABA, 'Apoe')\n",
        "Apoe_cluster_fig.savefig(\"apoe_cluster_plot.svg\", format=\"svg\")\n",
        "Apoe_cluster_fig.savefig(\"apoe_cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "Syn1_cluster_fig, ax = show_gene_loading(adata_GABA, 'Syn1')\n",
        "Syn1_cluster_fig.savefig(\"syn1_cluster_plot.svg\", format=\"svg\")\n",
        "Syn1_cluster_fig.savefig(\"syn1_cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "Gfap_cluster_fig, ax = show_gene_loading(adata_GABA, 'Gfap')\n",
        "Gfap_cluster_fig.savefig(\"gfap_cluster_plot.svg\", format=\"svg\")\n",
        "Gfap_cluster_fig.savefig(\"gfap_cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "Kcnt2_cluster_fig, ax = show_gene_loading(adata_GABA, 'Kcnt2')\n",
        "Kcnt2_cluster_fig.savefig(\"kcnt2_cluster_plot.svg\", format=\"svg\")\n",
        "Kcnt2_cluster_fig.savefig(\"kcnt2_cluster_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "# Volcano plots\n",
        "sst_volc, ax = single_deg_volcano(sst_cluster_deg_data[1]['15m+_E4vE3'], gene_list=kcn_list, title='SST Cluster 15-20m+',comp_string='ApoE4 vs ApoE3', special_genes=['Kcnt2'])\n",
        "sst_volc.savefig(\"sst_volcano_plot.svg\", format=\"svg\")\n",
        "sst_volc.savefig(\"sst_volcano_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "pv_volc, ax = single_deg_volcano(pv_cluster_deg_data[1]['15m+_E4vE3'], gene_list=kcn_list, title='PV Cluster 15-20m+',comp_string='ApoE4 vs ApoE3', special_genes=['Kcnt2'])\n",
        "pv_volc.savefig(\"pv_volcano_plot.svg\", format=\"svg\")\n",
        "pv_volc.savefig(\"pv_volcano_plot.jpeg\", format=\"jpeg\")\n",
        "\n",
        "sst_pv_volc, ax = single_deg_volcano(multimodal_cluster_deg_data[1]['15m_E4_SST vs PV_(celltype)'], gene_list=kcn_list, title='15-20m+ ApoE4' ,comp_string='SST vs PV Cluster', special_genes=['Kcnt2'])\n",
        "sst_pv_volc.savefig(\"sst_pv_volcano_plot.svg\", format=\"svg\")\n",
        "sst_pv_volc.savefig(\"sst_pv_volcano_plot.jpeg\", format=\"jpeg\")\n",
        "\n"
      ],
      "metadata": {
        "id": "4fgSG2fDKc8s"
      },
      "id": "4fgSG2fDKc8s",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install svgutils\n",
        "\n",
        "\n",
        "# from svgutils.compose import *\n",
        "from svgutils import compose as svguc\n",
        "\n",
        "\n",
        "\n",
        "row_1 = 0\n",
        "row_2 =row_1+240\n",
        "row_3 = row_2+160\n",
        "row_4 = row_3+160\n",
        "dpi = 96\n",
        "width = 7.25*dpi/2\n",
        "length = 9.4*dpi\n",
        "scale_factor= 1.33\n",
        "weight='bold'\n",
        "label_size = 12\n",
        "\n",
        "\n",
        "primary_figure_layout_1 = svguc.Figure(str(width), str(length),\n",
        "\n",
        "            svguc.Panel(svguc.SVG(\"cluster_plot.svg\"),\n",
        "            svguc.Text(\"a\", 0, 10, size=label_size,weight=weight,font='arial')\n",
        "            ).scale(scale_factor).move(width*0, row_1),\n",
        "\n",
        "            svguc.Panel(svguc.SVG(\"sst_cluster_plot.svg\"),\n",
        "            svguc.Text(\"b\", 0, 10, size=label_size,weight=weight,font='arial')\n",
        "            ).scale(scale_factor).move(width*0.0, row_2),\n",
        "\n",
        "            svguc.Panel(svguc.SVG(\"pv_cluster_plot.svg\"),\n",
        "            svguc.Text(\"c\", 0, 10, size=label_size,weight=weight,font='arial')\n",
        "            ).scale(scale_factor).move(width*0.5, row_2),\n",
        "\n",
        "            svguc.Panel(svguc.SVG(\"sst_volcano_plot.svg\"),\n",
        "            svguc.Text(\"d\", 0, 10, size=label_size,weight=weight,font='arial')\n",
        "            ).scale(scale_factor).move(width*0.0, row_3),\n",
        "\n",
        "            svguc.Panel(svguc.SVG(\"pv_volcano_plot.svg\"),\n",
        "            svguc.Text(\"e\", 0, 10, size=label_size,weight=weight,font='arial')\n",
        "            ).scale(scale_factor).move(width*0.5, row_3),\n",
        "\n",
        "            svguc.Panel(svguc.SVG(\"sst_pv_volcano_plot.svg\"),\n",
        "            svguc.Text(\"f\", 0, 10, size=label_size,weight=weight,font='arial')\n",
        "            ).scale(scale_factor).move(width*0, row_4),\n",
        "\n",
        "            svguc.Panel(svguc.SVG(\"kcnt2_cluster_plot.svg\"),\n",
        "            svguc.Text(\"g\", 0, 10, size=label_size,weight=weight,font='arial')\n",
        "            ).scale(scale_factor).move(width*0.5, row_4),\n",
        "\n",
        "    )\n",
        "\n",
        "display(primary_figure_layout_1)\n",
        "primary_figure_layout_1.save(\"primary_figure_layout_1.svg\")"
      ],
      "metadata": {
        "id": "psJYcIdVNwgg"
      },
      "id": "psJYcIdVNwgg",
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "private_outputs": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.4"
    },
    "varInspector": {
      "cols": {
        "lenName": 16,
        "lenType": 16,
        "lenVar": 40
      },
      "kernels_config": {
        "python": {
          "delete_cmd_postfix": "",
          "delete_cmd_prefix": "del ",
          "library": "var_list.py",
          "varRefreshCmd": "print(var_dic_list())"
        },
        "r": {
          "delete_cmd_postfix": ") ",
          "delete_cmd_prefix": "rm(",
          "library": "var_list.r",
          "varRefreshCmd": "cat(var_dic_list()) "
        }
      },
      "types_to_exclude": [
        "module",
        "function",
        "builtin_function_or_method",
        "instance",
        "_Feature"
      ],
      "window_display": false
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}