{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/dtabuena/Workshop/blob/main/Ref_Map/KZ_BuildRef.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "b294428a",
      "metadata": {
        "id": "b294428a",
        "scrolled": true,
        "tags": []
      },
      "outputs": [],
      "source": [
        "!pip install scanpy --quiet\n",
        "!pip install pybiomart --quiet\n",
        "!pip install python-igraph --quiet\n",
        "!pip install louvain --quiet\n",
        "!pip install pynndescent --quiet\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import h5py\n",
        "import numpy as np\n",
        "import scipy as sci\n",
        "from matplotlib import pyplot as plt\n",
        "import scanpy as sc\n",
        "import tarfile\n",
        "import os\n",
        "import anndata as ad\n",
        "import pandas as pd\n",
        "import pybiomart\n",
        "from tqdm import tqdm\n",
        "import urllib.request\n",
        "from IPython.display import clear_output\n",
        "from matplotlib.pyplot import rc_context\n",
        "from scipy import stats as st\n",
        "import logging\n",
        "logging.Logger('my_log').setLevel('INFO')\n",
        "\n",
        "############# SPECIFIC CONFIG #############\n",
        "working_dir = r\"C:\\Users\\dennis.tabuena\\Gladstone Dropbox\\Dennis Tabuena\\0_Projects\\_SeqRef\"\n",
        "os.chdir(working_dir)\n",
        "zalocusky_url = 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE167497&format=file'\n",
        "geo_zalo_filename = 'zalocusky_indiv.tar'\n",
        "\n",
        "import urllib\n",
        "response = urllib.request.urlretrieve('https://raw.githubusercontent.com/dtabuena/Resources/main/Matplotlib_Config/Load_FS6.py','Load_FS6.py')\n",
        "%run Load_FS6.py\n",
        "\n"
      ],
      "metadata": {
        "id": "YoAmeyEda0Sq"
      },
      "id": "YoAmeyEda0Sq",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4514aef3",
      "metadata": {
        "id": "4514aef3",
        "tags": []
      },
      "outputs": [],
      "source": [
        "def trim_key(k):\n",
        "    floxed_dict = {'GSM5106175_YH_KZ03_01':('E3fKI_Syn_Cre602_15m','GSM5106175_602_E3fKI_15_XX'),\n",
        "                   'GSM5106176_YH_KZ03_03':('E4fKI_Syn_Cre475_15m','GSM5106176_475_E4fKI_15_XX')}\n",
        "    for f in floxed_dict.keys():\n",
        "        if f in k: return floxed_dict[f][1]\n",
        "    k = k.replace('_raw_gene_bc_matrices_h5.h5',\"\")\n",
        "    return k\n",
        "\n",
        "def query_capitilaziation(gene,adata):\n",
        "    try:\n",
        "        return adata.var.index[ [g.lower() for g in list(adata.var.index)].index(gene.lower()) ]\n",
        "    except:\n",
        "        return gene + ' not_found'\n",
        "\n",
        "def z_score(x,axis=-1):\n",
        "    x=np.array(x)\n",
        "    return (x-np.mean(x,axis=axis))/np.std(x,axis=axis)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "c5a6ad5b",
      "metadata": {
        "tags": [],
        "id": "c5a6ad5b"
      },
      "outputs": [],
      "source": [
        "def pull_gene_annots(csv_loc='./mmusculus_coding_noncoding.csv',\n",
        "                     my_git='https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv',\n",
        "                     biomart_name='mmusculus',\n",
        "                     biomart_keys=[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\",\"peptide\"]):\n",
        "\n",
        "    if os.path.exists('./mmusculus_coding_noncoding.csv'):\n",
        "        print( 'Use local copy of musmus')\n",
        "        annot_dd = pd.read_csv('./mmusculus_coding_noncoding.csv').set_index(\"external_gene_name\")\n",
        "    else:\n",
        "        try:\n",
        "            print( 'attempting to pull mus mus from git...')\n",
        "            musmus_link = 'https://raw.githubusercontent.com/dtabuena/Resources/main/Genetics/mmusculus_coding_noncoding.csv'\n",
        "            filename = './mmusculus_coding_noncoding.csv'\n",
        "            urllib.request.urlretrieve(musmus_link, filename)\n",
        "            annot_dd = pd.read_csv('./mmusculus_coding_noncoding.csv').set_index(\"external_gene_name\")\n",
        "        except:\n",
        "            print('attempting to pull mus mus from biomart...')\n",
        "            annot = sc.queries.biomart_annotations(\"mmusculus\",[\"ensembl_gene_id\", \"chromosome_name\",\"transcript_biotype\",\"external_gene_name\"],).set_index('ensembl_gene_id')\n",
        "            uniq_inds = list(set(list(annot.index)))\n",
        "            for r in tqdm(uniq_inds):\n",
        "                match_bool = annot.index.str.contains(r)\n",
        "                if np.sum(match_bool)>1:\n",
        "                    new_val ='__'.join(list(annot.loc[r,'transcript_biotype']))\n",
        "                    annot.at[r,'transcript_biotype']=new_val\n",
        "            annot['is_coding']= annot.transcript_biotype.str.contains('coding')\n",
        "            annot_dd = annot.drop_duplicates().set_index(\"external_gene_name\")\n",
        "            annot_dd.to_csv('./mmusculus_coding_noncoding.csv')\n",
        "\n",
        "    coding_list = annot_dd.index[ annot_dd['is_coding'] ].to_list()\n",
        "    return coding_list, annot_dd\n",
        "\n",
        "\n",
        "def preprocess_andata10x(adata_og,pct_mito=0.25,min_genes=500,max_genes=2400,min_counts=500,max_counts=4500,keep_NC=False):\n",
        "\n",
        "    print('pulling gene annotations...')\n",
        "    coding_list, _ = pull_gene_annots()\n",
        "    adata_og.var['mt'] = adata_og.var_names.str.startswith('mt-')\n",
        "    adata_og.var['coding'] = [gene in coding_list for gene in adata_og.var_names]\n",
        "    sc.pp.calculate_qc_metrics(adata_og, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
        "\n",
        "    adata_QC = adata_og.copy()\n",
        "\n",
        "    print('Filtering...')\n",
        "    adata_QC = adata_QC[adata_QC.obs.pct_counts_mt < pct_mito, :]\n",
        "    print(str(np.sum(adata_og.obs.pct_counts_mt <pct_mito)) + f' cells with >{pct_mito}% mt removed')\n",
        "\n",
        "    if keep_NC:\n",
        "        print('keep noncoding')\n",
        "    else:\n",
        "        adata_QC = adata_QC[:, adata_QC.var.coding]\n",
        "        print(str(np.sum(np.logical_not(adata_og.var.coding))) + ' non coding genes removed')\n",
        "\n",
        "    sc.pp.filter_cells(adata_QC, min_genes=min_genes)\n",
        "    sc.pp.filter_cells(adata_QC, max_genes=max_genes)\n",
        "    sc.pp.filter_cells(adata_QC, min_counts=min_counts)\n",
        "    sc.pp.filter_cells(adata_QC, max_counts=max_counts)\n",
        "    fig,ax=plt.subplots(1,figsize=(1.5,1.5))\n",
        "    sc.pl.scatter(adata_QC, x='total_counts', y='n_genes_by_counts',ax=ax)\n",
        "\n",
        "    return adata_QC,adata_og\n",
        "\n",
        "def high_var_genes_dim_reduc(adata,min_mean = 0.25,max_mean = 4,min_disp=0.55):\n",
        "    ''' The gene expression matrices were then log-normalized with a scale factor of 10,000,\n",
        "    using the Seurat NormalizeData function57,58. Highly dispersed genes were selected using\n",
        "    the Seurat FindVariableGenes function57,58,filtering for an average expression range of\n",
        "    0.25â€“4 and a minimum dispersion of 0.55, resulting in a list of 2,197 genes.'''\n",
        "    adata.raw = adata\n",
        "    sc.pp.normalize_total(adata, target_sum=10000)\n",
        "    sc.pp.log1p(adata)\n",
        "    adata.uns['log1p'] = {'base': None}\n",
        "    print(adata.uns['log1p'])\n",
        "    sc.pp.highly_variable_genes(adata, min_mean=min_mean, max_mean=max_mean, min_disp=min_disp)\n",
        "    with rc_context({'figure.figsize': (1.5, 1.5)}):\n",
        "        sc.pl.highly_variable_genes(adata)\n",
        "    plt.tight_layout()\n",
        "    print(np.sum(adata.var['highly_variable']),'hv genes')\n",
        "\n",
        "\n",
        "    #### PCA\n",
        "    sc.tl.pca(adata, svd_solver='arpack',n_comps=50)\n",
        "    fig,ax=plt.subplots(figsize=(1,1))\n",
        "    ax.plot(adata.uns['pca']['variance_ratio'][:25],'ok',markersize=1)\n",
        "    quiet_PCA_plots(adata,['E_type','age_bin','mouse_ID'],pc_pairs=[(0,1),(2,3)])\n",
        "\n",
        "    return adata\n",
        "\n",
        "def quiet_PCA_plots(adata,key_list,figsize=(2,2),pc_pairs=[(0,1)]):\n",
        "    fig,ax=plt.subplots(1*len(pc_pairs),len(key_list),figsize=(figsize[0]*len(key_list),figsize[1]*len(pc_pairs)))\n",
        "    for ip,pair in enumerate(pc_pairs):\n",
        "        if len(key_list) == 1: ax=[ax]\n",
        "        for key_ind,key in enumerate(key_list):\n",
        "            key_types = sorted(list(set( adata.obs[key] )))\n",
        "            for k in key_types:\n",
        "                is_k = adata.obs[key]==k\n",
        "                ax[ip,key_ind].scatter(adata.obsm['X_pca'][is_k,pair[0]],adata.obsm['X_pca'][is_k,pair[1]],s=2,marker='.',linewidth=0,edgecolors=None,label=k)\n",
        "                ax[ip,key_ind].set_xlabel(f'PC{pair[0]}')\n",
        "                ax[ip,key_ind].set_ylabel(f'PC{pair[1]}')\n",
        "            if len(key_types)<8: ax[ip,key_ind].legend(key_types,loc='best',markerscale=3)\n",
        "            ax[ip,key_ind].set_title(key)\n",
        "            plt.tight_layout()\n",
        "    return None\n",
        "\n",
        "\n",
        "def umap_and_cluster(adata, n_neighbors=10, n_pcs=20,resolution=.6,plot_keys=['Cluster (nn)'],size = 1,to_plot=True):\n",
        "    sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=n_pcs,random_state=42, use_rep='X_pca')\n",
        "    sc.tl.louvain(adata,resolution=resolution,random_state=42)\n",
        "    sc.tl.paga(adata)\n",
        "    sc.tl.umap(adata,random_state=42)\n",
        "    adata.obs['Cluster (nn)']= adata.obs['louvain']\n",
        "    if to_plot:\n",
        "        with rc_context({'figure.figsize': (2.5, 2.5)}):\n",
        "            sc.pl.umap(adata,add_outline=False, legend_loc='on data', color=plot_keys,size=size)\n",
        "    return adata\n",
        "\n",
        "def explore_umap(adata_GABA,key_list=[],size=1,legend_loc=None):\n",
        "    with rc_context({'figure.figsize': (1.5,1.5)}):\n",
        "        sc.pl.umap(adata_GABA, legend_loc=legend_loc, color=key_list,vmin=-0.5,size=size,cmap='Purples') # add_outline=True,\n",
        "        plt.tight_layout()\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def marker_analysis(adata):\n",
        "    #### Get Marker Genes\n",
        "    sc.tl.rank_genes_groups(adata, 'louvain', method='wilcoxon',key_added='maker_genes',pts=True,use_raw =True)\n",
        "    sc.tl.rank_genes_groups(adata, 'louvain', method='logreg',pts=True,use_raw =True)\n",
        "    maker_genes_df = pd.DataFrame(adata.uns['maker_genes']['names'])\n",
        "    maker_genes_df_LFC = pd.DataFrame(adata.uns['maker_genes']['logfoldchanges'])\n",
        "    top_30 = maker_genes_df[:30]\n",
        "    lfc_thresh=4\n",
        "    marker_genes = [m for m in set(top_30[maker_genes_df_LFC[:30]>lfc_thresh].values.flatten()) if isinstance(m,str)]\n",
        "    sc.pl.stacked_violin(adata, marker_genes, groupby='louvain');\n",
        "    return marker_genes\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "def show_clusters(adata_GABA, key, cmap='tab20', legend_fontsize='small',size = 0.4):\n",
        "    assert key in adata_GABA.obs, f\"Key '{key}' not found in adata_GABA.obs.\"\n",
        "\n",
        "    umap_coords = adata_GABA.obsm['X_umap']\n",
        "\n",
        "\n",
        "    # Create the figure with subplots for scatter and legend\n",
        "    fig, (ax_scatter, ax_legend) = plt.subplots(1, 2, figsize=(3.5, 2.5), gridspec_kw={'width_ratios': [5, 1]})\n",
        "\n",
        "    # Get cluster colors\n",
        "    colors = adata_GABA.obs[key].values.astype(float)\n",
        "    scatter = ax_scatter.scatter(\n",
        "        umap_coords[:, 0], umap_coords[:, 1],\n",
        "        c=colors, cmap=cmap, s=size, vmin=-0.5, edgecolor='none'  # Set edgecolor to 'none'\n",
        "    )\n",
        "\n",
        "    # Set equal aspect ratio to make axes square\n",
        "    ax_scatter.set_aspect('equal', adjustable='box')\n",
        "    ax_scatter.set_xlabel('umap 1',ha='left')\n",
        "    ax_scatter.xaxis.set_label_coords(0, 0.0)\n",
        "    ax_scatter.set_ylabel('umap 2',ha='left')\n",
        "    ax_scatter.yaxis.set_label_coords(0.0, 0)\n",
        "    ax_scatter.set_xticks([])\n",
        "    ax_scatter.set_yticks([])\n",
        "\n",
        "    unique_clusters = np.unique(colors)\n",
        "    handles = []\n",
        "    labels = []\n",
        "\n",
        "    # Create custom legend handles for unique clusters\n",
        "    for cluster in unique_clusters:\n",
        "        color_i = scatter.cmap(scatter.norm(cluster))\n",
        "        handle = plt.Line2D([], [], marker='o', linestyle='None', color=color_i,\n",
        "                             markerfacecolor=color_i,\n",
        "                             markersize=4.5)  # Smaller markers\n",
        "        handles.append(handle)\n",
        "        labels.append(f'Cluster {int(cluster)}')\n",
        "\n",
        "    # Turn off the axis for the legend subplot\n",
        "    ax_legend.axis('off')\n",
        "\n",
        "    # Add the custom legend to the legend axis with specified font size\n",
        "    ax_legend.legend(handles, labels, loc='center', fontsize=legend_fontsize, markerscale=0.75,\n",
        "                     ncol=1, frameon=False)\n",
        "    plt.tight_layout()\n",
        "    plt.show()\n",
        "\n",
        "    return fig, (ax_scatter, ax_legend)\n",
        "\n",
        "# Example calls to the function\n",
        "# show_clusters(adata_GABA, 'Cluster (nn)')\n",
        "\n",
        "\n",
        "def show_gene_loading(adata, gene, cmap='Purples',size = 0.4,alpha=0.8):\n",
        "    assert gene in adata.var_names, f\"Gene '{gene}' not found in adata.var_names.\"\n",
        "\n",
        "    umap_coords = adata.obsm['X_umap']\n",
        "\n",
        "\n",
        "    # Get gene expression values for the specified gene\n",
        "    gene_expression = adata[:, gene].X.toarray().flatten()  # Convert to dense array and flatten\n",
        "\n",
        "    # Create the figure\n",
        "    fig = plt.figure(figsize=(3, 1.5))\n",
        "\n",
        "    # Define positions for scatter and colorbar axes\n",
        "    ax_scatter = fig.add_axes([0.0, 0.1, 0.5, 0.8])  # [left, bottom, width, height]\n",
        "    ax_cbar = fig.add_axes([0.50, 0.1, 0.01, 0.8])  # Slightly to the right of the scatter plot\n",
        "\n",
        "    # Scatter plot\n",
        "    scatter = ax_scatter.scatter(\n",
        "        umap_coords[:, 0], umap_coords[:, 1],\n",
        "        c=gene_expression, cmap=cmap, s=size, edgecolor='none',alpha=alpha,vmin=-.45,vmax=2.75)\n",
        "\n",
        "    # Set equal aspect ratio to make axes square\n",
        "    ax_scatter.set_aspect('equal', adjustable='box')\n",
        "    ax_scatter.set_xlabel('umap 1',ha='left')\n",
        "    ax_scatter.xaxis.set_label_coords(0, 0.0)\n",
        "    ax_scatter.set_ylabel('umap 2',ha='left')\n",
        "    ax_scatter.yaxis.set_label_coords(0.0, 0)\n",
        "    ax_scatter.set_xticks([])\n",
        "    ax_scatter.set_yticks([])\n",
        "\n",
        "    # Colorbar\n",
        "    cbar = fig.colorbar(scatter, cax=ax_cbar, orientation='vertical')\n",
        "    cbar.set_label(f'stdev/mean', rotation=270, labelpad=10)\n",
        "    ax_scatter.set_title(gene.upper())\n",
        "    # Ensure colorbar is labeled\n",
        "    # ax_cbar.set_ylabel(f'{gene} Expression', rotation=270, labelpad=10)\n",
        "\n",
        "    # Adjust layout and show the plot\n",
        "    plt.tight_layout()\n",
        "    plt.show()\n",
        "\n",
        "    return fig, (ax_scatter, ax_cbar)\n",
        "\n",
        "# Example call to the function\n",
        "# show_gene_loading(adata_GABA, 'Sst')\n",
        "\n",
        "# cluster_fig, ax = show_clusters(adata_GABA, 'Cluster (nn)')\n",
        "# cluster_fig.savefig(\"cluster_plot.svg\", format=\"svg\")\n",
        "# cluster_fig.savefig(\"cluster_plot.jpeg\", format=\"jpeg\")\n",
        "# show_gene_loading(adata_full, 'Prox1')"
      ],
      "metadata": {
        "id": "aM-_rLAPGNdx"
      },
      "id": "aM-_rLAPGNdx",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "################# INITIALIZE DIRECTORY DOWNLOAD FROM GEO\n",
        "os.chdir(working_dir)\n",
        "os.makedirs('./indiv_animal_results', exist_ok=True)\n",
        "urllib.request.urlretrieve(zalocusky_url, './indiv_animal_results/'+geo_zalo_filename)\n",
        "my_tar = tarfile.open('./indiv_animal_results/'+geo_zalo_filename)\n",
        "my_tar.extractall('./indiv_animal_results') # specify which folder to extract to\n",
        "my_tar.close()\n",
        "# for f in os.listdir('./indiv_animal_results'):\n",
        "#     print(f)\n"
      ],
      "metadata": {
        "id": "R4iZj93WnnKd"
      },
      "id": "R4iZj93WnnKd",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "################# Read, Combine, and Sample Split multiple 10x's\n",
        "\n",
        "adata_dict = {}\n",
        "for f in os.listdir('./indiv_animal_results'):\n",
        "    if '.h5' in f:\n",
        "        a = sc.read_10x_h5('./indiv_animal_results/'+f)\n",
        "        a.var_names_make_unique()\n",
        "        sample_code = trim_key(f)\n",
        "        a.obs['age_bin'] = str(int(np.ceil( int(sample_code.split(\"_\")[3])/5)*5))+'m'\n",
        "        a.obs['E_type'] = sample_code.split(\"_\")[2]\n",
        "        a.obs['mouse_ID'] = sample_code.split(\"_\")[1]\n",
        "        a.obs['well'] = sample_code.split(\"_\")[4]\n",
        "        a.obs['GSM'] = sample_code.split(\"_\")[0]\n",
        "        adata_dict[sample_code.split(\"_\")[0]] = a\n",
        "adata = ad.concat(adata_dict,axis = 0,label=\"Sample\",index_unique=\"_\")\n",
        "adata.obs.E_type\n",
        "# adata = adata[['fKI' not in t for t in adata.obs.E_type], :]\n",
        "adata_dict = {}\n",
        "clear_output()\n",
        "adata.write_h5ad(filename='./kz_adata_raw.h5')\n",
        "\n",
        "\n",
        "adata_QC = preprocess_andata10x(adata,keep_NC=True)[0]\n",
        "adata_QC.write_h5ad(filename='./kz_adata_qc.h5')\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "cx-L5v6n2FGP"
      },
      "id": "cx-L5v6n2FGP",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "############## QC Filter #####################\n",
        "adata_full = adata_QC.copy()\n",
        "age_dict = {'5m':'05m', '10m': '10m','15m': '15m','20m': '20m'}\n",
        "adata_full.obs['age_bin'] = [ age_dict[a] for a in adata_full.obs['age_bin'] ]\n",
        "\n",
        "sc.pp.filter_genes(adata_full, min_counts=50)\n",
        "\n",
        "############## Dim Reduction #####################\n",
        "adata_full = high_var_genes_dim_reduc(adata_full)\n",
        "adata_full.write_h5ad(filename='./adata_full.h5')\n",
        "print(adata_full.uns['log1p'])\n",
        "\n"
      ],
      "metadata": {
        "id": "thajIZvhtaW0"
      },
      "id": "thajIZvhtaW0",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "############## Dim Reduction #####################\n",
        "adata_full = ad.read_h5ad('./adata_full.h5')\n",
        "adata_full = umap_and_cluster(adata_full, n_neighbors=15, n_pcs=20,resolution=.6,to_plot=False)\n",
        "show_clusters(adata_full, 'Cluster (nn)')"
      ],
      "metadata": {
        "id": "8B2PtY0ibTHJ"
      },
      "id": "8B2PtY0ibTHJ",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "tab40_colors[0]"
      ],
      "metadata": {
        "id": "wtyegsdx1uAp"
      },
      "id": "wtyegsdx1uAp",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f246d0a6",
      "metadata": {
        "scrolled": true,
        "tags": [],
        "id": "f246d0a6"
      },
      "outputs": [],
      "source": [
        "tab20b_colors = plt.cm.get_cmap('tab20b').colors\n",
        "tab20c_colors = plt.cm.get_cmap('tab20c').colors\n",
        "tab40_colors = np.concatenate([tab20c_colors, tab20b_colors[4:]])\n",
        "tab40 = mpl.colors.ListedColormap(tab40_colors)\n",
        "\n",
        "\n",
        "\n",
        "show_clusters(adata_full, 'Cluster (nn)',cmap=tab40)\n",
        "# show_gene_loading(adata_full, 'Sst')\n",
        "# show_gene_loading(adata_full, 'Pvalb')\n",
        "# show_gene_loading(adata_full, 'Vip')\n",
        "# show_gene_loading(adata_full, 'Reln')\n",
        "# show_gene_loading(adata_full, 'Apoe')\n",
        "show_gene_loading(adata_full, 'Syn1',size=.1,alpha=1)\n",
        "show_gene_loading(adata_full, 'Slc17a7',size=.1,alpha=1)\n",
        "show_gene_loading(adata_full, 'Prox1',size=.1,alpha=1)\n",
        "show_gene_loading(adata_full, 'Pdzd2',size=.1,alpha=1)\n",
        "\n",
        "\n",
        "# show_gene_loading(adata_full, 'Gfap')\n",
        "# show_gene_loading(adata_full, 'Gad1')\n",
        "# show_gene_loading(adata_full, 'Gad2')\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "DGC_adata = adata_full[adata_full.obs['louvain'].isin(['0', '2', '4']), :]\n",
        "DGC_adata.obsm = {}\n",
        "DGC_adata.uns = {}\n",
        "sc.tl.pca(DGC_adata, n_comps=50)\n",
        "sc.pp.scale(DGC_adata)\n",
        "DGC_adata = umap_and_cluster(DGC_adata, n_neighbors=25, n_pcs=30,resolution=.6,to_plot=False)\n",
        "show_clusters(DGC_adata, 'Cluster (nn)')"
      ],
      "metadata": {
        "id": "F70xJ9FVpLCf"
      },
      "id": "F70xJ9FVpLCf",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sc.pl.pca_variance_ratio(DGC_adata)\n",
        "DGC_adata.write_h5ad(filename='./DGC_adata.h5')"
      ],
      "metadata": {
        "id": "hyq87n3ErAcg"
      },
      "id": "hyq87n3ErAcg",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# show_clusters(DGC_adata, 'Cluster (nn)')\n",
        "show_gene_loading(DGC_adata, 'Apoe',size = 0.5,alpha=0.5)\n",
        "show_gene_loading(DGC_adata, 'Syn1',size = 0.5,alpha=0.5)\n",
        "show_gene_loading(DGC_adata, 'Slc17a7',size = 0.5,alpha=0.5)\n",
        "show_gene_loading(DGC_adata, 'Prox1',size = 0.5,alpha=0.5)\n",
        "show_gene_loading(DGC_adata, 'Pdzd2',size = 0.5,alpha=0.5)\n"
      ],
      "metadata": {
        "id": "GH-G4O7PoW_e"
      },
      "id": "GH-G4O7PoW_e",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sc.tl.rank_genes_groups(DGC_adata, groupby='louvain', method='wilcoxon')"
      ],
      "metadata": {
        "id": "3k3BA_lEilBB"
      },
      "id": "3k3BA_lEilBB",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from scipy.stats import fisher_exact\n",
        "from statsmodels.stats.multitest import multipletests\n",
        "\n",
        "def calculate_enrichment_with_stats(adata, group_var, cluster_var, alpha=0.05):\n",
        "    # Contingency table of counts\n",
        "    contingency_table = pd.crosstab(adata.obs[cluster_var], adata.obs[group_var])\n",
        "    total_cells = contingency_table.values.sum()\n",
        "\n",
        "    # Observed and expected proportions\n",
        "    cluster_totals = contingency_table.sum(axis=1).values[:, None]  # Total in each cluster\n",
        "    group_totals = contingency_table.sum(axis=0).values[None, :]  # Total in each group\n",
        "    expected = cluster_totals * group_totals / total_cells  # Expected counts\n",
        "    enrichment = contingency_table / expected  # Observed-to-Expected ratio\n",
        "\n",
        "    # Perform Fisher's Exact Test for each cell\n",
        "    p_values = []\n",
        "    for cluster_idx in range(contingency_table.shape[0]):\n",
        "        for group_idx in range(contingency_table.shape[1]):\n",
        "            obs = contingency_table.iloc[cluster_idx, group_idx]\n",
        "            rest_cluster = cluster_totals[cluster_idx, 0] - obs\n",
        "            rest_group = group_totals[0, group_idx] - obs\n",
        "            rest_total = total_cells - obs - rest_cluster - rest_group\n",
        "\n",
        "            # Construct the 2x2 table for Fisher's test\n",
        "            table = np.array([[obs, rest_group], [rest_cluster, rest_total]])\n",
        "            _, p_value = fisher_exact(table, alternative=\"two-sided\")\n",
        "            p_values.append(p_value)\n",
        "\n",
        "    # FDR correction for multiple testing\n",
        "    p_values_corrected = multipletests(p_values, method=\"fdr_bh\")[1]\n",
        "\n",
        "    # Format results\n",
        "    p_corrected_matrix = pd.DataFrame(\n",
        "        np.array(p_values_corrected).reshape(contingency_table.shape),\n",
        "        index=contingency_table.index,\n",
        "        columns=contingency_table.columns,\n",
        "    )\n",
        "\n",
        "    return enrichment, p_corrected_matrix\n",
        "\n"
      ],
      "metadata": {
        "id": "Wj7a8C5dgbTa"
      },
      "id": "Wj7a8C5dgbTa",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def p_to_astk(p_val):\n",
        "        astk_range = [0.05,0.01,0.001,0.0001]\n",
        "        num_astk = [\"*\" for t in astk_range if p_val <= t ]\n",
        "        astk = \"\".join(num_astk)\n",
        "        if len(astk)==0:\n",
        "            astk='n.s.'\n",
        "        return astk\n"
      ],
      "metadata": {
        "id": "w-fWUSR2rxlm"
      },
      "id": "w-fWUSR2rxlm",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def plot_helicopter(df, pvalues, title, cbar_label=\"Enrichment\", ax=None):\n",
        "    \"\"\"\n",
        "    Plots a helicopter plot (heatmap) with enrichment values and p-values annotated.\n",
        "\n",
        "    Args:\n",
        "        df (pd.DataFrame): DataFrame containing enrichment values.\n",
        "        pvalues (pd.DataFrame): DataFrame containing corresponding p-values.\n",
        "        title (str): Title for the plot.\n",
        "        cbar_label (str): Label for the color bar.\n",
        "        ax (matplotlib.axes.Axes, optional): Axes to plot on. Creates a new figure if None.\n",
        "    \"\"\"\n",
        "    # Create a new figure only if ax is None\n",
        "    if ax is None:\n",
        "        fig, ax = plt.subplots(figsize=(2, 1.5))\n",
        "\n",
        "    # Calculate color scale limits for symmetry about 1\n",
        "    vrange = abs(df.values - 1).max()\n",
        "    vmin = 1 - vrange\n",
        "    vmax = 1 + vrange\n",
        "\n",
        "    # Function to format p-values\n",
        "    def format_pvalue(p):\n",
        "        if p < 0.001:\n",
        "            return f\"\\n(p={p:.1e})\"  # Scientific notation for p < 0.001\n",
        "        else:\n",
        "            return f\"\\n(p={p:.3f})\"  # Decimal format for p >= 0.001\n",
        "\n",
        "    # Initialize an empty DataFrame to hold the annotations\n",
        "    annotations = pd.DataFrame(index=df.index, columns=df.columns, dtype=object)\n",
        "\n",
        "    # Format the enrichment values and p-values\n",
        "    for i in range(df.shape[0]):\n",
        "        for j in range(df.shape[1]):\n",
        "            enrichment = f\"{df.iloc[i, j]:.2f}\"\n",
        "            pvalue = p_to_astk(pvalues.iloc[i, j])  # Use p_to_astk for p-value abbreviation\n",
        "            annotations.iloc[i, j] = enrichment + \"\\n\" + pvalue\n",
        "\n",
        "    sns.heatmap(\n",
        "        df,\n",
        "        annot=annotations.values,\n",
        "        fmt=\"\",\n",
        "        cmap=\"coolwarm\",\n",
        "        cbar_kws={\"label\": cbar_label},\n",
        "        vmin=vmin,\n",
        "        vmax=vmax,\n",
        "        ax=ax,  # Use the provided axes or create a new one\n",
        "    )\n",
        "    ax.set_title(title)\n",
        "    ax.set_xlabel(None)\n",
        "    ax.set_ylabel(\"Clusters\")\n",
        "    ax.set_xticklabels(ax.get_xticklabels(), rotation=45)\n",
        "\n",
        "    # Adjust layout only if no external ax is provided\n",
        "    if ax is None:\n",
        "        plt.tight_layout()\n",
        "        plt.show()\n",
        "\n",
        "    return ax\n",
        "\n",
        "\n",
        "\n",
        "# age_bin_enrichment, age_bin_corrected_p_values = calculate_enrichment_with_stats(DGC_adata, group_var=\"age_bin\", cluster_var=\"louvain\")\n",
        "# e_type_enrichment, e_type_corrected_p_values = calculate_enrichment_with_stats(DGC_adata, group_var=\"E_type\", cluster_var=\"louvain\")\n",
        "\n",
        "\n",
        "# # Generate helicopter plots\n",
        "# fig,ax = plt.subplots(1,2,figsize=(8,2))\n",
        "# plot_helicopter(age_bin_enrichment, age_bin_corrected_p_values, title=\"Helicopter Plot: Age Bin Enrichment Across Clusters\",ax=ax[0])\n",
        "# plot_helicopter(e_type_enrichment, e_type_corrected_p_values, title=\"Helicopter Plot: E_type Enrichment Across Clusters\",ax=ax[1])\n",
        "# plt.tight_layout()"
      ],
      "metadata": {
        "id": "oEd_fD_jbvta"
      },
      "id": "oEd_fD_jbvta",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def plot_enrichment_lines(enrichment, ax=None):\n",
        "    enrichment=enrichment.T\n",
        "    if ax is None:\n",
        "        fig, ax = plt.subplots(figsize=(10, 6))  # Create a new figure if no ax is provided\n",
        "\n",
        "    # Plotting the raw enrichment values as lines for each cluster\n",
        "    for cluster in enrichment.columns:\n",
        "        ax.plot(enrichment.index, enrichment[cluster], label=cluster,marker='o',linewidth=1,markersize=3)\n",
        "\n",
        "    # Adding labels and title\n",
        "    # ax.set_xlabel(\"E_type and Age_bin\")\n",
        "    ax.set_ylabel(\"Raw Enrichment\")\n",
        "    # ax.set_title(\"Raw Enrichment Across Clusters\")\n",
        "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
        "\n",
        "    # Add legend outside the plot\n",
        "    ax.legend(loc='upper left', bbox_to_anchor=(1.05, 1), title=\"Clusters\")\n",
        "\n",
        "    # Adjust layout to avoid overlap\n",
        "    plt.tight_layout()\n",
        "\n",
        "    return ax\n"
      ],
      "metadata": {
        "id": "D0BWWPuQDxzT"
      },
      "id": "D0BWWPuQDxzT",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "DGC_adata\n",
        "enrichment, corrected_p_values = calculate_enrichment_with_stats(DGC_adata, group_var=\"E_type_age_bin\", cluster_var=\"louvain\")\n",
        "\n",
        "\n",
        "\"\"\"\n",
        "E3 vs E4 Enrichment over time\n",
        "\"\"\"\n",
        "fig,ax = plt.subplots(1,2,figsize=(4,1.75), width_ratios=(1,1))\n",
        "\n",
        "subset = [c for c in enrichment.columns if 'fKI' not in c and 'E3' in c]\n",
        "plot_helicopter(enrichment[subset], corrected_p_values[subset], title=\"\" ,ax=ax[0])\n",
        "\n",
        "subset = [c for c in enrichment.columns if 'fKI' not in c and 'E4' in c]\n",
        "plot_helicopter(enrichment[subset], corrected_p_values[subset], title=\"\" ,ax=ax[1])\n",
        "\n",
        "plt.tight_layout()\n",
        "\n",
        "\"\"\"\n",
        "SynCre Compsre\n",
        "\"\"\"\n",
        "fig,ax = plt.subplots(1,2,figsize=(3,1.75), width_ratios=(1,1))\n",
        "\n",
        "subset = [c for c in enrichment.columns if '15m' in c and 'E3' in c]\n",
        "plot_helicopter(enrichment[subset], corrected_p_values[subset], title=\"\" ,ax=ax[0])\n",
        "\n",
        "subset = [c for c in enrichment.columns if '15m' in c and 'E4' in c]\n",
        "plot_helicopter(enrichment[subset], corrected_p_values[subset], title=\"\" ,ax=ax[1])\n",
        "\n",
        "plt.tight_layout()\n"
      ],
      "metadata": {
        "id": "2tecR7hRIJR2"
      },
      "id": "2tecR7hRIJR2",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "# enrichment_E3\n",
        "enrichment_E3 = enrichment[[c for c in enrichment.columns if 'fKI' not in c and 'E3' in c]]\n",
        "enrichment_E4 = enrichment[[c for c in enrichment.columns if 'fKI' not in c and 'E4' in c]]\n",
        "clusters = enrichment_s.index.to_list()\n",
        "num_clusters = len(clusters)\n",
        "\n",
        "fig,ax = plt.subplots(1,num_clusters, figsize=(1*num_clusters,1.2))\n",
        "for i,c in enumerate(clusters):\n",
        "    x = enrichment_s.columns\n",
        "    x = [a[:3] for a in x ]\n",
        "    ax[i].plot(x,enrichment_E3.loc[c],label='E3',color='b',marker='o')\n",
        "    ax[i].plot(x,enrichment_E4.loc[c],label='E4',color='r',marker='o')\n",
        "    ax[i].set_title(f'Cluster {c}')\n",
        "\n",
        "for a in ax:\n",
        "    a.set_ylim([.5,2])\n",
        "    a.axhline(1,linestyle='dotted',color='k')\n",
        "ax[0].legend()\n",
        "plt.tight_layout()\n"
      ],
      "metadata": {
        "id": "F1UdUNqE_K8r"
      },
      "id": "F1UdUNqE_K8r",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sc.tl.dendrogram(DGC_adata, groupby='louvain')\n",
        "ordered_clusters = DGC_adata.uns['dendrogram_louvain']['categories_ordered']\n",
        "\n",
        "# Extract top-ranked genes for each cluster\n",
        "top_n = 5\n",
        "ranked_genes = pd.DataFrame({\n",
        "    cluster: DGC_adata.uns['rank_genes_groups']['names'][cluster][:top_n]\n",
        "    for cluster in ordered_clusters\n",
        "})\n",
        "# display(ranked_genes.head())\n",
        "top_genes_all_clusters = ranked_genes.values.flatten().tolist()\n",
        "top_genes_all_clusters = list(set(top_genes_all_clusters))\n",
        "top_genes_all_clusters.append('Apoe')\n",
        "\n",
        "\n",
        "\n",
        "sc.pl.rank_genes_groups_dotplot(DGC_adata, n_genes=top_n)\n",
        "\n",
        "\n",
        "sc.pl.heatmap(DGC_adata, var_names=top_genes_all_clusters, groupby=\"louvain\", cmap=\"viridis\", dendrogram=False,figsize=(6, 1.5))\n",
        "\n",
        "\n",
        "sc.pl.stacked_violin(\n",
        "    DGC_adata,\n",
        "    var_names=top_genes_all_clusters,\n",
        "    groupby=\"louvain\",\n",
        "    swap_axes=False,\n",
        "    dendrogram=False,\n",
        "    figsize=(6,  1.5),\n",
        ")\n",
        "\n",
        "sc.pl.matrixplot(\n",
        "    DGC_adata,\n",
        "    var_names=top_genes_all_clusters,\n",
        "    groupby=\"louvain\",\n",
        "    dendrogram=False,\n",
        "    cmap=\"Blues\",\n",
        "    standard_scale=\"var\",\n",
        "    colorbar_title=\"column scaled\\nexpression\",\n",
        "    figsize=(6,  1.5),\n",
        ")\n",
        "\n",
        "print()"
      ],
      "metadata": {
        "id": "VxQn14SKfRGJ"
      },
      "id": "VxQn14SKfRGJ",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "fig,ax = plt.subplots(figsize=(2,1))\n",
        "sc.pl.violin(\n",
        "    DGC_adata,\n",
        "    keys='Apoe',\n",
        "    groupby=\"louvain\",\n",
        "    ax=ax\n",
        ")"
      ],
      "metadata": {
        "id": "ozF0g1Q31c8f"
      },
      "id": "ozF0g1Q31c8f",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "corr_method\n"
      ],
      "metadata": {
        "id": "hL-FU8KbmrJn"
      },
      "id": "hL-FU8KbmrJn",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "os.getcwd()"
      ],
      "metadata": {
        "id": "uRdJYdG7DusP"
      },
      "id": "uRdJYdG7DusP",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "marker_genes = DGC_adata.uns[\"rank_genes_groups\"]  # Assuming marker genes are stored here\n",
        "clus=4\n",
        "\n",
        "deg_df = pd.DataFrame(index=[row[clus] for row in marker_genes['names']],\n",
        "                      data={'scores':[ row[clus] for row in marker_genes['scores']],\n",
        "                            'pvals':[ row[clus] for row in marker_genes['pvals']],\n",
        "                            'pvals_adj':[ row[clus] for row in marker_genes['pvals_adj']],\n",
        "                            'logfoldchanges':[ row[clus] for row in marker_genes['logfoldchanges']],\n",
        "                            })\n",
        "deg_df.to_csv(f'Cluster_{clus}_deg_df.csv')\n",
        "fig_volcano, ax = plt.subplots(1, 1, figsize=(8, 4))\n",
        "single_deg_volcano(deg_df,min_p=-1e-30,maxlfc=4,ax=ax,title=f'Cluster {clus}')\n",
        "\n",
        "\n",
        "marker_genes = DGC_adata.uns[\"rank_genes_groups\"]  # Assuming marker genes are stored here\n",
        "clus=0\n",
        "\n",
        "deg_df = pd.DataFrame(index=[row[clus] for row in marker_genes['names']],\n",
        "                      data={'scores':[ row[clus] for row in marker_genes['scores']],\n",
        "                            'pvals':[ row[clus] for row in marker_genes['pvals']],\n",
        "                            'pvals_adj':[ row[clus] for row in marker_genes['pvals_adj']],\n",
        "                            'logfoldchanges':[ row[clus] for row in marker_genes['logfoldchanges']],\n",
        "                            })\n",
        "deg_df.to_csv(f'Cluster_{clus}_deg_df.csv')\n",
        "fig_volcano, ax = plt.subplots(1, 1, figsize=(8, 4))\n",
        "single_deg_volcano(deg_df,min_p=-1e-30,maxlfc=4,ax=ax,title=f'Cluster {clus}')"
      ],
      "metadata": {
        "id": "zhNotB2GrcVj"
      },
      "id": "zhNotB2GrcVj",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "F3E2EhUCnIBW"
      },
      "id": "F3E2EhUCnIBW",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def deg_set_up(adata,keys_of_interest):\n",
        "    combo_key = '_'.join(keys_of_interest)\n",
        "    adata.obs[combo_key] = adata.obs[keys_of_interest[0]]\n",
        "    for i,v in enumerate(keys_of_interest[1:]):\n",
        "        # to_add = adata.obs[v]\n",
        "        adata.obs[combo_key] = adata.obs[combo_key].str.cat(adata.obs[v].astype(str),sep='_')\n",
        "    return adata,combo_key\n",
        "\n",
        "\n",
        "def deg_analysis(adata_temp,new_file_path,analysis_pairs,group_key,adata_name='volcano',log2fc_extrema=[-15,15],label_cutoff=1,to_plot=True):\n",
        "    '''\n",
        "    Takes in anndata RNA object and a stratifying key and returns\n",
        "    DEGs based on given grouppings. Also builds volcano plots.    '''\n",
        "    adata_temp.raw = adata_temp\n",
        "    '''Perform statistics and write to DF'''\n",
        "    # print(os.getcwd())\n",
        "    deg_df_dict={}\n",
        "    for key,pair in analysis_pairs.items():\n",
        "        sc.tl.rank_genes_groups(adata_temp,use_raw=True,groupby=group_key,groups=pair,reference=pair[1],key_added=key,method='wilcoxon',tie_correct=True)\n",
        "        deg_df = sc.get.rank_genes_groups_df(adata_temp, group=pair[0],key=key,pval_cutoff=1,log2fc_min=log2fc_extrema[0], log2fc_max=log2fc_extrema[1])\n",
        "        deg_df = deg_df.set_index('names')\n",
        "        deg_df_dict[key] = deg_df\n",
        "        new_file_name = os.path.join(new_file_path,adata_name+'_'+key+'.csv')\n",
        "        print(new_file_name)\n",
        "        deg_df.to_csv(new_file_name)\n",
        "    return adata_temp, deg_df_dict\n",
        "\n",
        "def deg_volc_plot(deg_df_dict,gene_list=[],label_cutoff=0.1,suptitle='',lock_y_max=True,special_genes=[],use_adj=False):\n",
        "    ''' Plot Volcanoes '''\n",
        "    fig_volcano,ax=plt.subplots(1,len(deg_df_dict),figsize=(len(deg_df_dict)*1.5,2))\n",
        "    if len(deg_df_dict) == 1:  ax=[ax]\n",
        "    for key,df in deg_df_dict.items():\n",
        "        key_ind = list(deg_df_dict.keys()).index(key)\n",
        "        ax[key_ind].set_title(key.replace('_','\\n'))\n",
        "        genes= df.index\n",
        "        lfc = df['logfoldchanges']\n",
        "        if use_adj: pval = df['pvals_adj']\n",
        "        else: pval = df['pvals']\n",
        "\n",
        "        neg_log10_pval = -np.log10(pval)\n",
        "        for i,g in enumerate(genes):\n",
        "            if str(g) in gene_list:\n",
        "                if not np.isnan(lfc[i]*neg_log10_pval[i]):\n",
        "                    ax[key_ind].scatter(lfc[i],neg_log10_pval[i],c='k',s=3)\n",
        "                    if pval[i]<label_cutoff:\n",
        "                        ax[key_ind].text(lfc[i],neg_log10_pval[i],genes[i],rotation=45,fontsize=4)\n",
        "\n",
        "                if g in special_genes:\n",
        "                    if not np.isnan(lfc[i]*neg_log10_pval[i]):\n",
        "                        ax[key_ind].scatter(lfc[i],neg_log10_pval[i],c='r',s=3)\n",
        "                        ax[key_ind].text(lfc[i],neg_log10_pval[i],genes[i],rotation=45,fontsize=4,color='r')\n",
        "    x_lim=[]\n",
        "    y_lim=[]\n",
        "    for a in ax:\n",
        "        x_lim.append(a.get_xlim())\n",
        "        y_lim.append(a.get_ylim())\n",
        "\n",
        "    x_etr = np.max(np.abs(x_lim))\n",
        "    y_etr = np.max(np.abs(y_lim))\n",
        "\n",
        "    for a in ax:\n",
        "        a.set_xlim(-x_etr*1.1,x_etr*1.1)\n",
        "        if lock_y_max: a.set_ylim(0,y_etr*1.1)\n",
        "        else: a.set_ylim(0,a.get_ylim()[1]*1.1)\n",
        "        a.set_xlabel('Log2 Fold Change')\n",
        "        if use_adj: a.set_ylabel('-log10(adj-pvalue)')\n",
        "        else: a.set_ylabel('-log10(raw-pvalue)')\n",
        "        a.axhline(-np.log10(0.05),c='k',linestyle=\":\",linewidth=.6,label='p=0.05')\n",
        "\n",
        "    fig_volcano.suptitle(suptitle)\n",
        "    plt.tight_layout()\n",
        "    return fig_volcano,ax\n",
        "\n",
        "def deg_heatmap(deg_dict,gene_list=[],ax=None,lfcmax = 2.5,ax_cbar=None,suptitle=''):\n",
        "\n",
        "    try: ax.grid(visible=False)\n",
        "    except:\n",
        "        fig,axs = plt.subplots(1,2,figsize=(1+.5*len(deg_dict.keys()),len(gene_list)/14),width_ratios=[9,.5])\n",
        "\n",
        "        ax=axs[0]\n",
        "        ax_cbar = axs[1]\n",
        "        ax.grid(visible=False)\n",
        "    if len(gene_list)==0:gene_list = list()\n",
        "\n",
        "    heat_map_null=np.zeros([len(gene_list),len(deg_dict)])\n",
        "    heat_map_lfc = heat_map_null.copy()*np.nan\n",
        "    heat_map_p = heat_map_null.copy()*np.nan\n",
        "    for key,deg_df in deg_dict.items():\n",
        "        key_id = list(deg_dict.keys()).index(key)\n",
        "        for gene in gene_list:\n",
        "            i_gene = gene_list.index(gene)\n",
        "            try:\n",
        "                heat_map_lfc[i_gene,key_id] = deg_df.loc[gene,'logfoldchanges']\n",
        "            except:\n",
        "                None\n",
        "    cbar = ax.pcolorfast(heat_map_lfc,vmin=-lfcmax,vmax=lfcmax,cmap='bwr')\n",
        "    ax.set_yticks(np.arange(len(gene_list))+0.5, labels=gene_list,rotation=0)\n",
        "    ax.tick_params(length=0)\n",
        "\n",
        "    fmt_xlabels = list(deg_dict.keys())\n",
        "    fmt_xlabels = [xl.replace('_','\\n') for xl in fmt_xlabels]\n",
        "    ax.set_xticks(np.arange(len(deg_dict.keys()))+0.5, labels=fmt_xlabels,rotation=0)\n",
        "    ax.grid(visible=False)\n",
        "\n",
        "\n",
        "    for key,deg_df in deg_dict.items():\n",
        "        key_id = list(deg_dict.keys()).index(key)\n",
        "        for gene in gene_list:\n",
        "            i_gene = gene_list.index(gene)\n",
        "            try:\n",
        "                p = deg_df.loc[gene,'pvals']\n",
        "                if p<.1:\n",
        "                    p_str = str(p)[:5]\n",
        "                    l2fc = str(2**deg_df.loc[gene,'logfoldchanges'])[:3]\n",
        "                    tag = f\"{l2fc}X; p={p_str}\"\n",
        "                    tag = pval_to_star(p)\n",
        "                    ax.text(key_id+.5,i_gene+.5,tag,ha='center',va='top',rotation=0)\n",
        "            except: None\n",
        "\n",
        "    ax.tick_params(top=True, labeltop=True, bottom=False, labelbottom=False)\n",
        "\n",
        "    try:\n",
        "        plt.colorbar(cbar,cax=ax_cbar)\n",
        "        ax_cbar.set_ylabel('log2(Fold Change)')\n",
        "    except: plt.colorbar(cbar)\n",
        "    plt.suptitle(suptitle)\n",
        "    plt.tight_layout()\n",
        "    return plt.gcf(), ax, cbar\n",
        "\n",
        "def pval_to_star(p,specifics={(.05,.1):'\\'',(.05,1):'ns'},max_star=5):\n",
        "    for range,tag in specifics.items():\n",
        "        if p<=np.max(range) and p>np.min(range): return tag\n",
        "    if p<=0.5 and p>0.01: return '*'\n",
        "    nlogp=-np.log10(p)\n",
        "    return ('*'*int(nlogp))[:max_star]"
      ],
      "metadata": {
        "id": "hru563c6bCFU"
      },
      "id": "hru563c6bCFU",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "\n",
        "def single_deg_volcano(\n",
        "    deg_df,\n",
        "    label_cutoff=0.1,\n",
        "    title=None,\n",
        "    comp_string=None,\n",
        "    lock_y_max=True,\n",
        "    special_genes=[],\n",
        "    use_adj=False,\n",
        "    marker_size = .5,\n",
        "    min_p = 1e-30,\n",
        "    maxlfc =10,\n",
        "    ax=None):\n",
        "    '''\n",
        "    Plot a single volcano plot for a DE comparison.\n",
        "\n",
        "    Parameters:\n",
        "        deg_df (DataFrame): DataFrame of DE results with gene names as index and columns including\n",
        "                            'logfoldchanges', 'pvals', and optionally 'pvals_adj'.\n",
        "        gene_list (list, optional): List of genes to be highlighted in the plot (default is an empty list).\n",
        "        label_cutoff (float, optional): Minimum -log10(p-value) threshold for labeling genes (default is 0.1).\n",
        "        title (str, optional): Title for the plot (default is an empty string).\n",
        "        lock_y_max (bool, optional): If True, locks the y-axis max across subplots; adjust to data range otherwise.\n",
        "        special_genes (list, optional): Genes to be specially highlighted in red (default is an empty list).\n",
        "        use_adj (bool, optional): If True, use adjusted p-values; otherwise, use raw p-values (default is False).\n",
        "\n",
        "    Returns:\n",
        "        fig_volcano (Figure): Matplotlib figure object for the volcano plot.\n",
        "        ax (Axes): Matplotlib axes object for the volcano plot.\n",
        "    '''\n",
        "\n",
        "    # Initialize the figure and axis for the volcano plot\n",
        "    if ax is None:\n",
        "        fig_volcano, ax = plt.subplots(1, 1, figsize=(1.75, 1.75))\n",
        "    else:\n",
        "        fig_volcano = ax.figure\n",
        "\n",
        "    # Set the title for the plot\n",
        "    if title is not None:\n",
        "        ax.set_title(title.replace('_', '\\n'))\n",
        "\n",
        "    # Extract necessary data\n",
        "    genes = deg_df.index\n",
        "    lfc = deg_df['logfoldchanges']\n",
        "    pval = deg_df['pvals_adj'] if use_adj else deg_df['pvals']\n",
        "    pval[pval<min_p]=min_p\n",
        "    lfc[lfc<-maxlfc]=-maxlfc\n",
        "    lfc[lfc>maxlfc]=maxlfc\n",
        "    neg_log10_pval = -np.log10(pval)\n",
        "\n",
        "    # Plot each gene, marking those in gene_list and special_genes\n",
        "    ax.scatter(lfc, neg_log10_pval, c='gray', s=marker_size)\n",
        "\n",
        "    is_sig = neg_log10_pval>4\n",
        "    high_lfc = abs(lfc)>=1\n",
        "\n",
        "    selected = np.logical_and(is_sig,high_lfc)\n",
        "\n",
        "    lfc_sig = lfc[selected]\n",
        "    neg_log10_pval_sig = neg_log10_pval[selected]\n",
        "    genes_sig = genes[selected]\n",
        "\n",
        "    for i, g in enumerate(genes_sig):\n",
        "        if not np.isnan(lfc_sig[i] * neg_log10_pval_sig[i]):\n",
        "            ax.scatter(lfc_sig[i], neg_log10_pval_sig[i], c='r', s=marker_size)\n",
        "            ax.text(lfc_sig[i], neg_log10_pval_sig[i], genes_sig[i], rotation=45, fontsize=4, color='r')\n",
        "\n",
        "    # Set axis limits and labels\n",
        "    x_lim = ax.get_xlim()\n",
        "    y_lim = ax.get_ylim()\n",
        "    x_etr = np.max(np.abs(x_lim))\n",
        "    y_etr = np.max(np.abs(y_lim))\n",
        "\n",
        "    ax.set_xlim(-x_etr * 1.1, x_etr * 1.1)\n",
        "    if lock_y_max:\n",
        "        ax.set_ylim(0, y_etr * 1.1)\n",
        "    else:\n",
        "        ax.set_ylim(0, ax.get_ylim()[1] * 1.1)\n",
        "\n",
        "    if comp_string is not None:\n",
        "        xlabel = f'{comp_string}\\nLog2(fold change)'\n",
        "    else:\n",
        "        xlabel = 'Log2(fold change)'\n",
        "    ax.set_xlabel(xlabel)\n",
        "    ax.set_ylabel('-log10(adj-pvalue)' if use_adj else '-log10(raw-pvalue)')\n",
        "    ax.axhline(-np.log10(0.05), c='k', linestyle=\":\", linewidth=0.6, label='p=0.05')\n",
        "    ax.axvline(0, c='k', linestyle=\"-\", linewidth=0.6, label='p=0.05')\n",
        "\n",
        "    plt.tight_layout()\n",
        "\n",
        "    return fig_volcano, ax\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "H0eyzWEVJ0Ip"
      },
      "id": "H0eyzWEVJ0Ip",
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "private_outputs": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.4"
    },
    "varInspector": {
      "cols": {
        "lenName": 16,
        "lenType": 16,
        "lenVar": 40
      },
      "kernels_config": {
        "python": {
          "delete_cmd_postfix": "",
          "delete_cmd_prefix": "del ",
          "library": "var_list.py",
          "varRefreshCmd": "print(var_dic_list())"
        },
        "r": {
          "delete_cmd_postfix": ") ",
          "delete_cmd_prefix": "rm(",
          "library": "var_list.r",
          "varRefreshCmd": "cat(var_dic_list()) "
        }
      },
      "types_to_exclude": [
        "module",
        "function",
        "builtin_function_or_method",
        "instance",
        "_Feature"
      ],
      "window_display": false
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}